{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import gzip\n",
    "from bs4 import BeautifulSoup\n",
    "import lxml\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "import requests\n",
    "import random\n",
    "import sys\n",
    "import pathlib\n",
    "import csv\n",
    "\n",
    "\n",
    "\n",
    "import argparse\n",
    "\n",
    "# import multiprocessing\n",
    "from fuzzywuzzy import fuzz\n",
    "\n",
    "# url = 'http://ai-capo-api-lb/spaCy_ner_predictor?text_sentence='  # Load Balancer\n",
    "\n",
    "# result_path = '/nfs/gns/literature/machine-learning/evaluation/FP_Analysis/SpaCy/model_call/'\n",
    "# pathlib.Path(result_path).mkdir(parents=True, exist_ok=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/gns/literature/Santosh_Tirunagari/miniconda3/condabin/conda\r\n"
     ]
    }
   ],
   "source": [
    "!which conda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import en_ner_europepmc_md\n",
    "import en_relationv01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "relation_model1 = en_relationv01.load()\n",
    "relation_model2 = en_ner_europepmc_md.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(1, '/nfs/gns/literature/machine-learning/Santosh/Gitlab/biobertepmc/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import torch\n",
    "from torch.utils.data import DataLoader\n",
    "import pickle\n",
    "from biobert.model.bert_crf_model import BertCRF\n",
    "from biobert.data_loader.epmc_loader import NERDatasetBatch\n",
    "from biobert.utils.utils import my_collate\n",
    "from nltk.tokenize import WordPunctTokenizer\n",
    "from collections import namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_file_path = '/nfs/misc/literature/rdf_annotation_data/daily_pipeline_api/15_08_1947/fulltext/job_14/annotation/patch-total-417.xml.gz'\n",
    "\n",
    "# data_file_path = '/nfs/misc/literature/rdf_annotation_data/daily_pipeline_api/15_08_1947/abstract/job_21/annotation/patch-total-415.abstract.gz'\n",
    "\n",
    "# def getfileblocks(file_path):\n",
    "#     subFileBlocks = []\n",
    "\n",
    "#     with gzip.open(file_path, 'rt') as fh: #gzip.open\n",
    "#         for line in fh:\n",
    "#             if line.startswith('<!DOCTYPE article'):  # <!DOCTYPE \"JATS-archivearticle1.dtd\">\n",
    "#                 subFileBlocks.append(line)\n",
    "#             else:\n",
    "#                 subFileBlocks[-1] += line\n",
    "\n",
    "#     return subFileBlocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "data_file_path = '/nfs/production/literature/shyama/FullText20.11/Annot_PMC13900_PMC548680.xml'\n",
    "\n",
    "\n",
    "def getfileblocks(file_path):\n",
    "    subFileBlocks = []\n",
    "\n",
    "    with io.open(file_path, 'r', encoding='utf8') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith('<!DOCTYPE'):\n",
    "                subFileBlocks.append(line)\n",
    "            else:\n",
    "                subFileBlocks[-1] += line\n",
    "\n",
    "    return subFileBlocks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "files_list = getfileblocks(data_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [],
   "source": [
    "soup = BeautifulSoup(files_list[101], 'lxml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_size = 16\n",
    "class MLModel:\n",
    "    def __init__(self):\n",
    "        self.bertCrf_model = load_model()\n",
    "\n",
    "        # bertCrf_model.load_state_dict(torch.load('/homes/yangx/home/gitrepo/biobertepmc/model/bert_crf_model.states', map_location=device))\n",
    "        self.bertCrf_model.load_state_dict(torch.load(MODEL_PATH + 'bert_crf_model.states', map_location=device))\n",
    "        self.bertCrf_model.bert_model.bert_model.to(device)\n",
    "\n",
    "    def post(self, sentences):\n",
    "        BATCH_SIZE = 16\n",
    "        text = sentences\n",
    "        # print(text)\n",
    "        with torch.no_grad():\n",
    "            processor, tokens, spans = load_data_processor(text)\n",
    "            dataLoader = DataLoader(dataset=processor, batch_size=BATCH_SIZE, collate_fn=my_collate, num_workers=2)\n",
    "\n",
    "            idx2label = params['idx2label']\n",
    "            self.bertCrf_model.eval()\n",
    "            entities = []\n",
    "            for i_batch, sample_batched in enumerate(dataLoader):\n",
    "                inputs = sample_batched['input']\n",
    "\n",
    "                bert_inputs, bert_attention_mask, bert_token_mask, wordpiece_alignment, split_alignments, lengths, token_mask \\\n",
    "                    = processor.tokens_totensor(inputs)\n",
    "\n",
    "                _, preds = self.bertCrf_model.predict(input_ids=bert_inputs.to(device),\n",
    "                                                      bert_attention_mask=bert_attention_mask.to(device),\n",
    "                                                      bert_token_mask=bert_token_mask,\n",
    "                                                      alignment=wordpiece_alignment,\n",
    "                                                      splits=(split_alignments, lengths),\n",
    "                                                      token_mask=token_mask)\n",
    "                if idx2label:\n",
    "                    for i, (path, score) in enumerate(preds):\n",
    "                        labels = [idx2label[p] for p in path]\n",
    "                        offset_index = i_batch * BATCH_SIZE + i\n",
    "                        entities.append([[e.span[0], e.span[1], e.tag, e.text]\n",
    "                                         for e in extract_entity(labels, spans[offset_index], text[offset_index])])\n",
    "        return {'annotations': entities}\n",
    "\n",
    "\n",
    "def load_data_processor(inputs):\n",
    "    token_spans = []\n",
    "    tokens = []\n",
    "    for line in inputs:\n",
    "        token_spans.append(list(tokenizer.span_tokenize(line)))\n",
    "        tokens.append([line[start: end] for start, end in token_spans[-1]])\n",
    "\n",
    "    processor = NERDatasetBatch.from_params(params=params, inputs=tokens)\n",
    "    return processor, tokens, token_spans\n",
    "\n",
    "\n",
    "def load_model():\n",
    "    allowed_transitions = None\n",
    "    model = BertCRF(num_tags=params['num_tags'],\n",
    "                    model_name=params['model_name'],\n",
    "                    stride=params['stride'],\n",
    "                    include_start_end_transitions=True,\n",
    "                    constraints=allowed_transitions)\n",
    "    return model\n",
    "\n",
    "\n",
    "def extract_entity(preds, spans, text, length=20):\n",
    "    \"\"\"\n",
    "    extract entity from label sequence\n",
    "    :param preds: a list of labels in a sentence\n",
    "    :type preds: List[str\n",
    "    :param spans:\n",
    "    :type spans:\n",
    "    :return: A list of entity object\n",
    "    :rtype: List[Entity]\n",
    "    \"\"\"\n",
    "    entities = []\n",
    "    tmp = []\n",
    "\n",
    "    for i, token in enumerate(preds):\n",
    "        if token == 'O':\n",
    "            pos, tag = 'O', 'O'\n",
    "            label = None\n",
    "        else:\n",
    "            pos, tag = token.split('-')\n",
    "            label = Entity_Label(index=i, pos=pos, tag=tag, span=spans[i])\n",
    "\n",
    "        if pos in {'B', 'O'} and tmp:\n",
    "            start_span = tmp[0].span[0]\n",
    "            end_span = tmp[-1].span[1]\n",
    "            entities.append(Entity(span=(start_span, end_span),\n",
    "                                   tag=tmp[0].tag,\n",
    "                                   text=text[start_span:end_span],\n",
    "                                   pre=text[max(0, start_span - length):start_span],\n",
    "                                   post=text[end_span: end_span + length]))\n",
    "            tmp[:] = []\n",
    "        if pos == 'B' or pos == 'I':\n",
    "            tmp.append(label)\n",
    "\n",
    "    if tmp:\n",
    "        start_span = tmp[0].span[0]\n",
    "        end_span = tmp[-1].span[-1]\n",
    "        entities.append(\n",
    "            Entity(span=(start_span, end_span),\n",
    "                   tag=tmp[0].tag,\n",
    "                   text=text[start_span:end_span],\n",
    "                   pre=text[max(0, start_span - length):start_span],\n",
    "                   post=text[end_span:end_span + length])\n",
    "        )\n",
    "    return entities\n",
    "\n",
    "\n",
    "# Create a function called \"chunks\" with two arguments, l and n:\n",
    "def chunks(l, n):\n",
    "    # For item i in a range that is a length of l,\n",
    "    for i in range(0, len(l), n):\n",
    "        # Create an index range for l of n items:\n",
    "        yield l[i:i + n]\n",
    "\n",
    "\n",
    "def clean_Nones(ner_tags_):\n",
    "    ner_tags = []\n",
    "    # had to do this as the position of entity tag and entity are exchanged in CD\n",
    "    for each_ner_tag in ner_tags_:\n",
    "        if 'CD' == each_ner_tag[2]:\n",
    "            ner_tags.append([each_ner_tag[0], each_ner_tag[1], each_ner_tag[3], each_ner_tag[2]])\n",
    "        else:\n",
    "            ner_tags.append(each_ner_tag)\n",
    "\n",
    "    ner_tags = sorted(ner_tags, key=lambda x: len(x[3]), reverse=True)\n",
    "    if len(ner_tags) == 1 and 'None' in ner_tags:\n",
    "        return ner_tags\n",
    "    elif len(ner_tags) > 1 and 'None' in ner_tags:\n",
    "        ner_tags.remove('None')\n",
    "        return ner_tags\n",
    "    else:\n",
    "        return ner_tags\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_sentences = soup.find_all('sent')# all_sentences = soup.find_all('SENT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<!DOCTYPE >\n",
       "<html><body><p>JATS-archivearticle1.dtd\"&gt;\n",
       "<ebiroot xmlns:z=\"ebistuff\"><article article-type=\"research-article\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><front><journal-meta><journal-id journal-id-type=\"nlm-ta\">Crit Care</journal-id><journal-title>Critical Care</journal-title><issn pub-type=\"ppub\">1364-8535</issn><issn pub-type=\"epub\">1466-609X</issn><publisher><publisher-name>BioMed Central</publisher-name></publisher></journal-meta><article-meta><article-id pub-id-type=\"pmcid\">29020</article-id><article-id pub-id-type=\"publisher-id\">cc-3-4-091</article-id><article-id pub-id-type=\"pmid\">11056730</article-id><article-categories><subj-group subj-group-type=\"heading\"><subject>Research Paper</subject></subj-group></article-categories><title-group><article-title><text><sent pm=\".\" sid=\"40409\"><plain>Extracorporeal membrane oxygenation for overwhelming Blastomyces dermatitidis <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> </plain></sent>\n",
       "</text></article-title></title-group><contrib-group><contrib contrib-type=\"author\" id=\"A1\"><name><surname>Dalton</surname><given-names>Heidi J</given-names></name><xref ref-type=\"aff\" rid=\"I1\">1</xref></contrib><contrib contrib-type=\"author\" id=\"A2\"><name><surname>Hertzog</surname><given-names>James H</given-names></name><xref ref-type=\"aff\" rid=\"I1\">1</xref></contrib><contrib contrib-type=\"author\" id=\"A3\"><name><surname>Hannan</surname><given-names>Robert L</given-names></name><xref ref-type=\"aff\" rid=\"I2\">2</xref></contrib><contrib contrib-type=\"author\" id=\"A4\"><name><surname>Vezza</surname><given-names>Phyllis</given-names></name><xref ref-type=\"aff\" rid=\"I3\">3</xref></contrib><contrib contrib-type=\"author\" id=\"A5\"><name><surname>Hauser</surname><given-names>Gabriel J</given-names></name><xref ref-type=\"aff\" rid=\"I1\">1</xref></contrib></contrib-group><aff id=\"I1\"><label>1</label>Department of Pediatrics, Division of Pediatric Critical Care\n",
       "\t\t\t\tMedicine, Georgetown University Medical Center,\n",
       "\t\t\t\tWashington DC, USA.</aff><aff id=\"I2\"><label>2</label>Department of Surgery, Division of Cardiothoracic Surgery, Georgetown University Medical Center, Washington DC, USA.\n",
       "\t\t\t\t</aff><aff id=\"I3\"><label>3</label>Department of Pathology, Georgetown University Medical Center,\n",
       "\t\t\t\tWashington DC, USA.</aff><pub-date pub-type=\"ppub\"><year>1999</year></pub-date><pub-date pub-type=\"epub\"><day>19</day><month>7</month><year>1999</year></pub-date><volume>3</volume><issue>4</issue><fpage>91</fpage><lpage>94</lpage><ext-link ext-link-type=\"uri\" xlink:href=\"http://ccforum.com\"></ext-link><history><date date-type=\"received\"><day>29</day><month>10</month><year>1998</year></date><date date-type=\"rev-request\"><day>20</day><month>4</month><year>1999</year></date><date date-type=\"rev-recd\"><day>28</day><month>5</month><year>1999</year></date><date date-type=\"accepted\"><day>28</day><month>6</month><year>1999</year></date></history><permissions><copyright-statement>Copyright © 1999 Current Science Ltd</copyright-statement><copyright-year>1999</copyright-year><copyright-holder>Current Science Ltd</copyright-holder></permissions><kwd-group><kwd>ECMO</kwd><kwd>fungal infection</kwd><kwd>blastomycosis</kwd><kwd>respiratory failure</kwd><kwd>extracorporeal life support</kwd><kwd>pneumonia</kwd><kwd>ARDS</kwd></kwd-group></article-meta></front><sec><title><text><sent pm=\".\" sid=\"40410\"><plain>Abstract </plain></sent>\n",
       "</text></title><p><text><sent pm=\".\" sid=\"40411\"><plain>Blastomyces dermatitidis is a fungus that is found primarily in endemic areas of the midwestern and southcentral USA. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40412\"><plain><z:efo cat=\"disease\" ids=\"EFO_0007174\">Blastomycosis</z:efo> <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> may develop after the inhalation of spores. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40413\"><plain>While <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> may lead to acute or chronic <z:efo cat=\"disease\" ids=\"EFO_1001991\">pneumonitis</z:efo>, it will rarely result in the development of the <z:efo cat=\"disease\" ids=\"EFO_1000637\">acute respiratory distress syndrome</z:efo> (<z:efo cat=\"disease\" ids=\"EFO_1000637\">ARDS</z:efo>). </plain></sent>\n",
       "<sent pm=\".\" sid=\"40414\"><plain>In this situation, mortality rates are 50–80%. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40415\"><plain>Patients who survive, however, commonly show good recovery of pulmonary function. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40416\"><plain>Extracorporeal membrane oxygenation (ECMO), a modified form of cardiopulmonary bypass that allows systemic perfusion, oxygenation, and carbon dioxide removal, may be used to support patients with cardiorespiratory failure that is refractory to conventional therapies. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40417\"><plain>The use of ECMO allows the reduction of high levels of mechanical ventilatory support that may cause iatrogenic <z:efo cat=\"disease\" ids=\"EFO_0000546\">injury</z:efo> to the diseased lung. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40418\"><plain>Fungal diseases are often contraindications for ECMO use, since systemic fungal organisms may bind to the ECMO circuit and be difficult to eradicate. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40419\"><plain>It may be reasonable to utilize ECMO, however, if the fungal <z:efo cat=\"disease\" ids=\"EFO_0005741\">infection</z:efo> is isolated to the respiratory system. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40420\"><plain>We report the case of a patient with <z:efo cat=\"disease\" ids=\"EFO_1000637\">ARDS</z:efo> secondary to <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> who was treated with ECMO due to the failure of conventional cardiorespiratory supports. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40421\"><plain>To our knowledge, this is the first report of the use of ECMO for this condition. </plain></sent>\n",
       "</text></p></sec><sec><title><text><sent pm=\".\" sid=\"40422\"><plain>Introduction </plain></sent>\n",
       "</text></title><p><text><sent pm=\".\" sid=\"40423\"><plain>Blastomyces dermatitidis, a fungus that is endemic to the midwestern and southcentral USA, is a relatively common cause of pulmonary and <z:efo cat=\"disease\" ids=\"MONDO_0000254\">cutaneous mycosis</z:efo> in these areas. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40424\"><plain>Both respiratory and extrapulmonary <z:efo cat=\"disease\" ids=\"EFO_0005741\">infections</z:efo> are commonly initiated with the inhalation of fungal spores. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40425\"><plain>Following the introduction of amphotericin B, the fatality rate has fallen from 78% to between 21–28%. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40426\"><plain>If a total dose of 2 g amphotericin B is given, the cure rate may rise to 97% [1]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40427\"><plain>However, if <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> is associated with the development of the <z:efo cat=\"disease\" ids=\"EFO_1000637\">acute respiratory distress syndrome</z:efo> (<z:efo cat=\"disease\" ids=\"EFO_1000637\">ARDS</z:efo>), the mortality rate rises to 50%–80% [2]. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40428\"><plain>Extracorporeal membrane oxygenation (ECMO) has been used successfully to support newborn, pediatric and adult patients with cardiac and <z:efo cat=\"disease\" ids=\"EFO_0000684\">respiratory diseases</z:efo> that are refractory to conventional therapy. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40429\"><plain>Survival rates of 80% have been achieved in the 13000 neonates who have been supported with ECMO, while survival in the 1500 pediatric patients listed in the International Extracorporeal Life Support Registry who have received ECMO for <z:efo cat=\"disease\" ids=\"EFO_0009686\">respiratory failure</z:efo> is 53%. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40430\"><plain>Of the 547 adults listed in the registry, 47% have survived [3]. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40431\"><plain>In general, a primary fungal <z:efo cat=\"disease\" ids=\"EFO_0005741\">infection</z:efo>, especially when the <z:efo cat=\"disease\" ids=\"EFO_0005741\">infection</z:efo> is hematogenous, is a contraindication to the use of ECMO. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40432\"><plain>Historically, patients with hematogenous fungal <z:efo cat=\"disease\" ids=\"EFO_0005741\">infections</z:efo> do poorly on ECMO, primarily because the <z:efo cat=\"disease\" ids=\"MONDO_0002041\">mycosis</z:efo> cannot be eradicated from the synthetic surfaces of the ECMO circuit. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40433\"><plain>The restriction against ECMO support for cardiorespiratory failure secondary to fungal <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> without hematogenous spread is less certain. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40434\"><plain>In theory, this condition should not preclude the use of ECMO, but no reports about the use of ECMO in this situation are available. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40435\"><plain>We report our experience with the use of ECMO to support a previously healthy adolescent with <z:efo cat=\"disease\" ids=\"EFO_0009686\">respiratory failure</z:efo> secondary to <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> to raise awareness of this disease as a cause of <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> and better understand the use and limitation of extracorporeal support in this condition. </plain></sent>\n",
       "</text></p></sec><sec><title><text><sent pm=\".\" sid=\"40436\"><plain>Case summary </plain></sent>\n",
       "</text></title><p><text><sent pm=\".\" sid=\"40437\"><plain>A 15-year-old, 100kg black male developed <z:efo cat=\"phenotype\" ids=\"HP_0001945\">fever</z:efo>, <z:efo cat=\"phenotype\" ids=\"HP_0012735\">cough</z:efo>, and <z:efo cat=\"phenotype\" ids=\"HP_0002315\">headache</z:efo> 12 days prior to admission to Georgetown University Hospital. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40438\"><plain>Prior to this, the patient had been in good health with no evidence of <z:efo cat=\"disease\" ids=\"MONDO_0021094\">immunodeficiency</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40439\"><plain>There was no history of unusual travel or exposure to unusual environments. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40440\"><plain>He was treated with amoxicillin for presumed <z:efo cat=\"disease\" ids=\"EFO_0007486\">sinusitis</z:efo> but his symptoms persisted and he was hospitalized for 2 days with erythromycin treatment for presumed atypical <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40441\"><plain>His chest radiograph revealed a left lower lobe infiltrate. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40442\"><plain>After hospital discharge, the patient's private physician prescribed amantadine after his condition failure to improve. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40443\"><plain>The patient was readmitted to the hospital 3 days later with respiratory distress and complete opacification of the right hemithorax and left lower hemithorax on chest radiograph (Fig. 1). </plain></sent>\n",
       "<sent pm=\".\" sid=\"40444\"><plain>Emergent intubation and institution of mechanical ventilatory support was performed. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40445\"><plain>Despite supportive care and the institution of broad-spectrum antibiotic therapy, the patient's <z:efo cat=\"disease\" ids=\"EFO_0009686\">respiratory failure</z:efo> progressed. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40446\"><plain>Mean airway pressures of 30–40 cmH2O 48 h after his readmission were required to maintain arterial partial pressure of carbon dioxide (PaCO2) 70torr and oxygen (PaO2) 50torr. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40447\"><plain>Calculation of the oxygenation index (OI), a measure that relates airway pressure and PaO2 to presumed mortality, revealed that his OI was 36. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40448\"><plain>OI is the product of mean airway pressure (MAP) and fraction of inspired oxygen concentration (FiO2) divided by the PaO2 [OI=(MAP × FiO2 × 100)/PaO2]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40449\"><plain>Treatment with inverse ratio ventilation, permissive hypercapnia, prone positioning and pulmonary surfactant instillation failed to improve oxygenation. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40450\"><plain>A tracheal aspirate and subsequent bronchoscopy specimen revealed broad-based fungal elements consistent with Blastomyces dermatitidis. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40451\"><plain>No other organ involvement with <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> was evident. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40452\"><plain>Amphotericin B was added to the treatment regimen. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40453\"><plain>By day 5 of readmission (day 11 of illness), continued deterioration in respiratory function led to referral for ECMO rescue. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40454\"><plain>The patient was transported to Georgetown University Hospital without <z:efo cat=\"disease\" ids=\"EFO_0009518\">complication</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40455\"><plain>Despite multiple manipulations of mechanical ventilatory support (peak inspiratory pressure to 54 cmH2O, peak end-expiratory pressure to 18cm H2O) and a trial of inhaled nitric oxide, the OI remained &gt;40, PaCO2 &gt;99torr, PaO2 50–75torr and pulmonary artery pressure more than one-half systemic. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40456\"><plain>Hemodynamic instability developed, requiring treatment with dopamine and dobutamine infusions. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40457\"><plain>Progressive <z:efo cat=\"disease\" ids=\"EFO_1002048\">renal insufficiency</z:efo>, as demonstrated by rising creatinine and the development of <z:efo cat=\"disease\" ids=\"EFO_0009530\">anuria</z:efo>, was also present. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40458\"><plain>In an attempt to provide cardiorespiratory support while treatment for <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> was completed, and to allow time for potential lung recovery, ECMO was instituted. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40459\"><plain>Modified venoarterial ECMO was chosen because of the cardiopulmonary instability of the patient. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40460\"><plain>Cannulation was achieved using a 29 F catheter introduced into the right femoral vein and a 19F catheter placed into the right femoral artery. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40461\"><plain>The ECMO circuit has been previously described [4]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40462\"><plain>Since adequate ventilation and oxygenation can be maintained by the ECMO circuit, the peak inspiratory pressure, rate, and inspired oxygen concentration can be reduced to lessen iatrogenic lung <z:efo cat=\"disease\" ids=\"EFO_0000546\">injury</z:efo> and to provide a better environment for potential lung healing. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40463\"><plain>Initial ECMO flows of 50–60 cm3/kg per min maintained mixed venous oxygen saturations of &gt;70% and allowed a reduction in ventilator settings. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40464\"><plain>After 24 h stabilization, the patient underwent bronchoscopy, which revealed denuded airway mucosa and purulent material filling the airways. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40465\"><plain>Over the next few days, repeat bronchoscopy was performed to remove debris. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40466\"><plain>On ECMO day 3, the patient became hyperdynamic and developed <z:efo cat=\"phenotype\" ids=\"HP_0001974\">leukocytosis</z:efo> and decreasing mixed venous oxygen saturations. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40467\"><plain><z:efo cat=\"phenotype\" ids=\"HP_0100806\">Sepsis</z:efo> was presumed, although cultures remained <z:efo cat=\"disease\" ids=\"EFO_0000545\">sterile</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40468\"><plain>Membrane oxygenators on the ECMO circuit were changed without <z:efo cat=\"disease\" ids=\"EFO_0009518\">complication</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40469\"><plain>The patient continued to have episodes of decreased mixed venous oxygen saturation despite increased ECMO circuit flow. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40470\"><plain>Restriction of additional increases in circuit flow over the next few days led to the placement of a right internal jugular venous cannula and right carotid arterial cannula on ECMO day 10. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40471\"><plain>All other organ systems remained intact and all cultures (blood, bronchial washing and tissue, urine) remained <z:efo cat=\"disease\" ids=\"EFO_0000545\">sterile</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40472\"><plain>The patient was responsive to commands and questions while pharmacologic sedation was maintained. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40473\"><plain>By ECMO day 14, minimal clearing of right and left lung fields was noted with concurrent increases in exhaled tidal volumes. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40474\"><plain>Bronchoscopy was performed to remove airway debris with good results over the next few days and efforts to slowly re-expand the lungs were initiated. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40475\"><plain>On ECMO day 18, bilateral tension pneumothoraces developed and the patient suffered <z:efo cat=\"disease\" ids=\"EFO_0009492\">cardiac arrest</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40476\"><plain>Despite prompt resuscitation and continued support with ECMO, the patient deteriorated over the next 24h and ECMO support was electively withdrawn. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40477\"><plain>Native pulmonary function was insufficient and the patient expired. </plain></sent>\n",
       "</text></p><sec><title><text><sent pm=\".\" sid=\"40478\"><plain>Postmortem findings </plain></sent>\n",
       "</text></title><p><text><sent pm=\".\" sid=\"40479\"><plain>Despite systemic antifungal therapy and extracorporeal life support, postmortem examination revealed almost complete replacement of normal lung parenchyma with fungal elements and chronic inflammatory reaction. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40480\"><plain>Extensive <z:efo cat=\"phenotype\" ids=\"EFO_0009426\">necrosis</z:efo>, acute and chronic inflammatory cells, abscesses, fibrin deposition and <z:efo cat=\"phenotype\" ids=\"MP_0001914\">hemorrhage</z:efo> were seen (Fig. 2). </plain></sent>\n",
       "<sent pm=\".\" sid=\"40481\"><plain>Broad-based budding fungal organisms consistent with Blastomyces dermatitidis are seen in Figure 2 throughout the lung parenchyma. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40482\"><plain>They are not seen in association with vascular structures. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40483\"><plain>The left lung had regions of interstitial <z:efo cat=\"disease\" ids=\"EFO_0006890\">fibrosis</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40484\"><plain>The central <z:efo cat=\"disease\" ids=\"NCIT_C74532\">nervous</z:efo> system showed multiple areas of acute infarction and global <z:efo cat=\"disease\" ids=\"MONDO_0005299\">ischemic encephalopathy</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40485\"><plain>No evidence of invasion by Blastomyces dermatitidis was noted in the bloodstream or in any other organs outside of the lungs. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40486\"><plain>The final pathologic cause of death was listed as fungal <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo> with extensive destruction of lung tissue. </plain></sent>\n",
       "</text></p><fig id=\"F1\" position=\"float\"><label>Figure 1</label><caption><p><text><sent pm=\".\" sid=\"40487\"><plain>Chest radiograph on day of intubation. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40488\"><plain>Note consolidation of right hemithorax and left lower thorax. </plain></sent>\n",
       "</text></p></caption><graphic xlink:href=\"cc-3-4-091-1\"></graphic></fig><fig id=\"F2\" position=\"float\"><label>Figure 2</label><caption><p><text><sent pm=\".\" sid=\"40489\"><plain>Section of lung tissue from post-mortem examination. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40490\"><plain>H/E stain, magnification 200 ×. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40491\"><plain>Note multinucleated giant cells consistent with <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> and extensive destruction of lung. </plain></sent>\n",
       "</text></p></caption><graphic xlink:href=\"cc-3-4-091-2\"></graphic></fig></sec></sec><sec><title><text><sent pm=\".\" sid=\"40492\"><plain>Comments </plain></sent>\n",
       "</text></title><p><text><sent pm=\".\" sid=\"40493\"><plain>Unlike many fungal <z:efo cat=\"disease\" ids=\"EFO_0005741\">infections</z:efo>, <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> often occurs in immunocompetent hosts and many persons recover without antifungal treatment [5]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40494\"><plain>The typical patient with <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> is male, aged 25–50 years, and habitually exposed to the outdoor environment. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40495\"><plain>Children are rarely diagnosed with <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> (epidemics excepted). </plain></sent>\n",
       "<sent pm=\".\" sid=\"40496\"><plain>The occurrence of <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> in this patient, without living in an endemic area or with a history of prolonged outdoor activity, is extremely unusual. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40497\"><plain>The development of <z:efo cat=\"disease\" ids=\"EFO_1000637\">ARDS</z:efo> in association with pulmonary <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> is exceedingly rare. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40498\"><plain>However, when it does occur, mortality rates have been reported to be 50–80%. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40499\"><plain>It is likely that this rate of mortality is reduced with prompt recognition of the disease, aggressive supportive therapies, and administration of a total of 2 g amphotericin B [1,6]. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40500\"><plain>Our patient was referred to us as a potential candidate for ECMO. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40501\"><plain>Historical data (<z:efo cat=\"disease\" ids=\"EFO_0009686\">respiratory failure</z:efo> unresponsive to conventional therapies and an OI &gt; 40) suggested that this patient had a mortality risk of &gt;80% [4]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40502\"><plain>In our experience with pediatric <z:efo cat=\"disease\" ids=\"EFO_0009686\">respiratory failure</z:efo>, patients receiving mechanical ventilation for less than 7 days prior to ECMO initiation have a 67% survival rate. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40503\"><plain>In general, ECMO is not utilized during an active fungal <z:efo cat=\"disease\" ids=\"EFO_0005741\">infection</z:efo>, since fungal elements in the bloodstream adhere to the synthetic elements of the ECMO circuit, making it impossible to eradicate them with antifungal agents. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40504\"><plain>Pulmonary <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo>, however, is often localized to the respiratory tract, making contamination of the ECMO circuit less likely. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40505\"><plain>Since this patient was previously healthy, had single-organ disease and had failed less invasive forms of pulmonary support, ECMO was instituted to limit further iatrogenic lung <z:efo cat=\"disease\" ids=\"EFO_0000546\">injury</z:efo> from mechanical ventilation during completion of amphotericin B treatment. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40506\"><plain>Despite ECMO support, eradication of <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> and healing of diseased lung tissue did not occur and the patient did not survive. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40507\"><plain>While <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> was never isolated from the bloodstream and was only isolated from the respiratory tract prior to institution of amphotericin B therapy, it was identified throughout the lungs on postmortem examination. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40508\"><plain>In otherwise healthy individuals, it is common to see lung recovery after treatment with systemic antifungal therapy. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40509\"><plain>In fact, treatment with oral antifungal agents is sufficient for organism eradication and pulmonary recovery in most cases of <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> [7]. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40510\"><plain>There are several potential explanations for this lack of response to antifungal therapy. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40511\"><plain>One explanation is that ECMO itself impacted on the effectiveness of antifungal therapy. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40512\"><plain>Venoarterial ECMO has been shown to reduce pulmonary blood flow at high levels of bypass, and there is experimental evidence of pulmonary capillary bed <z:efo cat=\"phenotype\" ids=\"EFO_0000556\">ischemia</z:efo> with total bypass [8]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40513\"><plain>In this situation, severe lung <z:efo cat=\"disease\" ids=\"EFO_0000546\">injury</z:efo> may occur. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40514\"><plain>Another potential problem with high levels of bypass is the potential limitation of drug delivery to the infected area; inadequate pulmonary blood flow reaching the lungs will lead to impaired fungal eradication. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40515\"><plain>In this case, however, care was taken to limit the extent of cardiopulmonary bypass and to maintain pulmonary blood flow. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40516\"><plain>It is possible that local blood flow was limited by tissue and vascular <z:efo cat=\"disease\" ids=\"EFO_0000546\">injury</z:efo> from the <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo>, resulting in subtherapeutic tissue levels of amphotericin B. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40517\"><plain>This would be consistent with the pathologic findings of heavy concentrations of fungal elements in injured lung parenchyma that was not in proximity to normal vascular structures. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40518\"><plain>Another potential pitfall of ECMO is the capacity of the membrane oxygenators to remove certain medications [9]. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40519\"><plain>We monitored serum levels of amphotericin B in the patient and the ECMO circuit for standard peak and trough drug levels and when components of the circuit were changed. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40520\"><plain>All levels were within or exceeded the recommended therapeutic range and were not affected by changes of the membrane oxygenator or circuit tubing [10]. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40521\"><plain>It is also possible that this case was referred for ECMO too late in the course of disease to achieve any substantial benefit with the 'lung rest' provided by ECMO. </plain></sent>\n",
       "</text></p><p><text><sent pm=\".\" sid=\"40522\"><plain>In summary, we present the first description of the use of ECMO to support <z:efo cat=\"disease\" ids=\"EFO_0009686\">respiratory failure</z:efo> in an adolescent with <z:efo cat=\"disease\" ids=\"EFO_1000637\">ARDS</z:efo> secondary to <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> <z:efo cat=\"disease\" ids=\"EFO_0003106\">pneumonia</z:efo>. </plain></sent>\n",
       "<sent pm=\".\" sid=\"40523\"><plain>Despite the use of adequate antifungal agents, the reduction of ventilatory settings to non-toxic ranges, and adequate oxygen delivery to other organ systems, eradication of the <z:efo cat=\"disease\" ids=\"EFO_0007174\">blastomycosis</z:efo> did not occur and no lung recovery took place. </plain></sent>\n",
       "</text></p></sec><back><ref-list><ref id=\"B1\"><text><sent pm=\".\" sid=\"40524\">LockwoodWRAllisonFJrBatsonBEBuseyJFThe treatment of North American blastomycosis: ten years' experience.Am Rev Respir Dis19691003143205820331 </sent>\n",
       "</text></ref><ref id=\"B2\"><text><sent pm=\".\" sid=\"40525\">MeyerKCMcManusEJMakiDGOverwhelming pulmonary blastomycosis associated with the adult respiratory distress syndrome. N Engl J Med1993329123112368413389 </sent>\n",
       "</text></ref><ref id=\"B3\"><text><sent pm=\".\" sid=\"40526\">BeckRAndersonKDPearsonGDCriteria for extracorporeal membrane oxygenation in a population of infants with persistent pulmonary hypertension of the newborn.J Pediatr Surg 1986212973084751 </sent>\n",
       "</text></ref><ref id=\"B4\"><text><sent pm=\".\" sid=\"40527\">DaltonHJThompsonAEExtracorporeal membrane oxygenation.Pediatric Critical Care. </sent>\n",
       "<sent pm=\".\" sid=\"40528\">Edited by Fuhrman BP, Zimmerman JJ. </sent>\n",
       "<sent pm=\".\" sid=\"40529\">St Louis: Mosby1992545557 </sent>\n",
       "</text></ref><ref id=\"B5\"><text><sent pm=\".\" sid=\"40530\">BradsherRWBlastomycosis.Clin Infect Dis199214 (suppl 1)S82S901314106 </sent>\n",
       "</text></ref><ref id=\"B6\"><text><sent pm=\".\" sid=\"40531\">ParkerJDDotoILToshFEA decade of experience with blastomycosis and its treatment with amphotericin B.Am Rev Respir Dis19721058128185067267 </sent>\n",
       "</text></ref><ref id=\"B7\"><text><sent pm=\".\" sid=\"40532\">BradsherRWRiceDCAbernathyRSKetoconazole therapy for endemic blastomycosis.Ann Intern Med19851038728794062088 </sent>\n",
       "</text></ref><ref id=\"B8\"><text><sent pm=\".\" sid=\"40533\">KolobowTSpraggRGPierceJEMassive pulmonary infarction during total cardiopulmonary bypass in unaesthetized spontaneously breathing lambs.Int J Artif Organs1981476816792084 </sent>\n",
       "</text></ref><ref id=\"B9\"><text><sent pm=\".\" sid=\"40534\">ArnoldJHTruogRDOravEJScavoneJMHershensonMBTolerance and dependence in neonates sedated with fentanyl during extracorporeal membrane oxygenation.Anesthesiology199073113611402248393 </sent>\n",
       "</text></ref><ref id=\"B10\"><text><sent pm=\".\" sid=\"40535\">HertzogJHBrackettESaleMHauserGJDaltonHJAmphotericin Pharmacokinetics during ECMO.J Extracorp Technol1996289498 </sent>\n",
       "</text></ref></ref-list></back></article>\n",
       "</ebiroot>\n",
       "</p></body></html>"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "soup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [],
   "source": [
    "plain_sentences = []\n",
    "section_tags = defaultdict(set)\n",
    "gene_sentences = defaultdict(set)\n",
    "disease_setences = defaultdict(set)\n",
    "\n",
    "for each_sentence in all_sentences:\n",
    "    extracted_sentence = each_sentence.plain\n",
    "    if extracted_sentence:\n",
    "        try:\n",
    "#             section_tags[extracted_sentence.text].add(extracted_sentence.findParents('SecTag')[0]['type'])\n",
    "            section_tags[extracted_sentence.text].add(extracted_sentence.findParent('sec').title.text.strip())\n",
    "        except:\n",
    "            section_tags[extracted_sentence.text].add('')\n",
    "        plain_sentences.append(extracted_sentence.text.strip())\n",
    "        #### Extract GP sentences and their tags######\n",
    "        extracted_sentence_GP_ztags = extracted_sentence.find_all('z:uniprot')\n",
    "        if extracted_sentence_GP_ztags is not None:\n",
    "            for each_GP_z_tag in extracted_sentence_GP_ztags:\n",
    "                gene_sentences[each_sentence.plain.text].add(each_GP_z_tag.text)\n",
    "        #### Extract Disease sentences and their tags######\n",
    "        extracted_sentence_DS_ztags = extracted_sentence.find_all('z:disease')\n",
    "        if extracted_sentence_DS_ztags is not None:\n",
    "            for each_DS_z_tag in extracted_sentence_DS_ztags:\n",
    "                disease_setences[each_sentence.plain.text].add(each_DS_z_tag.text)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "from nltk.tokenize import WordPunctTokenizer\n",
    "tokenizer = WordPunctTokenizer()\n",
    "\n",
    "def get_all_tags(all_sentences):\n",
    "        ML_annotations = ml_model.post(all_sentences)\n",
    "        return ML_annotations['annotations']#list(filter(None, ML_annotations['annotations']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "MODEL_PATH = '/nfs/gns/literature/machine-learning/Santosh/Gitlab/biobertepmc/reproduce_GP_DS_OG_CD/1604049631/'\n",
    "\n",
    "# path to the file that has model parameters\n",
    "params_path = MODEL_PATH + \"params.pickle\"\n",
    "with open(params_path, 'rb') as f:\n",
    "    params = pickle.load(f)\n",
    "params['max_ner_token_len'] = -1\n",
    "params['max_bert_token_len'] = -1\n",
    "\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "ml_model = MLModel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Entity = namedtuple('Entity', ['span', 'tag', 'text', 'pre', 'post'])\n",
    "Entity_Label = namedtuple('Label', ['index', 'pos', 'tag', 'span'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = get_all_tags(plain_sentences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def roundoff(dict_y):\n",
    "    for k, v in dict_y.items():\n",
    "        v = round(v,2) \n",
    "        dict_y[k] = v \n",
    "    return dict_y\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 390/390 [00:00<00:00, 360860.04it/s]\n"
     ]
    }
   ],
   "source": [
    "ml_tagged_sentences = defaultdict(list)\n",
    "count = 0\n",
    "for each_sentence in tqdm(plain_sentences):\n",
    "#     rs = requests.post(url, data={'text': each_sentence})\n",
    "#     ml_annotations = rs.json()['annotations']\n",
    "    if annotations[count]:\n",
    "        ml_tagged_sentences[each_sentence].append(annotations[count])\n",
    "    count = count+1\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'Human Immunodeficiency Virus Type 1 DNA Sequences Genetically Damaged by Hypermutation Are Often Abundant in Patient Peripheral Blood Mononuclear Cells and May Be Generated during Near-Simultaneous Infection and Activation of CD4+ T Cells': [[[0,\n",
       "                35,\n",
       "                'OG',\n",
       "                'Human Immunodeficiency Virus Type 1'],\n",
       "               [198, 207, 'DS', 'Infection'],\n",
       "               [226, 229, 'GP', 'CD4']]],\n",
       "             'G-to-A hypermutation has been sporadically observed in human immunodeficiency virus type 1 (HIV-1) proviral sequences from patient peripheral blood mononuclear cells (PBMC) and virus cultures but has not been systematically evaluated.': [[[55,\n",
       "                90,\n",
       "                'OG',\n",
       "                'human immunodeficiency virus type 1'],\n",
       "               [92, 97, 'OG', 'HIV-1'],\n",
       "               [177, 182, 'OG', 'virus']]],\n",
       "             'PCR primers matched to normal and hypermutated sequences were used in conjunction with an agarose gel electrophoresis system incorporating an AT-binding dye to visualize, separate, clone, and sequence hypermutated and normal sequences in the 297-bp HIV-1 protease gene amplified from patient PBMC.': [[[249,\n",
       "                254,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [255, 263, 'GP', 'protease']]],\n",
       "             'Among 53 patients, including individuals infected with subtypes A through D and at different clinical stages, at least 43% of patients harbored abundant hypermutated, along with normal, protease genes.': [[[186,\n",
       "                194,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'In 70 hypermutated sequences, saturation of G residues in the GA or GG dinucleotide context ranged from 20 to 94%.': [[[71,\n",
       "                83,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'Levels of other mutants were not elevated, and G-to-A replacement was entirely restricted to GA or GG, and not GC or GT, dinucleotides.': [[[121,\n",
       "                134,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'To investigate the conditions under which hypermutation occurs in cell cultures, purified CD4+ T cells from normal donors were infected with cloned NL4-3 virus stocks at various times before and after phytohemagglutinin (PHA) activation.': [[[90,\n",
       "                93,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [148, 159, 'OG', 'NL4-3 virus'],\n",
       "               [201, 219, 'CD', 'phytohemagglutinin']]],\n",
       "             'Hypermutation was pronounced when HIV-1 infection occurred simultaneously with, or a few hours after, PHA activation, but after 12 h or more after PHA activation, most HIV-1 sequences were normal.': [[[34,\n",
       "                49,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [168, 173, 'OG', 'HIV-1']]],\n",
       "             'Near-simultaneous activation and infection of CD4+ T cells may represent a window of susceptibility where the informational content of HIV-1 sequences is lost due to hypermutation.': [[[33,\n",
       "                42,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [46, 49, 'GP', 'CD4'],\n",
       "               [135, 140, 'OG', 'HIV-1']]],\n",
       "             'There is loss of the information content of nucleotide sequences once mutation rates are increased beyond a tolerable error threshold (15).': [[[44,\n",
       "                54,\n",
       "                'CD',\n",
       "                'nucleotide']]],\n",
       "             'Some RNA viruses seem to tolerate mutation rates near this threshold, existing, not as a specific sequence, but as a quasispecies (12, 15, 26).': [[[5,\n",
       "                16,\n",
       "                'OG',\n",
       "                'RNA viruses']]],\n",
       "             'How surprising, then, that for some of the same viruses with the highest mutation rates, another mutation process has been described in which remarkable levels of one specific type of nucleotide substitution are observed.': [[[48,\n",
       "                55,\n",
       "                'OG',\n",
       "                'viruses'],\n",
       "               [184, 194, 'CD', 'nucleotide']]],\n",
       "             'A-to-G hypermutation occurs mostly in measles virus and vesicular stomatitis virus (8, 9, 49), and three cases in nonlentiviral retroviruses (avian leukosis virus and spleen necrosis virus) have been reported (16, 23, 33).': [[[38,\n",
       "                51,\n",
       "                'OG',\n",
       "                'measles virus'],\n",
       "               [56, 82, 'OG', 'vesicular stomatitis virus'],\n",
       "               [128, 140, 'OG', 'retroviruses'],\n",
       "               [142, 162, 'OG', 'avian leukosis virus'],\n",
       "               [167, 188, 'OG', 'spleen necrosis virus']]],\n",
       "             'G-to-A hypermutation is found primarily in the lentivirus family of retroviruses, along with two other examples in satellite tobacco mosaic virus (37) and hepatitis B virus (22).': [[[47,\n",
       "                57,\n",
       "                'OG',\n",
       "                'lentivirus'],\n",
       "               [68, 80, 'OG', 'retroviruses'],\n",
       "               [125, 145, 'OG', 'tobacco mosaic virus'],\n",
       "               [155, 172, 'OG', 'hepatitis B virus']]],\n",
       "             'G-to-A hypermutation in the lentivirus human immunodeficiency virus type 1 (HIV-1) is the subject of this report.': [[[28,\n",
       "                38,\n",
       "                'OG',\n",
       "                'lentivirus'],\n",
       "               [39, 74, 'OG', 'human immunodeficiency virus type 1'],\n",
       "               [76, 81, 'OG', 'HIV-1']]],\n",
       "             'G-to-A hypermutation occurs specifically within the GpA or GpG dinucleotide context (18, 60).': [[[52,\n",
       "                55,\n",
       "                'GP',\n",
       "                'GpA'],\n",
       "               [63, 75, 'CD', 'dinucleotide']]],\n",
       "             'Although G-to-A hypermutation was first described for spleen necrosis virus (SNV) by Pathak and Temin (50), G-to-A hypermutation has been mostly found among the lentiviral group of retroviruses, including HIV-1 (4, 11, 18, 21, 39, 40, 47, 60), HIV-2 (19), simian immunodeficiency virus (SIV) (30), equine infectious anemia virus (EIAV) (52), and caprine arthritis-encephalitis virus (CAEV) (64).': [[[54,\n",
       "                75,\n",
       "                'OG',\n",
       "                'spleen necrosis virus'],\n",
       "               [77, 80, 'OG', 'SNV'],\n",
       "               [181, 193, 'OG', 'retroviruses'],\n",
       "               [205, 210, 'OG', 'HIV-1'],\n",
       "               [244, 249, 'OG', 'HIV-2'],\n",
       "               [256, 285, 'OG', 'simian immunodeficiency virus'],\n",
       "               [287, 290, 'OG', 'SIV'],\n",
       "               [298, 328, 'OG', 'equine infectious anemia virus'],\n",
       "               [330, 334, 'OG', 'EIAV'],\n",
       "               [346, 382, 'OG', 'caprine arthritis-encephalitis virus'],\n",
       "               [384, 388, 'OG', 'CAEV']]],\n",
       "             'The susceptibility of lentiviruses to hypermutation is thought to be a property of their reverse transcriptase (RT), which, compared to nonlentiviral RTs, has a great capacity to elongate beyond nucleotide mismatches (51) and an increased ability to hypermutate in the presence of unbalanced nucleotide pools in vitro (43).': [[[22,\n",
       "                34,\n",
       "                'OG',\n",
       "                'lentiviruses'],\n",
       "               [89, 110, 'GP', 'reverse transcriptase'],\n",
       "               [112, 114, 'GP', 'RT'],\n",
       "               [150, 153, 'GP', 'RTs'],\n",
       "               [195, 205, 'CD', 'nucleotide'],\n",
       "               [292, 302, 'CD', 'nucleotide']]],\n",
       "             'This susceptibility is further evidenced by the elevated A content of lentivirus genomes (7).': [[[70,\n",
       "                80,\n",
       "                'OG',\n",
       "                'lentivirus']]],\n",
       "             'Hypermutation in HIV-1 was first detected during propagation of HIV-1 virus populations in vitro (11, 21, 60).': [[[17,\n",
       "                22,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [64, 75, 'OG', 'HIV-1 virus']]],\n",
       "             'Since then, several groups have recovered hypermutated HIV-1 sequences from clinical samples, confirming that they occur in vivo (4, 18, 40, 47).': [[[55,\n",
       "                60,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'With respect to genome region, hypermutation was initially found in small subregions of gag, env, nef, and U3/R elements of the long terminal repeat (LTR) (11, 21, 60), but others have found hypermutation throughout the genome (4).': [[[88,\n",
       "                91,\n",
       "                'GP',\n",
       "                'gag'],\n",
       "               [93, 96, 'GP', 'env'],\n",
       "               [98, 101, 'GP', 'nef']]],\n",
       "             'First thought to be the work of a mutant RT (18, 50), hypermutation has now been shown to be a property of wild-type RT (42) operating under suboptimal conditions.': [[[41,\n",
       "                43,\n",
       "                'GP',\n",
       "                'RT'],\n",
       "               [117, 119, 'GP', 'RT']]],\n",
       "             'G-to-A hypermutation is thought to occur when HIV minus-strand DNA synthesis takes place simultaneously with an increased intracellular concentration of dTTP relative to dCTP (60).': [[[46,\n",
       "                49,\n",
       "                'OG',\n",
       "                'HIV']]],\n",
       "             'Attempts have been made to generate hypermutants in a cell-free system in vitro with RNA, purified RT, and strongly biased deoxynucleoside triphosphate (dNTP) pools (43, 44, 59, 61).': [[[99,\n",
       "                101,\n",
       "                'GP',\n",
       "                'RT']]],\n",
       "             'While G-to-A hypermutants were produced, they lacked the marked preference for the GpA and GpG dinucleotide contexts that typify hypermutants produced in vivo or during virus cultivation in vitro.': [[[83,\n",
       "                86,\n",
       "                'GP',\n",
       "                'GpA'],\n",
       "               [95, 107, 'CD', 'dinucleotide'],\n",
       "               [169, 174, 'OG', 'virus']]],\n",
       "             'An intermediate approach where RT in virus particles was allowed to complete first-strand DNA synthesis in the presence of exogenously provided, highly biased dNTP pools also resulted in loss of the GpA and GpG dinucleotide preference (62).': [[[31,\n",
       "                33,\n",
       "                'GP',\n",
       "                'RT'],\n",
       "               [37, 42, 'OG', 'virus'],\n",
       "               [199, 202, 'GP', 'GpA'],\n",
       "               [211, 223, 'CD', 'dinucleotide']]],\n",
       "             'Hypermutants were also recovered at low frequency (0.5%) from cell cultures after the addition of deoxythymidine to supernatants, which resulted in an elevated intracellular [dTTP]/[dCTP] ratio (62), but because all possible dinucleotide contexts were not represented in the sequence that was studied, a complete analysis of context preference was not possible.': [[[98,\n",
       "                112,\n",
       "                'CD',\n",
       "                'deoxythymidine'],\n",
       "               [225, 237, 'CD', 'dinucleotide']]],\n",
       "             'More subtle perturbations in the intracellular environment may trigger dNTP pool biases sufficient to generate hypermutation while maintaining the proper dinucleotide context (67).': [[[154,\n",
       "                166,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'Indeed, a 40- fold [dTTP]/[dCTP] ratio produced hypermutants in which the GpA context preference was maintained (62), and hypermutants were recovered from 1 to 2% of unstimulated and phytohemagglutinin (PHA)-stimulated peripheral blood mononuclear cells (PBMC) without manipulation of dNTP pools (62).': [[[74,\n",
       "                77,\n",
       "                'GP',\n",
       "                'GpA'],\n",
       "               [183, 201, 'CD', 'phytohemagglutinin']]],\n",
       "             'Imbalanced and fluctuating nucleotide pools are a key element of many types of mutation, including hypermutation.': [[[27,\n",
       "                37,\n",
       "                'CD',\n",
       "                'nucleotide']]],\n",
       "             'Since the composition of dNTP pools changes as cells progress through the cell cycle (5, 20, 45), the timing of HIV-1 infection with respect to T-cell activation and entry into the cell cycle could be an important factor determining the generation of hypermutants.': [[[112,\n",
       "                127,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'In a retrovirus-based shuttle vector, G-to-A transitions predominated (31, 32) and were presumably a result of even these modest fluctuations in pools during the cell cycle.': [[[5,\n",
       "                15,\n",
       "                'OG',\n",
       "                'retrovirus']]],\n",
       "             'For example, a method based on the blue/white β-galactosidase complementation assay yielded a low recovery of hypermutants (0.5 to 2%) (62), not higher than that obtained when clones were randomly picked and sequenced.': [[[35,\n",
       "                61,\n",
       "                'GP',\n",
       "                'blue/white β-galactosidase']]],\n",
       "             'Given the lack of a specific genome location for hypermutation, the fact that hypermutants are almost always buried in a large excess of normal sequences, and the difficulty of precisely reproducing hypermutation in vitro, it is not surprising that this mutational process has been regarded as erratic, rare, and of minor importance in the HIV-1 life cycle.': [[[340,\n",
       "                345,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Here we describe the design and application of powerful new methods for systematic detection and recovery of hypermutants and their application to clinical samples and to the products of HIV-1 infection in cell culture.': [[[187,\n",
       "                202,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'The results call for a reassessment of the frequency of hypermutation in vivo, clarify the conditions that generate hypermutants in cell culture, and importantly, highlight a vulnerability of HIV-1 that could be exploited for clinical benefit.': [[[192,\n",
       "                197,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Purified PBMC from 53 HIV-1-positive individuals were the source of DNA for analysis.': [[[22,\n",
       "                27,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Twenty-six samples were from patients hospitalized in Tanzania in 1996 with symptoms compatible with AIDS (25).': [[[101,\n",
       "                105,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Fourteen others were from individuals who seroconverted to HIV-1 while in the U.S.': [[[59,\n",
       "                64,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The latter two groups were in the early, asymptomatic stage of HIV-1 infection, and the sample was typically drawn within 6 months of HIV-1 seroconversion.': [[[63,\n",
       "                78,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [134, 139, 'OG', 'HIV-1']]],\n",
       "             'DNA extracted from patient PBMC or virus cultures was the template for PCR amplification of HIV-1 protease sequences.': [[[35,\n",
       "                40,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [92, 97, 'OG', 'HIV-1'],\n",
       "               [98, 106, 'GP', 'protease']]],\n",
       "             'The primers for amplification of hypermutants contained either mixed bases (hyp primers) or G-to-A replacements (hypa primers) at some of the sites (GpA or GpG) susceptible to hypermutation.': [[[76,\n",
       "                79,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [113, 117, 'GP', 'hypa'],\n",
       "               [149, 152, 'GP', 'GpA'],\n",
       "               [156, 159, 'GP', 'GpG']]],\n",
       "             'Each PCR amplification was run in duplicate, once with normal primers and a second time with an equal mixture of hyp and hypa primers.': [[[113,\n",
       "                116,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [121, 125, 'GP', 'hypa']]],\n",
       "             'PCR products were analyzed in parallel 1% agarose gels, with or without HA yellow, a compound which consists of the DNA ligand bisbenzamide covalently linked to polyethylene glycol (PEG) (Hanse Analytik, Bremen, Germany).': [[[127,\n",
       "                139,\n",
       "                'CD',\n",
       "                'bisbenzamide']]],\n",
       "             'Bisbenzamide binds preferentially to AT-rich regions in DNA (in order from highest to lowest preference, AATT and AAAA, TAAT, ATAT, TATA, and TTAA) (1) and, when coupled to PEG, retards DNA mobility during gel electrophoresis according to AT content (41, 48, 65).': [[[0,\n",
       "                12,\n",
       "                'CD',\n",
       "                'Bisbenzamide']]],\n",
       "             'Products were visualized by poststaining with ethidium bromide.': [[[46,\n",
       "                62,\n",
       "                'CD',\n",
       "                'ethidium bromide']]],\n",
       "             'In experiments with virus cultivation in cell cultures, enrichment for hypermutated sequences by predigestion of the cellular DNA with restriction endonucleases ScrF1(CCAG∗G∗A) and AvaII (G∗G∗ACC), which preferentially cleave normal, as opposed to hypermutated, HIV-1 protease sequences, was used to recover rare hypermutated sequences from some samples.': [[[20,\n",
       "                25,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [135, 160, 'GP', 'restriction endonucleases'],\n",
       "               [161, 166, 'GP', 'ScrF1'],\n",
       "               [181, 186, 'GP', 'AvaII'],\n",
       "               [262, 267, 'OG', 'HIV-1'],\n",
       "               [268, 276, 'GP', 'protease']]],\n",
       "             'DNA was codigested with ScrF1 and AvaII at 37°C overnight prior to PCR amplification of HIV-1 protease.': [[[24,\n",
       "                29,\n",
       "                'GP',\n",
       "                'ScrF1'],\n",
       "               [34, 39, 'GP', 'AvaII'],\n",
       "               [88, 93, 'OG', 'HIV-1'],\n",
       "               [94, 102, 'GP', 'protease']]],\n",
       "             'Ligation into a plasmid vector, transformation of Escherichia coli, and plasmid purification were done using the TOPO TA cloning kit and One Shot Chemically Competent Cells from Invitrogen (San Diego, Calif.) and the Qiawell 8 ultra plasmid kit (Qiagen), respectively, as directed by the vendors.': [[[50,\n",
       "                66,\n",
       "                'OG',\n",
       "                'Escherichia coli']]],\n",
       "             'Eight hundred nanograms of purified plasmids was digested using 20 U of EcoR1 (New England Biolabs, Beverly, Mass.) for 1 h at 37°C to release the cloned protease gene, which was then rescreened on agarose gels with HA yellow.': [[[72,\n",
       "                77,\n",
       "                'GP',\n",
       "                'EcoR1'],\n",
       "               [154, 162, 'GP', 'protease']]],\n",
       "             'Both strands of the DNA were sequenced using fluorescent dye terminators present in the PRISM Ready Reaction Dyedeoxy Terminator kit with FS Taq (Applied Biosystems, Foster City, Calif.), and an Applied Biosystems Automated Sequencer as directed by the manufacturer.': [[[141,\n",
       "                144,\n",
       "                'OG',\n",
       "                'Taq']]],\n",
       "             'Determination of HIV-1 subtype.': [[[17, 22, 'OG', 'HIV-1']]],\n",
       "             'The genetic subtypes of normal protease sequences from the 53 patients were established by phylogenetic analysis (see Table 2).': [[[31,\n",
       "                39,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'Sequences were manually aligned with reference sequences of HIV-1 subtypes A through J.': [[[60,\n",
       "                65,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Among the AIDS patients more than half harbored subtype C; the remainder harbored mostly subtype A or D.': [[[10,\n",
       "                14,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Preparation and characterization of virus stocks.': [[[36,\n",
       "                41,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'Virus stocks generated from an infectious molecular clone of HIV-1 subtype B isolate NL4-3 (56, 57) were used to infect PBMC from a seronegative donor.': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'Virus'],\n",
       "               [61, 76, 'OG', 'HIV-1 subtype B'],\n",
       "               [85, 90, 'GP', 'NL4-3']]],\n",
       "             'Cells were activated with 1 μg of PHA/ml for 72 h, infected for 24 h, washed twice, and maintained in RPMI 1640 with 15% fetal calf serum (FCS), 1% PenStrep (Quality Biological, Gaithersburg, Md.), 1% l-glutamine, and 20 U of interleukin-2 (IL-2)/ml.': [[[201,\n",
       "                212,\n",
       "                'CD',\n",
       "                'l-glutamine'],\n",
       "               [226, 239, 'GP', 'interleukin-2'],\n",
       "               [241, 245, 'GP', 'IL-2']]],\n",
       "             'Culture supernatants were assayed for p24 antigen (HIV-1 p24 Antigen Assay Kit; Coulter Corporation, Miami, Fla.) at 3-day intervals.': [[[38,\n",
       "                41,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [51, 56, 'OG', 'HIV-1'],\n",
       "               [57, 60, 'GP', 'p24']]],\n",
       "             'Culture supernatants collected near the peak of p24 antigen production were treated with 50 U of DNAse I/ml (Boehringer Mannheim) for 30 min at room temperature and filtered sterilized (pore size, 0.22 μm).': [[[48,\n",
       "                51,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [97, 104, 'GP', 'DNAse I']]],\n",
       "             'To verify that the virus stocks were free of hypermutated sequences, RNA was extracted from 200 μl of virus stocks using the Nuclisens Extraction Kit (Organon Teknika), and RT-PCR was performed using the Reverse Transcription System (Promega, Madison, Wis.) with primer DP11.': [[[19,\n",
       "                24,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [102, 107, 'OG', 'virus']]],\n",
       "             'All virus stocks were free of hypermutants by these assays.': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'Fresh leucopacks were obtained from HIV-1-seronegative donors (RH Laboratories, Baltimore, Md.).': [[[36,\n",
       "                41,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'CD4+ T cells were purified by negative selection on magnetic beads with the MACS CD4+ T Cell Isolation Kit (Miltenyi Biotec, Auburn, Calif.).': [[[0,\n",
       "                3,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [81, 84, 'GP', 'CD4']]],\n",
       "             'Prior to infection, an aliquot of the purified CD4+ T cells was fixed with 2% formaldehyde and stained with fluorescein isothiocyanate (FITC)- or phycoerythrin (PE)-conjugated monoclonal antibodies specific for cell surface molecules CD3, CD4, CD8, CD14, CD20, CD56, and HLA-DR (Becton Dickinson, San Jose, Calif.).': [[[9,\n",
       "                18,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [47, 50, 'GP', 'CD4'],\n",
       "               [146, 159, 'CD', 'phycoerythrin'],\n",
       "               [234, 237, 'GP', 'CD3'],\n",
       "               [239, 242, 'GP', 'CD4'],\n",
       "               [244, 247, 'GP', 'CD8'],\n",
       "               [249, 253, 'GP', 'CD14'],\n",
       "               [255, 259, 'GP', 'CD20'],\n",
       "               [261, 265, 'GP', 'CD56'],\n",
       "               [271, 277, 'GP', 'HLA-DR']]],\n",
       "             'The majority of the cells in each experiment were CD4+ T cells (mean, 81%; range, 69 to 99%) representing most of the CD3+ T-cell subset (mean, 83%; range, 76 to 100%).': [[[50,\n",
       "                53,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [118, 121, 'GP', 'CD3']]],\n",
       "             'The other major cell type was B cells (CD20+; mean, 10.8%; range, 4.5 to 14.9%).': [[[39,\n",
       "                43,\n",
       "                'GP',\n",
       "                'CD20']]],\n",
       "             'Monocytes (CD14+), CD8+ T cells (CD8+), and NK cells (CD56+) were, on average, 2 to 4% of the population.': [[[11,\n",
       "                15,\n",
       "                'GP',\n",
       "                'CD14'],\n",
       "               [19, 22, 'GP', 'CD8'],\n",
       "               [33, 36, 'GP', 'CD8'],\n",
       "               [54, 58, 'GP', 'CD56']]],\n",
       "             'Between 14.6 and 17.7% of the cells were HLA-DR positive.': [[[41,\n",
       "                47,\n",
       "                'GP',\n",
       "                'HLA-DR']]],\n",
       "             'Cells were either left unstimulated or stimulated with 1 μg of PHA (Pharmacia)/ml for various intervals before, during, or after virus infection at a multiplicity of infection of 0.02 for 2 h at 37°C.': [[[129,\n",
       "                144,\n",
       "                'DS',\n",
       "                'virus infection'],\n",
       "               [166, 175, 'DS', 'infection']]],\n",
       "             'Cells were washed twice and maintained in RPMI 1640 containing 10% autologous plasma, and IL-2 at 20 U/ml (Amersham).': [[[90,\n",
       "                94,\n",
       "                'GP',\n",
       "                'IL-2']]],\n",
       "             'Autologous plasma, necessary to prevent activation of the unstimulated CD4+ T cells by foreign antigens, was used throughout.': [[[71,\n",
       "                74,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'Supernatants from cell cultures were evaluated for p24 antigen production using the Coulter HIV-1 p24 Antigen Assay Kit.': [[[51,\n",
       "                54,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [92, 97, 'OG', 'HIV-1'],\n",
       "               [98, 101, 'GP', 'p24']]],\n",
       "             'Nine different culture conditions were established (see Fig. 6) as follows: Culture I, virus infection without PHA stimulation; culture II, PHA and virus added simultaneously for 2 h; cultures III through VIII, PHA stimulation preceding virus infection by 3, 6, 12, 24, 48, or 72 h, respectively; culture IX, addition of virus for 2 h, washing away of virus, and PHA addition 24 h later.': [[[87,\n",
       "                102,\n",
       "                'DS',\n",
       "                'virus infection'],\n",
       "               [111, 114, 'GP', 'PHA'],\n",
       "               [148, 153, 'OG', 'virus'],\n",
       "               [211, 214, 'GP', 'PHA'],\n",
       "               [237, 252, 'DS', 'virus infection'],\n",
       "               [321, 326, 'OG', 'virus'],\n",
       "               [352, 357, 'OG', 'virus']]],\n",
       "             'Cells and culture supernatants were collected at 6, 12, 24, 48, and 72 h after infection, with the exception of culture condition IX, where collection took place at 6, 12, 24, 48, and 72 h after addition of PHA.': [[[79,\n",
       "                88,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'Parameters of hypermutation were evaluated with the Hypermut Program Package (54) (http://www.hiv.lanl.gov/HYPERMUT/hypermut.html) as implemented at the Los Alamos HIV Sequence Database, Los Alamos, N.': [[[164,\n",
       "                167,\n",
       "                'OG',\n",
       "                'HIV']]],\n",
       "             'This program identifies mutations and their dinucleotide context with respect to a reference sequence that is provided with each alignment.': [[[44,\n",
       "                56,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'For virus cultures, the sequence of NL4-3, obtained by RT-PCR and sequencing of the virus stock, was used as the reference.': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [36, 41, 'GP', 'NL4-3'],\n",
       "               [84, 89, 'OG', 'virus']]],\n",
       "             'Nucleotide sequence accession numbers.': [[[0,\n",
       "                10,\n",
       "                'CD',\n",
       "                'Nucleotide']]],\n",
       "             'All the HIV-1 sequences related to this work have been submitted to GenBank and were given accession numbers A Y036228 through A Y036577.': [[[8,\n",
       "                13,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Nested primers which amplify a 297-bp segment encoding HIV-1 protease were examined for GA and GG dinucleotides, which are susceptible to hypermutation.': [[[55,\n",
       "                60,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [61, 69, 'GP', 'protease'],\n",
       "               [95, 97, 'CD', 'GG'],\n",
       "               [98, 111, 'CD', 'dinucleotides']]],\n",
       "             'New primers were designed with either incorporated GA or CT mixtures at some of these sites (hyp primers) or replaced G with A or C with T (hypa primers).': [[[93,\n",
       "                96,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [140, 144, 'GP', 'hypa']]],\n",
       "             'The hyp and hypa primers were also shortened to bring potential mispairs closer to the 3′ end of the primer, where they could exert the maximal effect in destabilizing interaction with normal sequences.': [[[4,\n",
       "                7,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [12, 16, 'GP', 'hypa']]],\n",
       "             'The dye HA yellow, in which consists of the DNA-binding ligand bisbenzamide coupled to PEG, was incorporated into 1% agarose gels.': [[[8,\n",
       "                17,\n",
       "                'CD',\n",
       "                'HA yellow']]],\n",
       "             'A series of cloned protease sequences representing different A+T contents, amplified and sequenced previously in the pilot phase of this project, were used for evaluation of the gel system (Fig. 1).': [[[19,\n",
       "                27,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'HIV-1 proviral DNA sequences from 53 HIV-1-positive patients were investigated (Table 2).': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [37, 42, 'OG', 'HIV-1']]],\n",
       "             'The appearance of protease genes PCR amplified from patient PBMC is illustrated in Fig. 3.': [[[18,\n",
       "                26,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'While all of the protease genes were of uniform length without HA yellow, they often separated into multiple bands of different A+T content on HA yellow gels.': [[[17,\n",
       "                25,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'The sequence data established that, among the 53 patients, 23 (43%) harbored a mixture of normal and hypermutated HIV-1 sequences, while from the remaining 30 patients, only normal sequences were recovered by the techniques used (Table 2).': [[[114,\n",
       "                119,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Hypermutated sequences were found in both asymptomatic and AIDS patients, and in patients with infections with four different HIV-1 subtypes.': [[[59,\n",
       "                63,\n",
       "                'DS',\n",
       "                'AIDS'],\n",
       "               [95, 105, 'DS', 'infections'],\n",
       "               [126, 131, 'OG', 'HIV-1']]],\n",
       "             'This permitted, for the first time, a systematic evaluation of the parameters of HIV-1 hypermutation in vivo.': [[[81,\n",
       "                86,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The dinucleotide context of hypermutation was then considered (Fig. 5, bottom).': [[[4,\n",
       "                16,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'With respect to the saturation of the available GA and GG dinucleotides, at all levels of G-to-A replacement, the percentage of G replaced by A in the dinucleotide context GA was about twofold higher than the percentage of G replaced by A in the GG context.': [[[58,\n",
       "                71,\n",
       "                'CD',\n",
       "                'dinucleotides'],\n",
       "               [151, 163, 'CD', 'dinucleotide']]],\n",
       "             'Although G-to-A substitutions in the GC and GT dinucleotide contexts were previously referred as hypermutation related (60), the observed number of these replacements was limited, ranging from 0 to 3 (mean, 0.6) per hypermutated sequence.': [[[47,\n",
       "                59,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'At the highest levels of G-to-A mutation, virtually all of the GA and GG dinucleotides were mutated, while G-to-A mutations in GC or GT (other mutations; Fig. 5) were not elevated.': [[[73,\n",
       "                86,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'In summary, hypermutated HIV-1 protease sequences recovered from patient PBMC exhibited a broad range of G-to-A mutation that was entirely restricted to the GA or GG dinucleotide context, with a consistent preference for G-to-A mutation in the dinucleotide GA relative to GG at all levels of G-to-A saturation.': [[[25,\n",
       "                30,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [31, 39, 'GP', 'protease'],\n",
       "               [166, 178, 'CD', 'dinucleotide'],\n",
       "               [244, 256, 'CD', 'dinucleotide']]],\n",
       "             'In addition, there was a net accumulation of positively charged amino acids, mostly from decreased Gly, Asp, and Arg and increased Lys and Asn in hypermutants.': [[[64,\n",
       "                75,\n",
       "                'CD',\n",
       "                'amino acids']]],\n",
       "             'The proteases encoded by hypermutated sequences had, on average, a fivefold increase in net charge (Table 3).': [[[4,\n",
       "                13,\n",
       "                'GP',\n",
       "                'proteases']]],\n",
       "             'It is noteworthy that hypermutated sequences with the lowest levels of G-to-A substitutions sustained inactivating in-frame stop codons and many nonsynonymous nucleotide substitutions.': [[[159,\n",
       "                169,\n",
       "                'CD',\n",
       "                'nucleotide']]],\n",
       "             'These combined effects make it highly unlikely that functional proteases, and, consequently, viable viruses can be derived from hypermutated sequences.': [[[63,\n",
       "                72,\n",
       "                'GP',\n",
       "                'proteases'],\n",
       "               [100, 107, 'OG', 'viruses']]],\n",
       "             'The relationships between hypermutation, clinical stage, and HIV-1 subtype were investigated (Table 2).': [[[61,\n",
       "                66,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Hypermutated sequences were recovered at essentially the same rate from AIDS patients (42%) and early asymptomatic individuals (44%).': [[[72,\n",
       "                76,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Hypermutation was found in 57% of subtype A, 67% of subtype D, and 44% of subtype B infections, as well as in A+D dually infected and unclassified samples.': [[[84,\n",
       "                94,\n",
       "                'DS',\n",
       "                'infections']]],\n",
       "             'Although subtype C was the most frequently represented subtype among AIDS patients, hypermutation was detected only in 3 of 14 (21%) patients infected with this subtype.': [[[69,\n",
       "                73,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Experiments with cell-free systems have shown that unbalances in dNTP pools during reverse transcription can lead to hypermutation, but the specificity for the GA and GG dinucleotide context has been difficult to reproduce.': [[[170,\n",
       "                182,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'On the other hand, hypermutants in the correct dinucleotide context have been recovered from HIV-1 cultures in PBMC, but the exact nature of the cell population that produced them and the status of the dNTP pools in this population have not been elucidated.': [[[47,\n",
       "                59,\n",
       "                'CD',\n",
       "                'dinucleotide'],\n",
       "               [93, 98, 'OG', 'HIV-1']]],\n",
       "             'In CD4+ T cells, which are a main target of HIV 1 replication in vivo, dNTP pools are elevated when the cells are activated and in cycle (20) but are unbalanced and fluctuating in resting T cells and during the transition from a resting to an activated state (5, 10, 20, 58).': [[[3,\n",
       "                6,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [44, 49, 'OG', 'HIV 1']]],\n",
       "             'We designed an experiment in which the timing of HIV-1 infection was varied with respect to activation of resting CD4+ T cells by PHA; the outcome measure was recovery of hypermutated sequences whose characteristics corresponded exactly to those found in vivo.': [[[49,\n",
       "                64,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [114, 117, 'GP', 'CD4']]],\n",
       "             'Proviral DNA was established in all culture conditions, as evidenced by its recovery by DNA PCR; a mock-infected culture was HIV-1 DNA PCR negative (data not shown).': [[[125,\n",
       "                130,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'p24 antigen in the culture supernatant, indicative of virus production, was observed only when cultures were stimulated with PHA either before, or simultaneously with, virus infection.': [[[0,\n",
       "                3,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [54, 59, 'OG', 'virus'],\n",
       "               [168, 183, 'DS', 'virus infection']]],\n",
       "             'The HIV-1 protease gene was amplified from all time points and culture conditions (Materials and Methods) and evaluated for the presence of normal and hypermutated sequences using the HA yellow gel system (Fig. 6).': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [10, 18, 'GP', 'protease']]],\n",
       "             'The richest source of hypermutants was culture II, in which infection and activation occurred simultaneously (Fig. 6A, lanes b through e).': [[[60,\n",
       "                69,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'Hypermutated sequences could also be visualized in conditions III and IV (lanes g and h), which represented PHA activation for less than 8 h before infection.': [[[108,\n",
       "                111,\n",
       "                'GP',\n",
       "                'PHA'],\n",
       "               [148, 157, 'DS', 'infection']]],\n",
       "             'However, by pre-enrichment for hypermutants using ScrF1 and AvaII digestion (see Materials and Methods), rare hypermutants could be observed in late time points in most cultures (Fig. 6, lanes f∗, i∗, j∗, k∗, and l∗).': [[[50,\n",
       "                55,\n",
       "                'GP',\n",
       "                'ScrF1'],\n",
       "               [60, 65, 'GP', 'AvaII']]],\n",
       "             'Finally, hypermutated sequences were found in condition IX, where virus infection preceded PHA activation by 24 h.': [[[66,\n",
       "                81,\n",
       "                'DS',\n",
       "                'virus infection'],\n",
       "               [91, 94, 'GP', 'PHA']]],\n",
       "             'Overall, these analyses establish that HIV-1 infection of resting CD4+ T cells either shortly before or shortly after activation generated abundant hypermutated sequences with high levels of G-to-A mutation, while infection after longer periods of activation yielded principally normal sequences.': [[[39,\n",
       "                54,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [66, 69, 'GP', 'CD4'],\n",
       "               [214, 223, 'DS', 'infection']]],\n",
       "             'These combined approaches provided 131 different HIV-1 protease sequences from virus cultures for analysis.': [[[49,\n",
       "                54,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [55, 63, 'GP', 'protease'],\n",
       "               [79, 84, 'OG', 'virus']]],\n",
       "             'The normal sequences were narrowly distributed around 62% A+T, that of the NL4-3 virus innoculum.': [[[75,\n",
       "                86,\n",
       "                'OG',\n",
       "                'NL4-3 virus']]],\n",
       "             'However, when sequences from each culture condition were analyzed separately, an interesting pattern of dinucleotide context utilization was noted (Fig. 7B, lower graph).': [[[104,\n",
       "                116,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'In cultures IX, I, and II, where PHA either was not added or was added simultaneously with or after infection with the virus, the preference for GA over GG was not strong.': [[[100,\n",
       "                109,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [119, 124, 'OG', 'virus']]],\n",
       "             'With prior PHA stimulation for 3 h or more (cultures III to VIII), the saturation of available GA dinucleotides occurred at 4 to 6 times that of available GG dinucleotides.': [[[95,\n",
       "                111,\n",
       "                'CD',\n",
       "                'GA dinucleotides'],\n",
       "               [158, 171, 'CD', 'dinucleotides']]],\n",
       "             'Finally, hypermutated and normal protease gene sequences recovered from cell culture were examined for the loss of informational content (Table 3).': [[[33,\n",
       "                41,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'This study represents the first systematic evaluation of patient PBMC and virus cultures with a consistently applied methodology optimized for detection of hypermutated HIV-1 sequences.': [[[74,\n",
       "                79,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [169, 174, 'OG', 'HIV-1']]],\n",
       "             'Hypermutated sequences could be detected in 43% of patients and in virus cultures when T-cell stimulation was initiated at the time of, or shortly before, virus infection.': [[[67,\n",
       "                72,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [155, 170, 'DS', 'virus infection']]],\n",
       "             'The nearly identical quantitative parameters of the hypermutants recovered from patients and from virus cultures suggest that T cells in the process of becoming activated are also the main source of hypermutation in vivo.': [[[98,\n",
       "                103,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'The excessive accumulation of nonsynonymous substitutions and stop codon formation observed in hypermutants reinforces the perception that they can no longer generate viable viruses.': [[[174,\n",
       "                181,\n",
       "                'OG',\n",
       "                'viruses']]],\n",
       "             'We examined the question of whether hypermutation affects GC and GT dinucleotides, although at lower levels, in addition to the preferred GA and GG contexts.': [[[68,\n",
       "                81,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'However, our observations are limited to the HIV-1 protease gene, and examination of full HIV-1 genomes for the context of hypermutation would be required to fortify this conclusion.': [[[45,\n",
       "                50,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [51, 59, 'GP', 'protease'],\n",
       "               [90, 95, 'OG', 'HIV-1']]],\n",
       "             'A consistent preference for saturation of available GA over GG dinucleotides within hypermutated sequences was also noted.': [[[63,\n",
       "                76,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'Indeed, the DNA-binding preferences of bisbenzamide for runs of AT, particularly AAAA and AATT, would suggest that GA to AA would be more efficiently detected than GG to AG, but the fact that the vast majority of hypermutants we recovered used both contexts suggests that this was not a major obstacle to recovery.': [[[39,\n",
       "                51,\n",
       "                'CD',\n",
       "                'bisbenzamide']]],\n",
       "             'Many hypermutated proviruses may have been missed by our procedure.': [[[18,\n",
       "                28,\n",
       "                'OG',\n",
       "                'proviruses']]],\n",
       "             'Here we show that the process is completely independent of the normal accumulation of mutations in HIV-1 and covers a much broader dynamic range.': [[[99,\n",
       "                104,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The preferential utilization of available GA over GG dinucleotides over the whole dynamic range, and the fact that the vast majority of hypermutants utilized both GA and GG contexts, suggests that both contexts are recognized during a single reverse transcription event.': [[[50,\n",
       "                66,\n",
       "                'CD',\n",
       "                'GG dinucleotides']]],\n",
       "             'The high error rate of HIV-1 RT is apparently necessary to permit both “normal” mutation and hypermutation, but hypermutation should not be thought of as an elevated error rate for RT; it is clearly an independent process with separate biochemical requirements which must vary over a wide range during HIV-1 replication in vivo.': [[[23,\n",
       "                28,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [29, 31, 'GP', 'RT'],\n",
       "               [302, 307, 'OG', 'HIV-1']]],\n",
       "             'Could hypermutation be influenced by viral variation, or is it essentially a passive function of the condition of the cell population in which HIV-1 replicates?': [[[143,\n",
       "                148,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Stronger evidence would be needed, including larger numbers of patients with a broad range of clinical stages, to establish a definitive relationship between hypermutation and HIV-1 subtype.': [[[176,\n",
       "                181,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The experiments with virus cultures are illuminating with respect to the conditions under which hypermutation occurs.': [[[21,\n",
       "                26,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'First, the evidence that true hypermutants were generated in CD4+ T cells in culture should be reiterated.': [[[61,\n",
       "                64,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'The target cell populations were free of HIV-1 because DNA extracted from the cells of mock-infected cultures carried in parallel were HIV-1 PCR negative.': [[[41,\n",
       "                46,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [135, 140, 'OG', 'HIV-1']]],\n",
       "             'The virus stock was molecularly cloned and verified to be free of hypermutation by sequencing of the viral RNA by RT-PCR and by DNase treatment to remove any contaminating proviral DNA remaining from virus propagation.': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [128, 133, 'GP', 'DNase'],\n",
       "               [200, 205, 'OG', 'virus']]],\n",
       "             'DNA PCR of the culture supernatants after infection was negative.': [[[42,\n",
       "                51,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'The identity of the target cell population was carefully established; the culture contained principally, if not exclusively, CD4+ T cells.': [[[125,\n",
       "                128,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'The major contaminant was B cells, which do not support HIV-1 replication.': [[[56,\n",
       "                61,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Most, if not all, CD4+ T cells were quiescent.': [[[18,\n",
       "                21,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'It has been previously documented that infection of quiescent cells results in the absence of virus progeny (68) and that viral DNA can be detected after infection of unstimulated cells although it is not associated with virus production (69, 70).': [[[39,\n",
       "                48,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [94, 99, 'OG', 'virus'],\n",
       "               [154, 163, 'DS', 'infection'],\n",
       "               [221, 226, 'OG', 'virus']]],\n",
       "             'In our experiments, proviral DNA was established but there was no detectable virus production in the absence of PHA stimulation, which is consistent with the infection of quiescent cells.': [[[77,\n",
       "                82,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [158, 167, 'DS', 'infection']]],\n",
       "             'Most of the cells were HLA-DR negative.': [[[23,\n",
       "                29,\n",
       "                'GP',\n",
       "                'HLA-DR']]],\n",
       "             'It is highly probable that those cells that were HLA-DR+, ranging from 15 to 18% in different experiments, represented recently, but not currently, activated cells.': [[[49,\n",
       "                55,\n",
       "                'GP',\n",
       "                'HLA-DR']]],\n",
       "             'Hypermutants were recovered principally from virus cultures in which HIV-1 infection was simultaneous with, or briefly preceded by, PHA stimulation (Fig. 6).': [[[45,\n",
       "                50,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [69, 84, 'DS', 'HIV-1 infection']]],\n",
       "             'Even the dinucleotide context preference is partially parsed out by these experiments; we note that in culture conditions XI, I, and II, where PHA either was not added or was added along with or after the virus, utilization of available GA and GG was almost equivalent.': [[[9,\n",
       "                21,\n",
       "                'CD',\n",
       "                'dinucleotide'],\n",
       "               [205, 210, 'OG', 'virus']]],\n",
       "             'The strong bias for GA appeared once PHA preceded HIV-1 infection by only 3 h (Figure 7B).': [[[50,\n",
       "                65,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'A previous study demonstrated that hypermutants were recovered, although at low frequencies, after infection of unstimulated (2%) and stimulated (0.9%) bulk PBMC (62).': [[[99,\n",
       "                108,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'Such recovery of hypermutants was ascribed to virus infection of a small proportion of PBMC with distorted dNTP pools, reinforcing the concept of an occasional and sporadic event.': [[[46,\n",
       "                61,\n",
       "                'DS',\n",
       "                'virus infection']]],\n",
       "             'Because the composition of dNTP pools fluctuates as cells progress through the cell cycle (5, 20, 45), we investigated if the timing of HIV-1 infection with respect to T-cell stimulation could influence the generation of hypermutation.': [[[136,\n",
       "                151,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'The target cell population for hypermutation must be continually renewed in vivo, commensurate with the proportion of resting T cells that become newly activated during active virus replication.': [[[176,\n",
       "                181,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'Such sequences should be included in conceptualizations of the latent reservoir of HIV-1 in patients.': [[[83,\n",
       "                88,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Is hypermutation related to the observation that the genomes of HIV-1 and other lentiviruses are AT rich?': [[[64,\n",
       "                69,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [80, 92, 'OG', 'lentiviruses']]],\n",
       "             'The total and unrecoverable loss of coding potential of hypermutated HIV sequences (Table 3) would suggest that they are incapable of generating progeny virions and cannot contribute to the HIV-1 gene pool.': [[[69,\n",
       "                72,\n",
       "                'OG',\n",
       "                'HIV'],\n",
       "               [190, 195, 'OG', 'HIV-1']]],\n",
       "             'It will be important to determine what fraction of the genome in hypermutants remains unaffected, as well as the potential for hypermutated proviruses to generate viable viruses through recombination.': [[[140,\n",
       "                150,\n",
       "                'OG',\n",
       "                'proviruses'],\n",
       "               [170, 177, 'OG', 'viruses']]],\n",
       "             'Hypermutated proviruses could represent a difficult-to-eradicate remnant of HIV-1 during highly active antiretroviral therapy.': [[[13,\n",
       "                23,\n",
       "                'OG',\n",
       "                'proviruses'],\n",
       "               [76, 81, 'OG', 'HIV-1']]],\n",
       "             'Another aspect is the fact that the vast majority of hypermutated protease sequences present in this study were associated with a disrupted coding potential.': [[[66,\n",
       "                74,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'Since protease has a crucial role during the HIV-1 replication cycle, the loss of its coding potential integrity should lead to abortive infections.': [[[6,\n",
       "                14,\n",
       "                'GP',\n",
       "                'protease'],\n",
       "               [45, 50, 'OG', 'HIV-1'],\n",
       "               [137, 147, 'DS', 'infections']]],\n",
       "             'Our data demonstrated that hypermutation may reduce the pool of viable replicating genomes in an infecting HIV-1 population.': [[[107,\n",
       "                112,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Detection of HIV-1 hypermutated sequences in at least 43% of patients studied suggests that hypermutation may happen in a systematic way in HIV-1-infected individuals.': [[[13,\n",
       "                18,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [140, 145, 'OG', 'HIV-1']]],\n",
       "             'Based on these two observations one can speculate that hypermutation may be seen as a result of a host-associated mechanism which may decrease virus replication.': [[[143,\n",
       "                148,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'A better understanding of the metabolic profile of early proliferating T cells in relation to their expanding dNTP pools would lead to a better interpretation of the correlation between temporal aspects of cell cycle metabolism and HIV-1 hypermutation.': [[[232,\n",
       "                237,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'This interpretation could develop into new strategies of promoting a hypermutation-inducible state in HIV-1 target cells, leading to irreversible mutagenesis of viral genomes.': [[[102,\n",
       "                107,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'A better comprehension of the cellular mechanisms involved in HIV-1 hypermutation could open new avenues for virus control.': [[[62,\n",
       "                67,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [109, 114, 'OG', 'virus']]],\n",
       "             'They amplify a 297-bp segment of the HIV-1 genome encoding the protease gene, as indicated by their location in the reference sequence HXB2 (GenBank accession number AF033819).': [[[37,\n",
       "                42,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [63, 71, 'GP', 'protease']]],\n",
       "             'HIV-1 protease sequences amplified from patient PBMC and representing a range of G-to-A hypermutations were used to explore the performance of 1% agarose gels containing HA yellow, a dye that preferentially binds to AT-rich regions in DNA.': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [6, 14, 'GP', 'protease'],\n",
       "               [170, 172, 'GP', 'HA']]],\n",
       "             'The GA and GG dinucleotides that are susceptible to hypermutation are shaded.': [[[14,\n",
       "                27,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'The gel on the left shows the migration of these PCR products as a single band at 297 bp without HA yellow.': [[[97,\n",
       "                99,\n",
       "                'GP',\n",
       "                'HA']]],\n",
       "             'Primers were designed to incorporate G-to-A substitutions in GA or GG dinucleotides (Materials and Methods) and compared to primers matched to normal sequences for their ability to amplify hypermutated sequences from the PBMC of 10 patients.': [[[67,\n",
       "                69,\n",
       "                'CD',\n",
       "                'GG'],\n",
       "               [70, 83, 'CD', 'dinucleotides']]],\n",
       "             'TABLE 2Evaluation of patient PBMC for the presence of hypermutated HIV-1 sequences CountryStudyDateClinical StageSubtypeaNo. of patientsNo.': [[[67,\n",
       "                72,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Military Seroconverters1997–1998AsymptomaticB147\\u2009(50) TanzaniaSentinel Surveillance1996AIDSA74\\u2009(57) C143\\u2009(21) D32\\u2009(67) Other22\\u2009(100) \\u2003Total5323\\u2009(43)aThe genetic subtype of the normal HIV-1 protease sequence was determined by phylogenetic analysis (data not shown).': [[[183,\n",
       "                188,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [189, 197, 'GP', 'protease']]],\n",
       "             'Other, mixed infection or unclassified.\\xa0bHypermutated HIV-1 sequences could be recovered from primary PBMC as described in Materials and Methods.': [[[7,\n",
       "                22,\n",
       "                'DS',\n",
       "                'mixed infection'],\n",
       "               [54, 59, 'OG', 'HIV-1']]],\n",
       "             'Products were visualized by staining with ethidium bromide.': [[[42,\n",
       "                58,\n",
       "                'CD',\n",
       "                'ethidium bromide']]],\n",
       "             'HIV-1 protease genes were amplified using DNA extracted from the PBMC of 53 patients ranging from the early, asymptomatic stage of HIV-1 infection to late-stage AIDS, and including infections with HIV-1 subtypes A through D (Table 2).': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [6, 14, 'GP', 'protease'],\n",
       "               [131, 146, 'DS', 'HIV-1 infection'],\n",
       "               [161, 165, 'DS', 'AIDS'],\n",
       "               [181, 191, 'DS', 'infections'],\n",
       "               [197, 202, 'OG', 'HIV-1']]],\n",
       "             'TABLE 3Genetic damage in hypermutated sequences SourceSequence typenNo. of in-frame stop codonsMean (range) net charge of protein Patient PBMCNormal14932.5\\u2009(0.0–9.0) Hypermutated706912.6\\u2009(3.0–23.0) Virus culturesNormal7331.9\\u2009(0.0–4.0) Hypermutated584810.9\\u2009(2.0–23.0)': [[[198,\n",
       "                203,\n",
       "                'OG',\n",
       "                'Virus']]],\n",
       "             'Relationship of hypermutation to T-cell activation in virus cultures.': [[[54,\n",
       "                59,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'CD4+ T cells isolated from the PBMC of normal donors were infected with an NL4-3 virus stock for 2 h at a multiplicity of infection of 0.02 (stippled gray bar) under varying conditions of PHA activation (gray bars).': [[[0,\n",
       "                3,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [75, 80, 'OG', 'NL4-3'],\n",
       "               [81, 86, 'OG', 'virus'],\n",
       "               [122, 131, 'DS', 'infection']]],\n",
       "             'Condition I, infection without addition of PHA; condition II, simultaneous infection and PHA addition; conditions III to VIII, infection with prior PHA activation for 3 to 72 h respectively; condition IX, infection and addition of PHA 24 h later.': [[[13,\n",
       "                22,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [75, 84, 'DS', 'infection'],\n",
       "               [127, 136, 'DS', 'infection'],\n",
       "               [148, 151, 'GP', 'PHA'],\n",
       "               [205, 214, 'DS', 'infection']]],\n",
       "             'Cultures were sampled at intervals and assayed for the presence of p24 antigen in the culture supernatant (solid lines) and for the presence of hypermutated HIV-1 DNA sequences.': [[[67,\n",
       "                70,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [157, 162, 'OG', 'HIV-1']]],\n",
       "             'Samples from time points at which hypermutation was rare (f, i, j, k, and l) are shown both as bulk PCR products (B) and after pre-enrichment by ScrF1 and AvaII digestion (A).': [[[145,\n",
       "                150,\n",
       "                'GP',\n",
       "                'ScrF1'],\n",
       "               [155, 160, 'GP', 'AvaII']]],\n",
       "             'Parameters of hypermutation in virus cultures.': [[[31,\n",
       "                36,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'A total of 227 HIV-1 protease sequences, 131 of which were unique, were obtained from virus cultures.': [[[15,\n",
       "                20,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [21, 29, 'GP', 'protease'],\n",
       "               [86, 91, 'OG', 'virus']]],\n",
       "             '(C) Base composition of normal and hypermutated sequences from virus cultures.': [[[63,\n",
       "                68,\n",
       "                'OG',\n",
       "                'virus']]]})"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ml_tagged_sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ML not recognising covid terms in some cases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_list = ['covid-19', 'covid19', 'sarscov2', 'sars-cov-2', '2019-ncov']\n",
    "def compare_ml_annotations_with_dictionary_tagged(ml_tags_, z_tags_):\n",
    "    agreed_z_tags = set()\n",
    "#     print(z_tags_, ml_tags_)\n",
    "    for each_z_tag in z_tags_:\n",
    "        for each_ml_annotation in ml_tags_:\n",
    "            if each_z_tag.lower() in missing_list:\n",
    "                agreed_z_tags.add(each_z_tag)\n",
    "            else:\n",
    "                score = fuzz.partial_ratio(each_ml_annotation, each_z_tag) #token_set_ratio\n",
    "                if score == 100:\n",
    "                    agreed_z_tags.add(each_z_tag)\n",
    "    return list(z_tags_- agreed_z_tags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import itertools\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ml_tagged_sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_new_missing_tags(each_sentence):\n",
    "    new_entities = []\n",
    "    for missing_string in missing_list:\n",
    "        for i in re.finditer(missing_string, each_sentence.lower()):\n",
    "            indexlocation= i.span()\n",
    "    #         print(indexlocation)\n",
    "            startindex= i.start()\n",
    "            endindex= i.end()\n",
    "            entity = each_sentence[indexlocation[0]:indexlocation[1]]\n",
    "            new_entities.append([startindex,endindex, 'DS', entity])\n",
    "    return new_entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gp, ds, cd, and og sets\n",
    "gp_set = set()\n",
    "ds_set = set()\n",
    "cd_set = set()\n",
    "og_set = set()\n",
    "\n",
    "final_sentences = defaultdict(list)\n",
    "\n",
    "for sentence, ml_tags in ml_tagged_sentences.items():\n",
    "    all_tags =[]\n",
    "    new_missing_stuff = get_new_missing_tags(sentence)\n",
    "    if ml_tags[0]:\n",
    "        all_tags = new_missing_stuff+ml_tags[0]\n",
    "    else:\n",
    "        all_tags = new_missing_stuff\n",
    "    if all_tags: # only if the tags are present, so discard the sentence without any tags\n",
    "        final_sentences[sentence] = all_tags\n",
    "        for each_ml_tag in all_tags:\n",
    "            if each_ml_tag[2] =='GP':\n",
    "                gp_set.add(each_ml_tag[3])\n",
    "            elif each_ml_tag[2] =='DS':\n",
    "                ds_set.add(each_ml_tag[3])\n",
    "            if each_ml_tag[2] =='CD':\n",
    "                cd_set.add(each_ml_tag[3])\n",
    "            if each_ml_tag[2] =='OG':\n",
    "                og_set.add(each_ml_tag[3])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sentences_matches_tags(sentences_tags):\n",
    "    matches = defaultdict(list)\n",
    "    for each_sentence, ml_tags in sentences_tags.items():\n",
    "        for each_ml_tag in ml_tags:\n",
    "            if each_ml_tag[2]!= 'OG':\n",
    "                mini_dict = {}\n",
    "                mini_dict['label'] = each_ml_tag[3]\n",
    "                mini_dict['type'] = each_ml_tag[2]\n",
    "                mini_dict['startInSentence'] = each_ml_tag[0]\n",
    "                mini_dict['endInSentence'] = each_ml_tag[1]\n",
    "                matches[each_sentence].append(mini_dict)\n",
    "    return matches\n",
    "\n",
    "\n",
    "def get_mapped_list_from_annotations(annotation_list):\n",
    "    mapped_list = list(itertools.combinations(annotation_list, 2))\n",
    "\n",
    "    unique_maplist = []\n",
    "    for each_list in mapped_list:\n",
    "        if each_list[0][2] !=each_list[1][2] and each_list[1][2]!='OG' and each_list[0][2]!='OG':\n",
    "            unique_maplist.append((each_list[0], each_list[1]))\n",
    "\n",
    "    return unique_maplist  \n",
    "\n",
    "def get_sentences_offset_per_cooccurance(sentences_tags):\n",
    "    \n",
    "    dict_gp_ds = defaultdict(list)\n",
    "    dict_gp_cd = defaultdict(list)\n",
    "    dict_ds_cd = defaultdict(list)\n",
    "\n",
    "    for sentence, tags in sentences_tags.items():\n",
    "        if len(tags)>1: # only if more than 1 tag is available\n",
    "            check_tags =np.array(tags)\n",
    "            if 'GP' in  check_tags and 'DS' in check_tags:\n",
    "                dict_gp_ds[sentence] = get_mapped_list_from_annotations(tags)\n",
    "            if 'GP' in  check_tags and 'CD' in check_tags:\n",
    "                dict_gp_cd[sentence] = get_mapped_list_from_annotations(tags)\n",
    "            if 'DS' in  check_tags and 'CD' in check_tags:\n",
    "                dict_ds_cd[sentence]= get_mapped_list_from_annotations(tags)         \n",
    "                \n",
    "    return dict_gp_ds, dict_gp_cd, dict_ds_cd\n",
    "\n",
    "\n",
    "\n",
    "def get_article_offset_per_tag(soup, tagset):\n",
    "    dicttag = defaultdict(list)\n",
    "    \n",
    "    \n",
    "    for each_tag in tagset:\n",
    "        for i in re.finditer(each_tag, each_sentence.lower()):\n",
    "            indexlocation= i.span()\n",
    "    #         print(indexlocation)\n",
    "            startindex= i.start()\n",
    "            endindex= i.end()\n",
    "            entity = each_sentence[indexlocation[0]:indexlocation[1]]\n",
    "        dicttag[each_tag].append([each_sentence, each_ml_tag[0], each_ml_tag[1]])\n",
    "    return dicttag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "# GP\": [\n",
    "# {\n",
    "#   label: NR2B,\n",
    "#   abstract_matches: [{start: XXXX, end: XXXX}, {start: XXXX, end: XXXX}],\n",
    "#   full_text_matches: [{start: XXXX, end: XXXX}]\n",
    "# },\n",
    "# {\n",
    "#  label: \"NR2B\",\n",
    "#  abstract_matches:[{start:XXX, end:XXXXX}],\n",
    "# },\n",
    "\n",
    "## hard to get above notations are we process at sentences level, so the offset changes per sentence.\n",
    "\n",
    "dict_gp = get_sentences_offset_per_tag(gp_set, final_sentences, tag_type = 'GP')\n",
    "dict_ds = get_sentences_offset_per_tag(ds_set, final_sentences, tag_type = 'DS')\n",
    "dict_cd = get_sentences_offset_per_tag(cd_set, final_sentences, tag_type = 'CD')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xx_ = ' '.join(plain_sentences[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# {\n",
    "#   \"lit_id\": \"http://europepmc.org/abstract/MED/11250803\",\n",
    "#   \"mined_sentences\": [\n",
    "# \t{\n",
    "#   \t\"section\": \"abstract\",\n",
    "#   \t\"t_start\": 222,\n",
    "#   \t\"t_end\": 242,\n",
    "#   \t\"d_start\": 23,\n",
    "#   \t\"d_end\": 41,\n",
    "#   \t\"text\": \"Both further stimulate cardiac hypertrophy and, importantly, activate counterregulatory mechanisms including overexpression of atrial natriuretic peptide and b-type natriuretic peptide, and production of cytokines such as tumor necrosis factor-alpha.\"\n",
    "# \t}\n",
    "#   ]\n",
    "\n",
    "# }\n",
    "\n",
    "# old style\n",
    "\n",
    "\n",
    "dict_gp_ds, dict_gp_cd, dict_ds_cd = get_sentences_offset_per_cooccurance(final_sentences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "def swap_positions(cooccurance_list, pos1, pos2): \n",
    "    cooccurance_list[pos1], cooccurance_list[pos2] = cooccurance_list[pos2], cooccurance_list[pos1] \n",
    "    return cooccurance_list\n",
    "\n",
    "def get_relations(gp_ds_text_sentence):\n",
    "    docs = relation_model2(gp_ds_text_sentence)\n",
    "    rel_list =[]\n",
    "    for ent in docs.ents:\n",
    "        if ent.label_!='GP' and ent.label_!='DS':\n",
    "            rel_dict = {}\n",
    "            rel_dict['startr'] = ent.start_char\n",
    "            rel_dict['endr'] = ent.end_char\n",
    "            rel_dict['labelr'] = ent.text\n",
    "            rel_dict['typer'] = ent.label_\n",
    "            rel_list.append(rel_dict)\n",
    "    return rel_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cooccurance_evidence(dict_tags, tag_type_1, tag_type_2):\n",
    "    co_occurance_sentences = defaultdict(list)\n",
    "#     mined_sentences = []\n",
    "    for each_sent_map, mappedtags in dict_tags.items():\n",
    "        # always see that GP-DS, GP-CD and CD-DS is followed\n",
    "        \n",
    "        if tag_type_1 not in mappedtags[0][0][2]:\n",
    "            mappedtags[0] = swap_positions(list(mappedtags[0]),0,1)\n",
    "        else:\n",
    "            mappedtags[0] = list(mappedtags[0])\n",
    "\n",
    "        for eachtag in mappedtags:\n",
    "            if tag_type_1==eachtag[0][2] and tag_type_2==eachtag[1][2]:\n",
    "                mini_dict = {}\n",
    "                mini_dict['start1'] = eachtag[0][0]\n",
    "                mini_dict['end1']= eachtag[0][1]\n",
    "                mini_dict['label1']= eachtag[0][3]\n",
    "                mini_dict['start2'] = eachtag[1][0]\n",
    "                mini_dict['end2']= eachtag[1][1]\n",
    "                mini_dict['label2']= eachtag[1][3]\n",
    "                mini_dict['type'] = tag_type_1+'-'+tag_type_2\n",
    "                \n",
    "                if tag_type_1=='GP' and tag_type_2=='DS':\n",
    "                    # get associations scores\n",
    "                    mini_dict['association'] = roundoff(relation_model1(each_sent_map).cats)\n",
    "                    # get relations\n",
    "                    rels = get_relations(each_sent_map)\n",
    "                    if rels:\n",
    "                        mini_dict['relation'] = rels\n",
    "                \n",
    "                co_occurance_sentences[each_sent_map].append(mini_dict)\n",
    "    return co_occurance_sentences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "co_occurance_gp_ds = get_cooccurance_evidence(dict_gp_ds, tag_type_1='GP', tag_type_2='DS')\n",
    "co_occurance_gp_cd = get_cooccurance_evidence(dict_gp_cd, tag_type_1='GP', tag_type_2='CD')\n",
    "co_occurance_ds_cd = get_cooccurance_evidence(dict_ds_cd, tag_type_1='DS', tag_type_2='CD')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'Human Immunodeficiency Virus Type 1 DNA Sequences Genetically Damaged by Hypermutation Are Often Abundant in Patient Peripheral Blood Mononuclear Cells and May Be Generated during Near-Simultaneous Infection and Activation of CD4+ T Cells': [{'start1': 226,\n",
       "               'end1': 229,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 198,\n",
       "               'end2': 207,\n",
       "               'label2': 'Infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 1.0,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.59,\n",
       "                'Negative': 0.0,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.15,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.0},\n",
       "               'relation': [{'startr': 212,\n",
       "                 'endr': 222,\n",
       "                 'labelr': 'Activation',\n",
       "                 'typer': 'POSITIVE_REGULATION'}]}],\n",
       "             'Near-simultaneous activation and infection of CD4+ T cells may represent a window of susceptibility where the informational content of HIV-1 sequences is lost due to hypermutation.': [{'start1': 46,\n",
       "               'end1': 49,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 33,\n",
       "               'end2': 42,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.98,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.85,\n",
       "                'Negative': 0.02,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.07,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.06,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Prior to infection, an aliquot of the purified CD4+ T cells was fixed with 2% formaldehyde and stained with fluorescein isothiocyanate (FITC)- or phycoerythrin (PE)-conjugated monoclonal antibodies specific for cell surface molecules CD3, CD4, CD8, CD14, CD20, CD56, and HLA-DR (Becton Dickinson, San Jose, Calif.).': [{'start1': 47,\n",
       "               'end1': 50,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 9,\n",
       "               'end2': 18,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 1.0,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.86,\n",
       "                'Negative': 0.01,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.13,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Nine different culture conditions were established (see Fig. 6) as follows: Culture I, virus infection without PHA stimulation; culture II, PHA and virus added simultaneously for 2 h; cultures III through VIII, PHA stimulation preceding virus infection by 3, 6, 12, 24, 48, or 72 h, respectively; culture IX, addition of virus for 2 h, washing away of virus, and PHA addition 24 h later.': [{'start1': 111,\n",
       "               'end1': 114,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 87,\n",
       "               'end2': 102,\n",
       "               'label2': 'virus infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.01,\n",
       "                'Positive': 0.21,\n",
       "                'Negative': 0.02,\n",
       "                'Neutral': 0.02,\n",
       "                'Altered Expression': 0.1,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.0}},\n",
       "              {'start1': 111,\n",
       "               'end1': 114,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 237,\n",
       "               'end2': 252,\n",
       "               'label2': 'virus infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.01,\n",
       "                'Positive': 0.21,\n",
       "                'Negative': 0.02,\n",
       "                'Neutral': 0.02,\n",
       "                'Altered Expression': 0.1,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.0}},\n",
       "              {'start1': 211,\n",
       "               'end1': 214,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 237,\n",
       "               'end2': 252,\n",
       "               'label2': 'virus infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.01,\n",
       "                'Positive': 0.21,\n",
       "                'Negative': 0.02,\n",
       "                'Neutral': 0.02,\n",
       "                'Altered Expression': 0.1,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'We designed an experiment in which the timing of HIV-1 infection was varied with respect to activation of resting CD4+ T cells by PHA; the outcome measure was recovery of hypermutated sequences whose characteristics corresponded exactly to those found in vivo.': [{'start1': 114,\n",
       "               'end1': 117,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 49,\n",
       "               'end2': 64,\n",
       "               'label2': 'HIV-1 infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.21,\n",
       "                'Negative': 0.01,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.09,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.02,\n",
       "                'Regulatory modification': 0.0},\n",
       "               'relation': [{'startr': 92,\n",
       "                 'endr': 102,\n",
       "                 'labelr': 'activation',\n",
       "                 'typer': 'POSITIVE_REGULATION'}]}],\n",
       "             'p24 antigen in the culture supernatant, indicative of virus production, was observed only when cultures were stimulated with PHA either before, or simultaneously with, virus infection.': [{'start1': 0,\n",
       "               'end1': 3,\n",
       "               'label1': 'p24',\n",
       "               'start2': 168,\n",
       "               'end2': 183,\n",
       "               'label2': 'virus infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.98,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.03,\n",
       "                'Negative': 0.02,\n",
       "                'Neutral': 0.04,\n",
       "                'Altered Expression': 0.01,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.05,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Hypermutated sequences could also be visualized in conditions III and IV (lanes g and h), which represented PHA activation for less than 8 h before infection.': [{'start1': 108,\n",
       "               'end1': 111,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 148,\n",
       "               'end2': 157,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.98,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.01,\n",
       "                'Negative': 0.01,\n",
       "                'Neutral': 0.56,\n",
       "                'Altered Expression': 0.06,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.01,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Finally, hypermutated sequences were found in condition IX, where virus infection preceded PHA activation by 24 h.': [{'start1': 91,\n",
       "               'end1': 94,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 66,\n",
       "               'end2': 81,\n",
       "               'label2': 'virus infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 1.0,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.0,\n",
       "                'Negative': 0.0,\n",
       "                'Neutral': 0.87,\n",
       "                'Altered Expression': 0.01,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Overall, these analyses establish that HIV-1 infection of resting CD4+ T cells either shortly before or shortly after activation generated abundant hypermutated sequences with high levels of G-to-A mutation, while infection after longer periods of activation yielded principally normal sequences.': [{'start1': 66,\n",
       "               'end1': 69,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 39,\n",
       "               'end2': 54,\n",
       "               'label2': 'HIV-1 infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.55,\n",
       "                'Negative': 0.03,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.02,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.13,\n",
       "                'Regulatory modification': 0.0}},\n",
       "              {'start1': 66,\n",
       "               'end1': 69,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 214,\n",
       "               'end2': 223,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.55,\n",
       "                'Negative': 0.03,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.02,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.13,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Since protease has a crucial role during the HIV-1 replication cycle, the loss of its coding potential integrity should lead to abortive infections.': [{'start1': 6,\n",
       "               'end1': 14,\n",
       "               'label1': 'protease',\n",
       "               'start2': 137,\n",
       "               'end2': 147,\n",
       "               'label2': 'infections',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.02,\n",
       "                'Negative': 0.01,\n",
       "                'Neutral': 0.01,\n",
       "                'Altered Expression': 0.0,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.02,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'HIV-1 protease genes were amplified using DNA extracted from the PBMC of 53 patients ranging from the early, asymptomatic stage of HIV-1 infection to late-stage AIDS, and including infections with HIV-1 subtypes A through D (Table 2).': [{'start1': 6,\n",
       "               'end1': 14,\n",
       "               'label1': 'protease',\n",
       "               'start2': 131,\n",
       "               'end2': 146,\n",
       "               'label2': 'HIV-1 infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.96,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.0,\n",
       "                'Negative': 0.04,\n",
       "                'Neutral': 0.93,\n",
       "                'Altered Expression': 0.0,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.04,\n",
       "                'Regulatory modification': 0.0}},\n",
       "              {'start1': 6,\n",
       "               'end1': 14,\n",
       "               'label1': 'protease',\n",
       "               'start2': 161,\n",
       "               'end2': 165,\n",
       "               'label2': 'AIDS',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.96,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.0,\n",
       "                'Negative': 0.04,\n",
       "                'Neutral': 0.93,\n",
       "                'Altered Expression': 0.0,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.04,\n",
       "                'Regulatory modification': 0.0}},\n",
       "              {'start1': 6,\n",
       "               'end1': 14,\n",
       "               'label1': 'protease',\n",
       "               'start2': 181,\n",
       "               'end2': 191,\n",
       "               'label2': 'infections',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.96,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.0,\n",
       "                'Negative': 0.04,\n",
       "                'Neutral': 0.93,\n",
       "                'Altered Expression': 0.0,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.04,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'CD4+ T cells isolated from the PBMC of normal donors were infected with an NL4-3 virus stock for 2 h at a multiplicity of infection of 0.02 (stippled gray bar) under varying conditions of PHA activation (gray bars).': [{'start1': 0,\n",
       "               'end1': 3,\n",
       "               'label1': 'CD4',\n",
       "               'start2': 122,\n",
       "               'end2': 131,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.84,\n",
       "                'Negative': 0.0,\n",
       "                'Neutral': 0.0,\n",
       "                'Altered Expression': 0.47,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.01,\n",
       "                'Regulatory modification': 0.0}}],\n",
       "             'Condition I, infection without addition of PHA; condition II, simultaneous infection and PHA addition; conditions III to VIII, infection with prior PHA activation for 3 to 72 h respectively; condition IX, infection and addition of PHA 24 h later.': [{'start1': 148,\n",
       "               'end1': 151,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 13,\n",
       "               'end2': 22,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.09,\n",
       "                'Negative': 0.01,\n",
       "                'Neutral': 0.64,\n",
       "                'Altered Expression': 0.08,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.01},\n",
       "               'relation': [{'startr': 152,\n",
       "                 'endr': 162,\n",
       "                 'labelr': 'activation',\n",
       "                 'typer': 'POSITIVE_REGULATION'}]},\n",
       "              {'start1': 148,\n",
       "               'end1': 151,\n",
       "               'label1': 'PHA',\n",
       "               'start2': 205,\n",
       "               'end2': 214,\n",
       "               'label2': 'infection',\n",
       "               'type': 'GP-DS',\n",
       "               'association': {'Yes': 0.99,\n",
       "                'No': 0.0,\n",
       "                'Positive': 0.09,\n",
       "                'Negative': 0.01,\n",
       "                'Neutral': 0.64,\n",
       "                'Altered Expression': 0.08,\n",
       "                'Genetic Variation': 0.0,\n",
       "                'Any': 0.0,\n",
       "                'Regulatory modification': 0.01},\n",
       "               'relation': [{'startr': 152,\n",
       "                 'endr': 162,\n",
       "                 'labelr': 'activation',\n",
       "                 'typer': 'POSITIVE_REGULATION'}]}]})"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "co_occurance_gp_ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_defaultdicts(dict1,dict2):\n",
    "    for key, value in dict2.items():\n",
    "        for subkey, subvalue in value.items():\n",
    "            dict1[key][subkey] = subvalue\n",
    "    return dict1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "mactch_gp_ds_cd = get_sentences_matches_tags(final_sentences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'Human Immunodeficiency Virus Type 1 DNA Sequences Genetically Damaged by Hypermutation Are Often Abundant in Patient Peripheral Blood Mononuclear Cells and May Be Generated during Near-Simultaneous Infection and Activation of CD4+ T Cells': [{'label': 'Infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 198,\n",
       "               'endInSentence': 207},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 226,\n",
       "               'endInSentence': 229}],\n",
       "             'PCR primers matched to normal and hypermutated sequences were used in conjunction with an agarose gel electrophoresis system incorporating an AT-binding dye to visualize, separate, clone, and sequence hypermutated and normal sequences in the 297-bp HIV-1 protease gene amplified from patient PBMC.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 255,\n",
       "               'endInSentence': 263}],\n",
       "             'Among 53 patients, including individuals infected with subtypes A through D and at different clinical stages, at least 43% of patients harbored abundant hypermutated, along with normal, protease genes.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 186,\n",
       "               'endInSentence': 194}],\n",
       "             'In 70 hypermutated sequences, saturation of G residues in the GA or GG dinucleotide context ranged from 20 to 94%.': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 71,\n",
       "               'endInSentence': 83}],\n",
       "             'Levels of other mutants were not elevated, and G-to-A replacement was entirely restricted to GA or GG, and not GC or GT, dinucleotides.': [{'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 121,\n",
       "               'endInSentence': 134}],\n",
       "             'To investigate the conditions under which hypermutation occurs in cell cultures, purified CD4+ T cells from normal donors were infected with cloned NL4-3 virus stocks at various times before and after phytohemagglutinin (PHA) activation.': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 90,\n",
       "               'endInSentence': 93},\n",
       "              {'label': 'phytohemagglutinin',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 201,\n",
       "               'endInSentence': 219}],\n",
       "             'Hypermutation was pronounced when HIV-1 infection occurred simultaneously with, or a few hours after, PHA activation, but after 12 h or more after PHA activation, most HIV-1 sequences were normal.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 34,\n",
       "               'endInSentence': 49}],\n",
       "             'Near-simultaneous activation and infection of CD4+ T cells may represent a window of susceptibility where the informational content of HIV-1 sequences is lost due to hypermutation.': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 33,\n",
       "               'endInSentence': 42},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 46,\n",
       "               'endInSentence': 49}],\n",
       "             'There is loss of the information content of nucleotide sequences once mutation rates are increased beyond a tolerable error threshold (15).': [{'label': 'nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 44,\n",
       "               'endInSentence': 54}],\n",
       "             'How surprising, then, that for some of the same viruses with the highest mutation rates, another mutation process has been described in which remarkable levels of one specific type of nucleotide substitution are observed.': [{'label': 'nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 184,\n",
       "               'endInSentence': 194}],\n",
       "             'G-to-A hypermutation occurs specifically within the GpA or GpG dinucleotide context (18, 60).': [{'label': 'GpA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 52,\n",
       "               'endInSentence': 55},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 63,\n",
       "               'endInSentence': 75}],\n",
       "             'The susceptibility of lentiviruses to hypermutation is thought to be a property of their reverse transcriptase (RT), which, compared to nonlentiviral RTs, has a great capacity to elongate beyond nucleotide mismatches (51) and an increased ability to hypermutate in the presence of unbalanced nucleotide pools in vitro (43).': [{'label': 'reverse transcriptase',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 89,\n",
       "               'endInSentence': 110},\n",
       "              {'label': 'RT',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 112,\n",
       "               'endInSentence': 114},\n",
       "              {'label': 'RTs',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 150,\n",
       "               'endInSentence': 153},\n",
       "              {'label': 'nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 195,\n",
       "               'endInSentence': 205},\n",
       "              {'label': 'nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 292,\n",
       "               'endInSentence': 302}],\n",
       "             'With respect to genome region, hypermutation was initially found in small subregions of gag, env, nef, and U3/R elements of the long terminal repeat (LTR) (11, 21, 60), but others have found hypermutation throughout the genome (4).': [{'label': 'gag',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 88,\n",
       "               'endInSentence': 91},\n",
       "              {'label': 'env',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 93,\n",
       "               'endInSentence': 96},\n",
       "              {'label': 'nef',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 98,\n",
       "               'endInSentence': 101}],\n",
       "             'First thought to be the work of a mutant RT (18, 50), hypermutation has now been shown to be a property of wild-type RT (42) operating under suboptimal conditions.': [{'label': 'RT',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 41,\n",
       "               'endInSentence': 43},\n",
       "              {'label': 'RT',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 117,\n",
       "               'endInSentence': 119}],\n",
       "             'Attempts have been made to generate hypermutants in a cell-free system in vitro with RNA, purified RT, and strongly biased deoxynucleoside triphosphate (dNTP) pools (43, 44, 59, 61).': [{'label': 'RT',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 99,\n",
       "               'endInSentence': 101}],\n",
       "             'While G-to-A hypermutants were produced, they lacked the marked preference for the GpA and GpG dinucleotide contexts that typify hypermutants produced in vivo or during virus cultivation in vitro.': [{'label': 'GpA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 83,\n",
       "               'endInSentence': 86},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 95,\n",
       "               'endInSentence': 107}],\n",
       "             'An intermediate approach where RT in virus particles was allowed to complete first-strand DNA synthesis in the presence of exogenously provided, highly biased dNTP pools also resulted in loss of the GpA and GpG dinucleotide preference (62).': [{'label': 'RT',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 31,\n",
       "               'endInSentence': 33},\n",
       "              {'label': 'GpA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 199,\n",
       "               'endInSentence': 202},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 211,\n",
       "               'endInSentence': 223}],\n",
       "             'Hypermutants were also recovered at low frequency (0.5%) from cell cultures after the addition of deoxythymidine to supernatants, which resulted in an elevated intracellular [dTTP]/[dCTP] ratio (62), but because all possible dinucleotide contexts were not represented in the sequence that was studied, a complete analysis of context preference was not possible.': [{'label': 'deoxythymidine',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 98,\n",
       "               'endInSentence': 112},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 225,\n",
       "               'endInSentence': 237}],\n",
       "             'More subtle perturbations in the intracellular environment may trigger dNTP pool biases sufficient to generate hypermutation while maintaining the proper dinucleotide context (67).': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 154,\n",
       "               'endInSentence': 166}],\n",
       "             'Indeed, a 40- fold [dTTP]/[dCTP] ratio produced hypermutants in which the GpA context preference was maintained (62), and hypermutants were recovered from 1 to 2% of unstimulated and phytohemagglutinin (PHA)-stimulated peripheral blood mononuclear cells (PBMC) without manipulation of dNTP pools (62).': [{'label': 'GpA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 74,\n",
       "               'endInSentence': 77},\n",
       "              {'label': 'phytohemagglutinin',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 183,\n",
       "               'endInSentence': 201}],\n",
       "             'Imbalanced and fluctuating nucleotide pools are a key element of many types of mutation, including hypermutation.': [{'label': 'nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 27,\n",
       "               'endInSentence': 37}],\n",
       "             'Since the composition of dNTP pools changes as cells progress through the cell cycle (5, 20, 45), the timing of HIV-1 infection with respect to T-cell activation and entry into the cell cycle could be an important factor determining the generation of hypermutants.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 112,\n",
       "               'endInSentence': 127}],\n",
       "             'For example, a method based on the blue/white β-galactosidase complementation assay yielded a low recovery of hypermutants (0.5 to 2%) (62), not higher than that obtained when clones were randomly picked and sequenced.': [{'label': 'blue/white β-galactosidase',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 35,\n",
       "               'endInSentence': 61}],\n",
       "             'Here we describe the design and application of powerful new methods for systematic detection and recovery of hypermutants and their application to clinical samples and to the products of HIV-1 infection in cell culture.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 187,\n",
       "               'endInSentence': 202}],\n",
       "             'Twenty-six samples were from patients hospitalized in Tanzania in 1996 with symptoms compatible with AIDS (25).': [{'label': 'AIDS',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 101,\n",
       "               'endInSentence': 105}],\n",
       "             'The latter two groups were in the early, asymptomatic stage of HIV-1 infection, and the sample was typically drawn within 6 months of HIV-1 seroconversion.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 63,\n",
       "               'endInSentence': 78}],\n",
       "             'DNA extracted from patient PBMC or virus cultures was the template for PCR amplification of HIV-1 protease sequences.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 98,\n",
       "               'endInSentence': 106}],\n",
       "             'The primers for amplification of hypermutants contained either mixed bases (hyp primers) or G-to-A replacements (hypa primers) at some of the sites (GpA or GpG) susceptible to hypermutation.': [{'label': 'hyp',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 76,\n",
       "               'endInSentence': 79},\n",
       "              {'label': 'hypa',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 113,\n",
       "               'endInSentence': 117},\n",
       "              {'label': 'GpA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 149,\n",
       "               'endInSentence': 152},\n",
       "              {'label': 'GpG',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 156,\n",
       "               'endInSentence': 159}],\n",
       "             'Each PCR amplification was run in duplicate, once with normal primers and a second time with an equal mixture of hyp and hypa primers.': [{'label': 'hyp',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 113,\n",
       "               'endInSentence': 116},\n",
       "              {'label': 'hypa',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 121,\n",
       "               'endInSentence': 125}],\n",
       "             'PCR products were analyzed in parallel 1% agarose gels, with or without HA yellow, a compound which consists of the DNA ligand bisbenzamide covalently linked to polyethylene glycol (PEG) (Hanse Analytik, Bremen, Germany).': [{'label': 'bisbenzamide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 127,\n",
       "               'endInSentence': 139}],\n",
       "             'Bisbenzamide binds preferentially to AT-rich regions in DNA (in order from highest to lowest preference, AATT and AAAA, TAAT, ATAT, TATA, and TTAA) (1) and, when coupled to PEG, retards DNA mobility during gel electrophoresis according to AT content (41, 48, 65).': [{'label': 'Bisbenzamide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 0,\n",
       "               'endInSentence': 12}],\n",
       "             'Products were visualized by poststaining with ethidium bromide.': [{'label': 'ethidium bromide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 46,\n",
       "               'endInSentence': 62}],\n",
       "             'In experiments with virus cultivation in cell cultures, enrichment for hypermutated sequences by predigestion of the cellular DNA with restriction endonucleases ScrF1(CCAG∗G∗A) and AvaII (G∗G∗ACC), which preferentially cleave normal, as opposed to hypermutated, HIV-1 protease sequences, was used to recover rare hypermutated sequences from some samples.': [{'label': 'restriction endonucleases',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 135,\n",
       "               'endInSentence': 160},\n",
       "              {'label': 'ScrF1',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 161,\n",
       "               'endInSentence': 166},\n",
       "              {'label': 'AvaII',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 181,\n",
       "               'endInSentence': 186},\n",
       "              {'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 268,\n",
       "               'endInSentence': 276}],\n",
       "             'DNA was codigested with ScrF1 and AvaII at 37°C overnight prior to PCR amplification of HIV-1 protease.': [{'label': 'ScrF1',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 24,\n",
       "               'endInSentence': 29},\n",
       "              {'label': 'AvaII',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 34,\n",
       "               'endInSentence': 39},\n",
       "              {'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 94,\n",
       "               'endInSentence': 102}],\n",
       "             'Eight hundred nanograms of purified plasmids was digested using 20 U of EcoR1 (New England Biolabs, Beverly, Mass.) for 1 h at 37°C to release the cloned protease gene, which was then rescreened on agarose gels with HA yellow.': [{'label': 'EcoR1',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 72,\n",
       "               'endInSentence': 77},\n",
       "              {'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 154,\n",
       "               'endInSentence': 162}],\n",
       "             'The genetic subtypes of normal protease sequences from the 53 patients were established by phylogenetic analysis (see Table 2).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 31,\n",
       "               'endInSentence': 39}],\n",
       "             'Among the AIDS patients more than half harbored subtype C; the remainder harbored mostly subtype A or D.': [{'label': 'AIDS',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 10,\n",
       "               'endInSentence': 14}],\n",
       "             'Virus stocks generated from an infectious molecular clone of HIV-1 subtype B isolate NL4-3 (56, 57) were used to infect PBMC from a seronegative donor.': [{'label': 'NL4-3',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 85,\n",
       "               'endInSentence': 90}],\n",
       "             'Cells were activated with 1 μg of PHA/ml for 72 h, infected for 24 h, washed twice, and maintained in RPMI 1640 with 15% fetal calf serum (FCS), 1% PenStrep (Quality Biological, Gaithersburg, Md.), 1% l-glutamine, and 20 U of interleukin-2 (IL-2)/ml.': [{'label': 'l-glutamine',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 201,\n",
       "               'endInSentence': 212},\n",
       "              {'label': 'interleukin-2',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 226,\n",
       "               'endInSentence': 239},\n",
       "              {'label': 'IL-2',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 241,\n",
       "               'endInSentence': 245}],\n",
       "             'Culture supernatants were assayed for p24 antigen (HIV-1 p24 Antigen Assay Kit; Coulter Corporation, Miami, Fla.) at 3-day intervals.': [{'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 38,\n",
       "               'endInSentence': 41},\n",
       "              {'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 57,\n",
       "               'endInSentence': 60}],\n",
       "             'Culture supernatants collected near the peak of p24 antigen production were treated with 50 U of DNAse I/ml (Boehringer Mannheim) for 30 min at room temperature and filtered sterilized (pore size, 0.22 μm).': [{'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 48,\n",
       "               'endInSentence': 51},\n",
       "              {'label': 'DNAse I',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 97,\n",
       "               'endInSentence': 104}],\n",
       "             'CD4+ T cells were purified by negative selection on magnetic beads with the MACS CD4+ T Cell Isolation Kit (Miltenyi Biotec, Auburn, Calif.).': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 0,\n",
       "               'endInSentence': 3},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 81,\n",
       "               'endInSentence': 84}],\n",
       "             'Prior to infection, an aliquot of the purified CD4+ T cells was fixed with 2% formaldehyde and stained with fluorescein isothiocyanate (FITC)- or phycoerythrin (PE)-conjugated monoclonal antibodies specific for cell surface molecules CD3, CD4, CD8, CD14, CD20, CD56, and HLA-DR (Becton Dickinson, San Jose, Calif.).': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 9,\n",
       "               'endInSentence': 18},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 47,\n",
       "               'endInSentence': 50},\n",
       "              {'label': 'phycoerythrin',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 146,\n",
       "               'endInSentence': 159},\n",
       "              {'label': 'CD3',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 234,\n",
       "               'endInSentence': 237},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 239,\n",
       "               'endInSentence': 242},\n",
       "              {'label': 'CD8',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 244,\n",
       "               'endInSentence': 247},\n",
       "              {'label': 'CD14',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 249,\n",
       "               'endInSentence': 253},\n",
       "              {'label': 'CD20',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 255,\n",
       "               'endInSentence': 259},\n",
       "              {'label': 'CD56',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 261,\n",
       "               'endInSentence': 265},\n",
       "              {'label': 'HLA-DR',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 271,\n",
       "               'endInSentence': 277}],\n",
       "             'The majority of the cells in each experiment were CD4+ T cells (mean, 81%; range, 69 to 99%) representing most of the CD3+ T-cell subset (mean, 83%; range, 76 to 100%).': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 50,\n",
       "               'endInSentence': 53},\n",
       "              {'label': 'CD3',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 118,\n",
       "               'endInSentence': 121}],\n",
       "             'The other major cell type was B cells (CD20+; mean, 10.8%; range, 4.5 to 14.9%).': [{'label': 'CD20',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 39,\n",
       "               'endInSentence': 43}],\n",
       "             'Monocytes (CD14+), CD8+ T cells (CD8+), and NK cells (CD56+) were, on average, 2 to 4% of the population.': [{'label': 'CD14',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 11,\n",
       "               'endInSentence': 15},\n",
       "              {'label': 'CD8',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 19,\n",
       "               'endInSentence': 22},\n",
       "              {'label': 'CD8',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 33,\n",
       "               'endInSentence': 36},\n",
       "              {'label': 'CD56',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 54,\n",
       "               'endInSentence': 58}],\n",
       "             'Between 14.6 and 17.7% of the cells were HLA-DR positive.': [{'label': 'HLA-DR',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 41,\n",
       "               'endInSentence': 47}],\n",
       "             'Cells were either left unstimulated or stimulated with 1 μg of PHA (Pharmacia)/ml for various intervals before, during, or after virus infection at a multiplicity of infection of 0.02 for 2 h at 37°C.': [{'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 129,\n",
       "               'endInSentence': 144},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 166,\n",
       "               'endInSentence': 175}],\n",
       "             'Cells were washed twice and maintained in RPMI 1640 containing 10% autologous plasma, and IL-2 at 20 U/ml (Amersham).': [{'label': 'IL-2',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 90,\n",
       "               'endInSentence': 94}],\n",
       "             'Autologous plasma, necessary to prevent activation of the unstimulated CD4+ T cells by foreign antigens, was used throughout.': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 71,\n",
       "               'endInSentence': 74}],\n",
       "             'Supernatants from cell cultures were evaluated for p24 antigen production using the Coulter HIV-1 p24 Antigen Assay Kit.': [{'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 51,\n",
       "               'endInSentence': 54},\n",
       "              {'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 98,\n",
       "               'endInSentence': 101}],\n",
       "             'Nine different culture conditions were established (see Fig. 6) as follows: Culture I, virus infection without PHA stimulation; culture II, PHA and virus added simultaneously for 2 h; cultures III through VIII, PHA stimulation preceding virus infection by 3, 6, 12, 24, 48, or 72 h, respectively; culture IX, addition of virus for 2 h, washing away of virus, and PHA addition 24 h later.': [{'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 87,\n",
       "               'endInSentence': 102},\n",
       "              {'label': 'PHA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 111,\n",
       "               'endInSentence': 114},\n",
       "              {'label': 'PHA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 211,\n",
       "               'endInSentence': 214},\n",
       "              {'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 237,\n",
       "               'endInSentence': 252}],\n",
       "             'Cells and culture supernatants were collected at 6, 12, 24, 48, and 72 h after infection, with the exception of culture condition IX, where collection took place at 6, 12, 24, 48, and 72 h after addition of PHA.': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 79,\n",
       "               'endInSentence': 88}],\n",
       "             'This program identifies mutations and their dinucleotide context with respect to a reference sequence that is provided with each alignment.': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 44,\n",
       "               'endInSentence': 56}],\n",
       "             'For virus cultures, the sequence of NL4-3, obtained by RT-PCR and sequencing of the virus stock, was used as the reference.': [{'label': 'NL4-3',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 36,\n",
       "               'endInSentence': 41}],\n",
       "             'Nucleotide sequence accession numbers.': [{'label': 'Nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 0,\n",
       "               'endInSentence': 10}],\n",
       "             'Nested primers which amplify a 297-bp segment encoding HIV-1 protease were examined for GA and GG dinucleotides, which are susceptible to hypermutation.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 61,\n",
       "               'endInSentence': 69},\n",
       "              {'label': 'GG',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 95,\n",
       "               'endInSentence': 97},\n",
       "              {'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 98,\n",
       "               'endInSentence': 111}],\n",
       "             'New primers were designed with either incorporated GA or CT mixtures at some of these sites (hyp primers) or replaced G with A or C with T (hypa primers).': [{'label': 'hyp',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 93,\n",
       "               'endInSentence': 96},\n",
       "              {'label': 'hypa',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 140,\n",
       "               'endInSentence': 144}],\n",
       "             'The hyp and hypa primers were also shortened to bring potential mispairs closer to the 3′ end of the primer, where they could exert the maximal effect in destabilizing interaction with normal sequences.': [{'label': 'hyp',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 4,\n",
       "               'endInSentence': 7},\n",
       "              {'label': 'hypa',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 12,\n",
       "               'endInSentence': 16}],\n",
       "             'The dye HA yellow, in which consists of the DNA-binding ligand bisbenzamide coupled to PEG, was incorporated into 1% agarose gels.': [{'label': 'HA yellow',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 8,\n",
       "               'endInSentence': 17}],\n",
       "             'A series of cloned protease sequences representing different A+T contents, amplified and sequenced previously in the pilot phase of this project, were used for evaluation of the gel system (Fig. 1).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 19,\n",
       "               'endInSentence': 27}],\n",
       "             'The appearance of protease genes PCR amplified from patient PBMC is illustrated in Fig. 3.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 18,\n",
       "               'endInSentence': 26}],\n",
       "             'While all of the protease genes were of uniform length without HA yellow, they often separated into multiple bands of different A+T content on HA yellow gels.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 17,\n",
       "               'endInSentence': 25}],\n",
       "             'Hypermutated sequences were found in both asymptomatic and AIDS patients, and in patients with infections with four different HIV-1 subtypes.': [{'label': 'AIDS',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 59,\n",
       "               'endInSentence': 63},\n",
       "              {'label': 'infections',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 95,\n",
       "               'endInSentence': 105}],\n",
       "             'The dinucleotide context of hypermutation was then considered (Fig. 5, bottom).': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 4,\n",
       "               'endInSentence': 16}],\n",
       "             'With respect to the saturation of the available GA and GG dinucleotides, at all levels of G-to-A replacement, the percentage of G replaced by A in the dinucleotide context GA was about twofold higher than the percentage of G replaced by A in the GG context.': [{'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 58,\n",
       "               'endInSentence': 71},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 151,\n",
       "               'endInSentence': 163}],\n",
       "             'Although G-to-A substitutions in the GC and GT dinucleotide contexts were previously referred as hypermutation related (60), the observed number of these replacements was limited, ranging from 0 to 3 (mean, 0.6) per hypermutated sequence.': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 47,\n",
       "               'endInSentence': 59}],\n",
       "             'At the highest levels of G-to-A mutation, virtually all of the GA and GG dinucleotides were mutated, while G-to-A mutations in GC or GT (other mutations; Fig. 5) were not elevated.': [{'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 73,\n",
       "               'endInSentence': 86}],\n",
       "             'In summary, hypermutated HIV-1 protease sequences recovered from patient PBMC exhibited a broad range of G-to-A mutation that was entirely restricted to the GA or GG dinucleotide context, with a consistent preference for G-to-A mutation in the dinucleotide GA relative to GG at all levels of G-to-A saturation.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 31,\n",
       "               'endInSentence': 39},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 166,\n",
       "               'endInSentence': 178},\n",
       "              {'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 244,\n",
       "               'endInSentence': 256}],\n",
       "             'In addition, there was a net accumulation of positively charged amino acids, mostly from decreased Gly, Asp, and Arg and increased Lys and Asn in hypermutants.': [{'label': 'amino acids',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 64,\n",
       "               'endInSentence': 75}],\n",
       "             'The proteases encoded by hypermutated sequences had, on average, a fivefold increase in net charge (Table 3).': [{'label': 'proteases',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 4,\n",
       "               'endInSentence': 13}],\n",
       "             'It is noteworthy that hypermutated sequences with the lowest levels of G-to-A substitutions sustained inactivating in-frame stop codons and many nonsynonymous nucleotide substitutions.': [{'label': 'nucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 159,\n",
       "               'endInSentence': 169}],\n",
       "             'These combined effects make it highly unlikely that functional proteases, and, consequently, viable viruses can be derived from hypermutated sequences.': [{'label': 'proteases',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 63,\n",
       "               'endInSentence': 72}],\n",
       "             'Hypermutated sequences were recovered at essentially the same rate from AIDS patients (42%) and early asymptomatic individuals (44%).': [{'label': 'AIDS',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 72,\n",
       "               'endInSentence': 76}],\n",
       "             'Hypermutation was found in 57% of subtype A, 67% of subtype D, and 44% of subtype B infections, as well as in A+D dually infected and unclassified samples.': [{'label': 'infections',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 84,\n",
       "               'endInSentence': 94}],\n",
       "             'Although subtype C was the most frequently represented subtype among AIDS patients, hypermutation was detected only in 3 of 14 (21%) patients infected with this subtype.': [{'label': 'AIDS',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 69,\n",
       "               'endInSentence': 73}],\n",
       "             'Experiments with cell-free systems have shown that unbalances in dNTP pools during reverse transcription can lead to hypermutation, but the specificity for the GA and GG dinucleotide context has been difficult to reproduce.': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 170,\n",
       "               'endInSentence': 182}],\n",
       "             'On the other hand, hypermutants in the correct dinucleotide context have been recovered from HIV-1 cultures in PBMC, but the exact nature of the cell population that produced them and the status of the dNTP pools in this population have not been elucidated.': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 47,\n",
       "               'endInSentence': 59}],\n",
       "             'In CD4+ T cells, which are a main target of HIV 1 replication in vivo, dNTP pools are elevated when the cells are activated and in cycle (20) but are unbalanced and fluctuating in resting T cells and during the transition from a resting to an activated state (5, 10, 20, 58).': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 3,\n",
       "               'endInSentence': 6}],\n",
       "             'We designed an experiment in which the timing of HIV-1 infection was varied with respect to activation of resting CD4+ T cells by PHA; the outcome measure was recovery of hypermutated sequences whose characteristics corresponded exactly to those found in vivo.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 49,\n",
       "               'endInSentence': 64},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 114,\n",
       "               'endInSentence': 117}],\n",
       "             'p24 antigen in the culture supernatant, indicative of virus production, was observed only when cultures were stimulated with PHA either before, or simultaneously with, virus infection.': [{'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 0,\n",
       "               'endInSentence': 3},\n",
       "              {'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 168,\n",
       "               'endInSentence': 183}],\n",
       "             'The HIV-1 protease gene was amplified from all time points and culture conditions (Materials and Methods) and evaluated for the presence of normal and hypermutated sequences using the HA yellow gel system (Fig. 6).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 10,\n",
       "               'endInSentence': 18}],\n",
       "             'The richest source of hypermutants was culture II, in which infection and activation occurred simultaneously (Fig. 6A, lanes b through e).': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 60,\n",
       "               'endInSentence': 69}],\n",
       "             'Hypermutated sequences could also be visualized in conditions III and IV (lanes g and h), which represented PHA activation for less than 8 h before infection.': [{'label': 'PHA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 108,\n",
       "               'endInSentence': 111},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 148,\n",
       "               'endInSentence': 157}],\n",
       "             'However, by pre-enrichment for hypermutants using ScrF1 and AvaII digestion (see Materials and Methods), rare hypermutants could be observed in late time points in most cultures (Fig. 6, lanes f∗, i∗, j∗, k∗, and l∗).': [{'label': 'ScrF1',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 50,\n",
       "               'endInSentence': 55},\n",
       "              {'label': 'AvaII',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 60,\n",
       "               'endInSentence': 65}],\n",
       "             'Finally, hypermutated sequences were found in condition IX, where virus infection preceded PHA activation by 24 h.': [{'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 66,\n",
       "               'endInSentence': 81},\n",
       "              {'label': 'PHA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 91,\n",
       "               'endInSentence': 94}],\n",
       "             'Overall, these analyses establish that HIV-1 infection of resting CD4+ T cells either shortly before or shortly after activation generated abundant hypermutated sequences with high levels of G-to-A mutation, while infection after longer periods of activation yielded principally normal sequences.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 39,\n",
       "               'endInSentence': 54},\n",
       "              {'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 66,\n",
       "               'endInSentence': 69},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 214,\n",
       "               'endInSentence': 223}],\n",
       "             'These combined approaches provided 131 different HIV-1 protease sequences from virus cultures for analysis.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 55,\n",
       "               'endInSentence': 63}],\n",
       "             'However, when sequences from each culture condition were analyzed separately, an interesting pattern of dinucleotide context utilization was noted (Fig. 7B, lower graph).': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 104,\n",
       "               'endInSentence': 116}],\n",
       "             'In cultures IX, I, and II, where PHA either was not added or was added simultaneously with or after infection with the virus, the preference for GA over GG was not strong.': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 100,\n",
       "               'endInSentence': 109}],\n",
       "             'With prior PHA stimulation for 3 h or more (cultures III to VIII), the saturation of available GA dinucleotides occurred at 4 to 6 times that of available GG dinucleotides.': [{'label': 'GA dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 95,\n",
       "               'endInSentence': 111},\n",
       "              {'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 158,\n",
       "               'endInSentence': 171}],\n",
       "             'Finally, hypermutated and normal protease gene sequences recovered from cell culture were examined for the loss of informational content (Table 3).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 33,\n",
       "               'endInSentence': 41}],\n",
       "             'Hypermutated sequences could be detected in 43% of patients and in virus cultures when T-cell stimulation was initiated at the time of, or shortly before, virus infection.': [{'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 155,\n",
       "               'endInSentence': 170}],\n",
       "             'We examined the question of whether hypermutation affects GC and GT dinucleotides, although at lower levels, in addition to the preferred GA and GG contexts.': [{'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 68,\n",
       "               'endInSentence': 81}],\n",
       "             'However, our observations are limited to the HIV-1 protease gene, and examination of full HIV-1 genomes for the context of hypermutation would be required to fortify this conclusion.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 51,\n",
       "               'endInSentence': 59}],\n",
       "             'A consistent preference for saturation of available GA over GG dinucleotides within hypermutated sequences was also noted.': [{'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 63,\n",
       "               'endInSentence': 76}],\n",
       "             'Indeed, the DNA-binding preferences of bisbenzamide for runs of AT, particularly AAAA and AATT, would suggest that GA to AA would be more efficiently detected than GG to AG, but the fact that the vast majority of hypermutants we recovered used both contexts suggests that this was not a major obstacle to recovery.': [{'label': 'bisbenzamide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 39,\n",
       "               'endInSentence': 51}],\n",
       "             'The preferential utilization of available GA over GG dinucleotides over the whole dynamic range, and the fact that the vast majority of hypermutants utilized both GA and GG contexts, suggests that both contexts are recognized during a single reverse transcription event.': [{'label': 'GG dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 50,\n",
       "               'endInSentence': 66}],\n",
       "             'The high error rate of HIV-1 RT is apparently necessary to permit both “normal” mutation and hypermutation, but hypermutation should not be thought of as an elevated error rate for RT; it is clearly an independent process with separate biochemical requirements which must vary over a wide range during HIV-1 replication in vivo.': [{'label': 'RT',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 29,\n",
       "               'endInSentence': 31}],\n",
       "             'First, the evidence that true hypermutants were generated in CD4+ T cells in culture should be reiterated.': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 61,\n",
       "               'endInSentence': 64}],\n",
       "             'The virus stock was molecularly cloned and verified to be free of hypermutation by sequencing of the viral RNA by RT-PCR and by DNase treatment to remove any contaminating proviral DNA remaining from virus propagation.': [{'label': 'DNase',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 128,\n",
       "               'endInSentence': 133}],\n",
       "             'DNA PCR of the culture supernatants after infection was negative.': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 42,\n",
       "               'endInSentence': 51}],\n",
       "             'The identity of the target cell population was carefully established; the culture contained principally, if not exclusively, CD4+ T cells.': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 125,\n",
       "               'endInSentence': 128}],\n",
       "             'Most, if not all, CD4+ T cells were quiescent.': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 18,\n",
       "               'endInSentence': 21}],\n",
       "             'It has been previously documented that infection of quiescent cells results in the absence of virus progeny (68) and that viral DNA can be detected after infection of unstimulated cells although it is not associated with virus production (69, 70).': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 39,\n",
       "               'endInSentence': 48},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 154,\n",
       "               'endInSentence': 163}],\n",
       "             'In our experiments, proviral DNA was established but there was no detectable virus production in the absence of PHA stimulation, which is consistent with the infection of quiescent cells.': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 158,\n",
       "               'endInSentence': 167}],\n",
       "             'Most of the cells were HLA-DR negative.': [{'label': 'HLA-DR',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 23,\n",
       "               'endInSentence': 29}],\n",
       "             'It is highly probable that those cells that were HLA-DR+, ranging from 15 to 18% in different experiments, represented recently, but not currently, activated cells.': [{'label': 'HLA-DR',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 49,\n",
       "               'endInSentence': 55}],\n",
       "             'Hypermutants were recovered principally from virus cultures in which HIV-1 infection was simultaneous with, or briefly preceded by, PHA stimulation (Fig. 6).': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 69,\n",
       "               'endInSentence': 84}],\n",
       "             'Even the dinucleotide context preference is partially parsed out by these experiments; we note that in culture conditions XI, I, and II, where PHA either was not added or was added along with or after the virus, utilization of available GA and GG was almost equivalent.': [{'label': 'dinucleotide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 9,\n",
       "               'endInSentence': 21}],\n",
       "             'The strong bias for GA appeared once PHA preceded HIV-1 infection by only 3 h (Figure 7B).': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 50,\n",
       "               'endInSentence': 65}],\n",
       "             'A previous study demonstrated that hypermutants were recovered, although at low frequencies, after infection of unstimulated (2%) and stimulated (0.9%) bulk PBMC (62).': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 99,\n",
       "               'endInSentence': 108}],\n",
       "             'Such recovery of hypermutants was ascribed to virus infection of a small proportion of PBMC with distorted dNTP pools, reinforcing the concept of an occasional and sporadic event.': [{'label': 'virus infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 46,\n",
       "               'endInSentence': 61}],\n",
       "             'Because the composition of dNTP pools fluctuates as cells progress through the cell cycle (5, 20, 45), we investigated if the timing of HIV-1 infection with respect to T-cell stimulation could influence the generation of hypermutation.': [{'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 136,\n",
       "               'endInSentence': 151}],\n",
       "             'Another aspect is the fact that the vast majority of hypermutated protease sequences present in this study were associated with a disrupted coding potential.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 66,\n",
       "               'endInSentence': 74}],\n",
       "             'Since protease has a crucial role during the HIV-1 replication cycle, the loss of its coding potential integrity should lead to abortive infections.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 6,\n",
       "               'endInSentence': 14},\n",
       "              {'label': 'infections',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 137,\n",
       "               'endInSentence': 147}],\n",
       "             'They amplify a 297-bp segment of the HIV-1 genome encoding the protease gene, as indicated by their location in the reference sequence HXB2 (GenBank accession number AF033819).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 63,\n",
       "               'endInSentence': 71}],\n",
       "             'HIV-1 protease sequences amplified from patient PBMC and representing a range of G-to-A hypermutations were used to explore the performance of 1% agarose gels containing HA yellow, a dye that preferentially binds to AT-rich regions in DNA.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 6,\n",
       "               'endInSentence': 14},\n",
       "              {'label': 'HA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 170,\n",
       "               'endInSentence': 172}],\n",
       "             'The GA and GG dinucleotides that are susceptible to hypermutation are shaded.': [{'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 14,\n",
       "               'endInSentence': 27}],\n",
       "             'The gel on the left shows the migration of these PCR products as a single band at 297 bp without HA yellow.': [{'label': 'HA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 97,\n",
       "               'endInSentence': 99}],\n",
       "             'Primers were designed to incorporate G-to-A substitutions in GA or GG dinucleotides (Materials and Methods) and compared to primers matched to normal sequences for their ability to amplify hypermutated sequences from the PBMC of 10 patients.': [{'label': 'GG',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 67,\n",
       "               'endInSentence': 69},\n",
       "              {'label': 'dinucleotides',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 70,\n",
       "               'endInSentence': 83}],\n",
       "             'Military Seroconverters1997–1998AsymptomaticB147\\u2009(50) TanzaniaSentinel Surveillance1996AIDSA74\\u2009(57) C143\\u2009(21) D32\\u2009(67) Other22\\u2009(100) \\u2003Total5323\\u2009(43)aThe genetic subtype of the normal HIV-1 protease sequence was determined by phylogenetic analysis (data not shown).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 189,\n",
       "               'endInSentence': 197}],\n",
       "             'Other, mixed infection or unclassified.\\xa0bHypermutated HIV-1 sequences could be recovered from primary PBMC as described in Materials and Methods.': [{'label': 'mixed infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 7,\n",
       "               'endInSentence': 22}],\n",
       "             'Products were visualized by staining with ethidium bromide.': [{'label': 'ethidium bromide',\n",
       "               'type': 'CD',\n",
       "               'startInSentence': 42,\n",
       "               'endInSentence': 58}],\n",
       "             'HIV-1 protease genes were amplified using DNA extracted from the PBMC of 53 patients ranging from the early, asymptomatic stage of HIV-1 infection to late-stage AIDS, and including infections with HIV-1 subtypes A through D (Table 2).': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 6,\n",
       "               'endInSentence': 14},\n",
       "              {'label': 'HIV-1 infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 131,\n",
       "               'endInSentence': 146},\n",
       "              {'label': 'AIDS',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 161,\n",
       "               'endInSentence': 165},\n",
       "              {'label': 'infections',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 181,\n",
       "               'endInSentence': 191}],\n",
       "             'CD4+ T cells isolated from the PBMC of normal donors were infected with an NL4-3 virus stock for 2 h at a multiplicity of infection of 0.02 (stippled gray bar) under varying conditions of PHA activation (gray bars).': [{'label': 'CD4',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 0,\n",
       "               'endInSentence': 3},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 122,\n",
       "               'endInSentence': 131}],\n",
       "             'Condition I, infection without addition of PHA; condition II, simultaneous infection and PHA addition; conditions III to VIII, infection with prior PHA activation for 3 to 72 h respectively; condition IX, infection and addition of PHA 24 h later.': [{'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 13,\n",
       "               'endInSentence': 22},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 75,\n",
       "               'endInSentence': 84},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 127,\n",
       "               'endInSentence': 136},\n",
       "              {'label': 'PHA',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 148,\n",
       "               'endInSentence': 151},\n",
       "              {'label': 'infection',\n",
       "               'type': 'DS',\n",
       "               'startInSentence': 205,\n",
       "               'endInSentence': 214}],\n",
       "             'Cultures were sampled at intervals and assayed for the presence of p24 antigen in the culture supernatant (solid lines) and for the presence of hypermutated HIV-1 DNA sequences.': [{'label': 'p24',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 67,\n",
       "               'endInSentence': 70}],\n",
       "             'Samples from time points at which hypermutation was rare (f, i, j, k, and l) are shown both as bulk PCR products (B) and after pre-enrichment by ScrF1 and AvaII digestion (A).': [{'label': 'ScrF1',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 145,\n",
       "               'endInSentence': 150},\n",
       "              {'label': 'AvaII',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 155,\n",
       "               'endInSentence': 160}],\n",
       "             'A total of 227 HIV-1 protease sequences, 131 of which were unique, were obtained from virus cultures.': [{'label': 'protease',\n",
       "               'type': 'GP',\n",
       "               'startInSentence': 21,\n",
       "               'endInSentence': 29}]})"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mactch_gp_ds_cd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# {\n",
    "#   \"pmid\": \"PMID31911178\",\n",
    "#   \"pubDate\": \"YYYY-MM-DD\",\n",
    "#   \"organisms\": [\n",
    "#     \"TEXT\",\n",
    "#     \"TEXT\"\n",
    "#   ],\n",
    "#   \"sentences\": [\n",
    "#     {\n",
    "#       \"text\": \"TEXT\",\n",
    "#       \"section\": \"TEXT\",\n",
    "#       \"matches\": [\n",
    "#         {\n",
    "#           \"label\": \"NR2B\",\n",
    "#           \"type\": \"GP\",\n",
    "#           \"startInSentence\": \"INT\",\n",
    "#           \"endInSentence\": \"INT\",\n",
    "#           \"sectionStart\": \"INT, (nullable) - only for section == abstract\",\n",
    "#           \"sectionEnd\": \"INT (nullable) - only for section == abstract}]\"\n",
    "#         }\n",
    "#       ],\n",
    "#       \"co-occurrence\": [\n",
    "#         {\n",
    "#           \"label1\": \"TEXT\",\n",
    "#           \"label2\": \"TEXT\",\n",
    "#           \"type\": \"GP-DS\",\n",
    "#           \"start1\": \"INT\",\n",
    "#           \"end1\": \"INT\",\n",
    "#           \"start2\": \"INT\",\n",
    "#           \"end2\": \"INT\",\n",
    "#           \"scores\": [\n",
    "#             {\n",
    "#               \"evidence\": \"FLOAT\",\n",
    "#               \"Association\": \"FLOAT\",\n",
    "#               \"sentiment\": \"FLOAT\"\n",
    "#             },\n",
    "#             \"verbs\": [\n",
    "#             {\n",
    "#               \"positive_regulation\": \"TEXT\",\n",
    "#               \"negative_regulation\": \"TEXT\",\n",
    "#               \"expression\": \"TEXT\"\n",
    "#             }\n",
    "#           ]\n",
    "#         }\n",
    "#       ]\n",
    "#     }\n",
    "#   ]\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'Human Immunodeficiency Virus Type 1 DNA Sequences Genetically Damaged by Hypermutation Are Often Abundant in Patient Peripheral Blood Mononuclear Cells and May Be Generated during Near-Simultaneous Infection and Activation of CD4+ T Cells': [[[0,\n",
       "                35,\n",
       "                'OG',\n",
       "                'Human Immunodeficiency Virus Type 1'],\n",
       "               [198, 207, 'DS', 'Infection'],\n",
       "               [226, 229, 'GP', 'CD4']]],\n",
       "             'G-to-A hypermutation has been sporadically observed in human immunodeficiency virus type 1 (HIV-1) proviral sequences from patient peripheral blood mononuclear cells (PBMC) and virus cultures but has not been systematically evaluated.': [[[55,\n",
       "                90,\n",
       "                'OG',\n",
       "                'human immunodeficiency virus type 1'],\n",
       "               [92, 97, 'OG', 'HIV-1'],\n",
       "               [177, 182, 'OG', 'virus']]],\n",
       "             'PCR primers matched to normal and hypermutated sequences were used in conjunction with an agarose gel electrophoresis system incorporating an AT-binding dye to visualize, separate, clone, and sequence hypermutated and normal sequences in the 297-bp HIV-1 protease gene amplified from patient PBMC.': [[[249,\n",
       "                254,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [255, 263, 'GP', 'protease']]],\n",
       "             'Among 53 patients, including individuals infected with subtypes A through D and at different clinical stages, at least 43% of patients harbored abundant hypermutated, along with normal, protease genes.': [[[186,\n",
       "                194,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'In 70 hypermutated sequences, saturation of G residues in the GA or GG dinucleotide context ranged from 20 to 94%.': [[[71,\n",
       "                83,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'Levels of other mutants were not elevated, and G-to-A replacement was entirely restricted to GA or GG, and not GC or GT, dinucleotides.': [[[121,\n",
       "                134,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'To investigate the conditions under which hypermutation occurs in cell cultures, purified CD4+ T cells from normal donors were infected with cloned NL4-3 virus stocks at various times before and after phytohemagglutinin (PHA) activation.': [[[90,\n",
       "                93,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [148, 159, 'OG', 'NL4-3 virus'],\n",
       "               [201, 219, 'CD', 'phytohemagglutinin']]],\n",
       "             'Hypermutation was pronounced when HIV-1 infection occurred simultaneously with, or a few hours after, PHA activation, but after 12 h or more after PHA activation, most HIV-1 sequences were normal.': [[[34,\n",
       "                49,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [168, 173, 'OG', 'HIV-1']]],\n",
       "             'Near-simultaneous activation and infection of CD4+ T cells may represent a window of susceptibility where the informational content of HIV-1 sequences is lost due to hypermutation.': [[[33,\n",
       "                42,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [46, 49, 'GP', 'CD4'],\n",
       "               [135, 140, 'OG', 'HIV-1']]],\n",
       "             'There is loss of the information content of nucleotide sequences once mutation rates are increased beyond a tolerable error threshold (15).': [[[44,\n",
       "                54,\n",
       "                'CD',\n",
       "                'nucleotide']]],\n",
       "             'Some RNA viruses seem to tolerate mutation rates near this threshold, existing, not as a specific sequence, but as a quasispecies (12, 15, 26).': [[[5,\n",
       "                16,\n",
       "                'OG',\n",
       "                'RNA viruses']]],\n",
       "             'How surprising, then, that for some of the same viruses with the highest mutation rates, another mutation process has been described in which remarkable levels of one specific type of nucleotide substitution are observed.': [[[48,\n",
       "                55,\n",
       "                'OG',\n",
       "                'viruses'],\n",
       "               [184, 194, 'CD', 'nucleotide']]],\n",
       "             'A-to-G hypermutation occurs mostly in measles virus and vesicular stomatitis virus (8, 9, 49), and three cases in nonlentiviral retroviruses (avian leukosis virus and spleen necrosis virus) have been reported (16, 23, 33).': [[[38,\n",
       "                51,\n",
       "                'OG',\n",
       "                'measles virus'],\n",
       "               [56, 82, 'OG', 'vesicular stomatitis virus'],\n",
       "               [128, 140, 'OG', 'retroviruses'],\n",
       "               [142, 162, 'OG', 'avian leukosis virus'],\n",
       "               [167, 188, 'OG', 'spleen necrosis virus']]],\n",
       "             'G-to-A hypermutation is found primarily in the lentivirus family of retroviruses, along with two other examples in satellite tobacco mosaic virus (37) and hepatitis B virus (22).': [[[47,\n",
       "                57,\n",
       "                'OG',\n",
       "                'lentivirus'],\n",
       "               [68, 80, 'OG', 'retroviruses'],\n",
       "               [125, 145, 'OG', 'tobacco mosaic virus'],\n",
       "               [155, 172, 'OG', 'hepatitis B virus']]],\n",
       "             'G-to-A hypermutation in the lentivirus human immunodeficiency virus type 1 (HIV-1) is the subject of this report.': [[[28,\n",
       "                38,\n",
       "                'OG',\n",
       "                'lentivirus'],\n",
       "               [39, 74, 'OG', 'human immunodeficiency virus type 1'],\n",
       "               [76, 81, 'OG', 'HIV-1']]],\n",
       "             'G-to-A hypermutation occurs specifically within the GpA or GpG dinucleotide context (18, 60).': [[[52,\n",
       "                55,\n",
       "                'GP',\n",
       "                'GpA'],\n",
       "               [63, 75, 'CD', 'dinucleotide']]],\n",
       "             'Although G-to-A hypermutation was first described for spleen necrosis virus (SNV) by Pathak and Temin (50), G-to-A hypermutation has been mostly found among the lentiviral group of retroviruses, including HIV-1 (4, 11, 18, 21, 39, 40, 47, 60), HIV-2 (19), simian immunodeficiency virus (SIV) (30), equine infectious anemia virus (EIAV) (52), and caprine arthritis-encephalitis virus (CAEV) (64).': [[[54,\n",
       "                75,\n",
       "                'OG',\n",
       "                'spleen necrosis virus'],\n",
       "               [77, 80, 'OG', 'SNV'],\n",
       "               [181, 193, 'OG', 'retroviruses'],\n",
       "               [205, 210, 'OG', 'HIV-1'],\n",
       "               [244, 249, 'OG', 'HIV-2'],\n",
       "               [256, 285, 'OG', 'simian immunodeficiency virus'],\n",
       "               [287, 290, 'OG', 'SIV'],\n",
       "               [298, 328, 'OG', 'equine infectious anemia virus'],\n",
       "               [330, 334, 'OG', 'EIAV'],\n",
       "               [346, 382, 'OG', 'caprine arthritis-encephalitis virus'],\n",
       "               [384, 388, 'OG', 'CAEV']]],\n",
       "             'The susceptibility of lentiviruses to hypermutation is thought to be a property of their reverse transcriptase (RT), which, compared to nonlentiviral RTs, has a great capacity to elongate beyond nucleotide mismatches (51) and an increased ability to hypermutate in the presence of unbalanced nucleotide pools in vitro (43).': [[[22,\n",
       "                34,\n",
       "                'OG',\n",
       "                'lentiviruses'],\n",
       "               [89, 110, 'GP', 'reverse transcriptase'],\n",
       "               [112, 114, 'GP', 'RT'],\n",
       "               [150, 153, 'GP', 'RTs'],\n",
       "               [195, 205, 'CD', 'nucleotide'],\n",
       "               [292, 302, 'CD', 'nucleotide']]],\n",
       "             'This susceptibility is further evidenced by the elevated A content of lentivirus genomes (7).': [[[70,\n",
       "                80,\n",
       "                'OG',\n",
       "                'lentivirus']]],\n",
       "             'Hypermutation in HIV-1 was first detected during propagation of HIV-1 virus populations in vitro (11, 21, 60).': [[[17,\n",
       "                22,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [64, 75, 'OG', 'HIV-1 virus']]],\n",
       "             'Since then, several groups have recovered hypermutated HIV-1 sequences from clinical samples, confirming that they occur in vivo (4, 18, 40, 47).': [[[55,\n",
       "                60,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'With respect to genome region, hypermutation was initially found in small subregions of gag, env, nef, and U3/R elements of the long terminal repeat (LTR) (11, 21, 60), but others have found hypermutation throughout the genome (4).': [[[88,\n",
       "                91,\n",
       "                'GP',\n",
       "                'gag'],\n",
       "               [93, 96, 'GP', 'env'],\n",
       "               [98, 101, 'GP', 'nef']]],\n",
       "             'First thought to be the work of a mutant RT (18, 50), hypermutation has now been shown to be a property of wild-type RT (42) operating under suboptimal conditions.': [[[41,\n",
       "                43,\n",
       "                'GP',\n",
       "                'RT'],\n",
       "               [117, 119, 'GP', 'RT']]],\n",
       "             'G-to-A hypermutation is thought to occur when HIV minus-strand DNA synthesis takes place simultaneously with an increased intracellular concentration of dTTP relative to dCTP (60).': [[[46,\n",
       "                49,\n",
       "                'OG',\n",
       "                'HIV']]],\n",
       "             'Attempts have been made to generate hypermutants in a cell-free system in vitro with RNA, purified RT, and strongly biased deoxynucleoside triphosphate (dNTP) pools (43, 44, 59, 61).': [[[99,\n",
       "                101,\n",
       "                'GP',\n",
       "                'RT']]],\n",
       "             'While G-to-A hypermutants were produced, they lacked the marked preference for the GpA and GpG dinucleotide contexts that typify hypermutants produced in vivo or during virus cultivation in vitro.': [[[83,\n",
       "                86,\n",
       "                'GP',\n",
       "                'GpA'],\n",
       "               [95, 107, 'CD', 'dinucleotide'],\n",
       "               [169, 174, 'OG', 'virus']]],\n",
       "             'An intermediate approach where RT in virus particles was allowed to complete first-strand DNA synthesis in the presence of exogenously provided, highly biased dNTP pools also resulted in loss of the GpA and GpG dinucleotide preference (62).': [[[31,\n",
       "                33,\n",
       "                'GP',\n",
       "                'RT'],\n",
       "               [37, 42, 'OG', 'virus'],\n",
       "               [199, 202, 'GP', 'GpA'],\n",
       "               [211, 223, 'CD', 'dinucleotide']]],\n",
       "             'Hypermutants were also recovered at low frequency (0.5%) from cell cultures after the addition of deoxythymidine to supernatants, which resulted in an elevated intracellular [dTTP]/[dCTP] ratio (62), but because all possible dinucleotide contexts were not represented in the sequence that was studied, a complete analysis of context preference was not possible.': [[[98,\n",
       "                112,\n",
       "                'CD',\n",
       "                'deoxythymidine'],\n",
       "               [225, 237, 'CD', 'dinucleotide']]],\n",
       "             'More subtle perturbations in the intracellular environment may trigger dNTP pool biases sufficient to generate hypermutation while maintaining the proper dinucleotide context (67).': [[[154,\n",
       "                166,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'Indeed, a 40- fold [dTTP]/[dCTP] ratio produced hypermutants in which the GpA context preference was maintained (62), and hypermutants were recovered from 1 to 2% of unstimulated and phytohemagglutinin (PHA)-stimulated peripheral blood mononuclear cells (PBMC) without manipulation of dNTP pools (62).': [[[74,\n",
       "                77,\n",
       "                'GP',\n",
       "                'GpA'],\n",
       "               [183, 201, 'CD', 'phytohemagglutinin']]],\n",
       "             'Imbalanced and fluctuating nucleotide pools are a key element of many types of mutation, including hypermutation.': [[[27,\n",
       "                37,\n",
       "                'CD',\n",
       "                'nucleotide']]],\n",
       "             'Since the composition of dNTP pools changes as cells progress through the cell cycle (5, 20, 45), the timing of HIV-1 infection with respect to T-cell activation and entry into the cell cycle could be an important factor determining the generation of hypermutants.': [[[112,\n",
       "                127,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'In a retrovirus-based shuttle vector, G-to-A transitions predominated (31, 32) and were presumably a result of even these modest fluctuations in pools during the cell cycle.': [[[5,\n",
       "                15,\n",
       "                'OG',\n",
       "                'retrovirus']]],\n",
       "             'For example, a method based on the blue/white β-galactosidase complementation assay yielded a low recovery of hypermutants (0.5 to 2%) (62), not higher than that obtained when clones were randomly picked and sequenced.': [[[35,\n",
       "                61,\n",
       "                'GP',\n",
       "                'blue/white β-galactosidase']]],\n",
       "             'Given the lack of a specific genome location for hypermutation, the fact that hypermutants are almost always buried in a large excess of normal sequences, and the difficulty of precisely reproducing hypermutation in vitro, it is not surprising that this mutational process has been regarded as erratic, rare, and of minor importance in the HIV-1 life cycle.': [[[340,\n",
       "                345,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Here we describe the design and application of powerful new methods for systematic detection and recovery of hypermutants and their application to clinical samples and to the products of HIV-1 infection in cell culture.': [[[187,\n",
       "                202,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'The results call for a reassessment of the frequency of hypermutation in vivo, clarify the conditions that generate hypermutants in cell culture, and importantly, highlight a vulnerability of HIV-1 that could be exploited for clinical benefit.': [[[192,\n",
       "                197,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Purified PBMC from 53 HIV-1-positive individuals were the source of DNA for analysis.': [[[22,\n",
       "                27,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Twenty-six samples were from patients hospitalized in Tanzania in 1996 with symptoms compatible with AIDS (25).': [[[101,\n",
       "                105,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Fourteen others were from individuals who seroconverted to HIV-1 while in the U.S.': [[[59,\n",
       "                64,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The latter two groups were in the early, asymptomatic stage of HIV-1 infection, and the sample was typically drawn within 6 months of HIV-1 seroconversion.': [[[63,\n",
       "                78,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [134, 139, 'OG', 'HIV-1']]],\n",
       "             'DNA extracted from patient PBMC or virus cultures was the template for PCR amplification of HIV-1 protease sequences.': [[[35,\n",
       "                40,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [92, 97, 'OG', 'HIV-1'],\n",
       "               [98, 106, 'GP', 'protease']]],\n",
       "             'The primers for amplification of hypermutants contained either mixed bases (hyp primers) or G-to-A replacements (hypa primers) at some of the sites (GpA or GpG) susceptible to hypermutation.': [[[76,\n",
       "                79,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [113, 117, 'GP', 'hypa'],\n",
       "               [149, 152, 'GP', 'GpA'],\n",
       "               [156, 159, 'GP', 'GpG']]],\n",
       "             'Each PCR amplification was run in duplicate, once with normal primers and a second time with an equal mixture of hyp and hypa primers.': [[[113,\n",
       "                116,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [121, 125, 'GP', 'hypa']]],\n",
       "             'PCR products were analyzed in parallel 1% agarose gels, with or without HA yellow, a compound which consists of the DNA ligand bisbenzamide covalently linked to polyethylene glycol (PEG) (Hanse Analytik, Bremen, Germany).': [[[127,\n",
       "                139,\n",
       "                'CD',\n",
       "                'bisbenzamide']]],\n",
       "             'Bisbenzamide binds preferentially to AT-rich regions in DNA (in order from highest to lowest preference, AATT and AAAA, TAAT, ATAT, TATA, and TTAA) (1) and, when coupled to PEG, retards DNA mobility during gel electrophoresis according to AT content (41, 48, 65).': [[[0,\n",
       "                12,\n",
       "                'CD',\n",
       "                'Bisbenzamide']]],\n",
       "             'Products were visualized by poststaining with ethidium bromide.': [[[46,\n",
       "                62,\n",
       "                'CD',\n",
       "                'ethidium bromide']]],\n",
       "             'In experiments with virus cultivation in cell cultures, enrichment for hypermutated sequences by predigestion of the cellular DNA with restriction endonucleases ScrF1(CCAG∗G∗A) and AvaII (G∗G∗ACC), which preferentially cleave normal, as opposed to hypermutated, HIV-1 protease sequences, was used to recover rare hypermutated sequences from some samples.': [[[20,\n",
       "                25,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [135, 160, 'GP', 'restriction endonucleases'],\n",
       "               [161, 166, 'GP', 'ScrF1'],\n",
       "               [181, 186, 'GP', 'AvaII'],\n",
       "               [262, 267, 'OG', 'HIV-1'],\n",
       "               [268, 276, 'GP', 'protease']]],\n",
       "             'DNA was codigested with ScrF1 and AvaII at 37°C overnight prior to PCR amplification of HIV-1 protease.': [[[24,\n",
       "                29,\n",
       "                'GP',\n",
       "                'ScrF1'],\n",
       "               [34, 39, 'GP', 'AvaII'],\n",
       "               [88, 93, 'OG', 'HIV-1'],\n",
       "               [94, 102, 'GP', 'protease']]],\n",
       "             'Ligation into a plasmid vector, transformation of Escherichia coli, and plasmid purification were done using the TOPO TA cloning kit and One Shot Chemically Competent Cells from Invitrogen (San Diego, Calif.) and the Qiawell 8 ultra plasmid kit (Qiagen), respectively, as directed by the vendors.': [[[50,\n",
       "                66,\n",
       "                'OG',\n",
       "                'Escherichia coli']]],\n",
       "             'Eight hundred nanograms of purified plasmids was digested using 20 U of EcoR1 (New England Biolabs, Beverly, Mass.) for 1 h at 37°C to release the cloned protease gene, which was then rescreened on agarose gels with HA yellow.': [[[72,\n",
       "                77,\n",
       "                'GP',\n",
       "                'EcoR1'],\n",
       "               [154, 162, 'GP', 'protease']]],\n",
       "             'Both strands of the DNA were sequenced using fluorescent dye terminators present in the PRISM Ready Reaction Dyedeoxy Terminator kit with FS Taq (Applied Biosystems, Foster City, Calif.), and an Applied Biosystems Automated Sequencer as directed by the manufacturer.': [[[141,\n",
       "                144,\n",
       "                'OG',\n",
       "                'Taq']]],\n",
       "             'Determination of HIV-1 subtype.': [[[17, 22, 'OG', 'HIV-1']]],\n",
       "             'The genetic subtypes of normal protease sequences from the 53 patients were established by phylogenetic analysis (see Table 2).': [[[31,\n",
       "                39,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'Sequences were manually aligned with reference sequences of HIV-1 subtypes A through J.': [[[60,\n",
       "                65,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Among the AIDS patients more than half harbored subtype C; the remainder harbored mostly subtype A or D.': [[[10,\n",
       "                14,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Preparation and characterization of virus stocks.': [[[36,\n",
       "                41,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'Virus stocks generated from an infectious molecular clone of HIV-1 subtype B isolate NL4-3 (56, 57) were used to infect PBMC from a seronegative donor.': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'Virus'],\n",
       "               [61, 76, 'OG', 'HIV-1 subtype B'],\n",
       "               [85, 90, 'GP', 'NL4-3']]],\n",
       "             'Cells were activated with 1 μg of PHA/ml for 72 h, infected for 24 h, washed twice, and maintained in RPMI 1640 with 15% fetal calf serum (FCS), 1% PenStrep (Quality Biological, Gaithersburg, Md.), 1% l-glutamine, and 20 U of interleukin-2 (IL-2)/ml.': [[[201,\n",
       "                212,\n",
       "                'CD',\n",
       "                'l-glutamine'],\n",
       "               [226, 239, 'GP', 'interleukin-2'],\n",
       "               [241, 245, 'GP', 'IL-2']]],\n",
       "             'Culture supernatants were assayed for p24 antigen (HIV-1 p24 Antigen Assay Kit; Coulter Corporation, Miami, Fla.) at 3-day intervals.': [[[38,\n",
       "                41,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [51, 56, 'OG', 'HIV-1'],\n",
       "               [57, 60, 'GP', 'p24']]],\n",
       "             'Culture supernatants collected near the peak of p24 antigen production were treated with 50 U of DNAse I/ml (Boehringer Mannheim) for 30 min at room temperature and filtered sterilized (pore size, 0.22 μm).': [[[48,\n",
       "                51,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [97, 104, 'GP', 'DNAse I']]],\n",
       "             'To verify that the virus stocks were free of hypermutated sequences, RNA was extracted from 200 μl of virus stocks using the Nuclisens Extraction Kit (Organon Teknika), and RT-PCR was performed using the Reverse Transcription System (Promega, Madison, Wis.) with primer DP11.': [[[19,\n",
       "                24,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [102, 107, 'OG', 'virus']]],\n",
       "             'All virus stocks were free of hypermutants by these assays.': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'Fresh leucopacks were obtained from HIV-1-seronegative donors (RH Laboratories, Baltimore, Md.).': [[[36,\n",
       "                41,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'CD4+ T cells were purified by negative selection on magnetic beads with the MACS CD4+ T Cell Isolation Kit (Miltenyi Biotec, Auburn, Calif.).': [[[0,\n",
       "                3,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [81, 84, 'GP', 'CD4']]],\n",
       "             'Prior to infection, an aliquot of the purified CD4+ T cells was fixed with 2% formaldehyde and stained with fluorescein isothiocyanate (FITC)- or phycoerythrin (PE)-conjugated monoclonal antibodies specific for cell surface molecules CD3, CD4, CD8, CD14, CD20, CD56, and HLA-DR (Becton Dickinson, San Jose, Calif.).': [[[9,\n",
       "                18,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [47, 50, 'GP', 'CD4'],\n",
       "               [146, 159, 'CD', 'phycoerythrin'],\n",
       "               [234, 237, 'GP', 'CD3'],\n",
       "               [239, 242, 'GP', 'CD4'],\n",
       "               [244, 247, 'GP', 'CD8'],\n",
       "               [249, 253, 'GP', 'CD14'],\n",
       "               [255, 259, 'GP', 'CD20'],\n",
       "               [261, 265, 'GP', 'CD56'],\n",
       "               [271, 277, 'GP', 'HLA-DR']]],\n",
       "             'The majority of the cells in each experiment were CD4+ T cells (mean, 81%; range, 69 to 99%) representing most of the CD3+ T-cell subset (mean, 83%; range, 76 to 100%).': [[[50,\n",
       "                53,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [118, 121, 'GP', 'CD3']]],\n",
       "             'The other major cell type was B cells (CD20+; mean, 10.8%; range, 4.5 to 14.9%).': [[[39,\n",
       "                43,\n",
       "                'GP',\n",
       "                'CD20']]],\n",
       "             'Monocytes (CD14+), CD8+ T cells (CD8+), and NK cells (CD56+) were, on average, 2 to 4% of the population.': [[[11,\n",
       "                15,\n",
       "                'GP',\n",
       "                'CD14'],\n",
       "               [19, 22, 'GP', 'CD8'],\n",
       "               [33, 36, 'GP', 'CD8'],\n",
       "               [54, 58, 'GP', 'CD56']]],\n",
       "             'Between 14.6 and 17.7% of the cells were HLA-DR positive.': [[[41,\n",
       "                47,\n",
       "                'GP',\n",
       "                'HLA-DR']]],\n",
       "             'Cells were either left unstimulated or stimulated with 1 μg of PHA (Pharmacia)/ml for various intervals before, during, or after virus infection at a multiplicity of infection of 0.02 for 2 h at 37°C.': [[[129,\n",
       "                144,\n",
       "                'DS',\n",
       "                'virus infection'],\n",
       "               [166, 175, 'DS', 'infection']]],\n",
       "             'Cells were washed twice and maintained in RPMI 1640 containing 10% autologous plasma, and IL-2 at 20 U/ml (Amersham).': [[[90,\n",
       "                94,\n",
       "                'GP',\n",
       "                'IL-2']]],\n",
       "             'Autologous plasma, necessary to prevent activation of the unstimulated CD4+ T cells by foreign antigens, was used throughout.': [[[71,\n",
       "                74,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'Supernatants from cell cultures were evaluated for p24 antigen production using the Coulter HIV-1 p24 Antigen Assay Kit.': [[[51,\n",
       "                54,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [92, 97, 'OG', 'HIV-1'],\n",
       "               [98, 101, 'GP', 'p24']]],\n",
       "             'Nine different culture conditions were established (see Fig. 6) as follows: Culture I, virus infection without PHA stimulation; culture II, PHA and virus added simultaneously for 2 h; cultures III through VIII, PHA stimulation preceding virus infection by 3, 6, 12, 24, 48, or 72 h, respectively; culture IX, addition of virus for 2 h, washing away of virus, and PHA addition 24 h later.': [[[87,\n",
       "                102,\n",
       "                'DS',\n",
       "                'virus infection'],\n",
       "               [111, 114, 'GP', 'PHA'],\n",
       "               [148, 153, 'OG', 'virus'],\n",
       "               [211, 214, 'GP', 'PHA'],\n",
       "               [237, 252, 'DS', 'virus infection'],\n",
       "               [321, 326, 'OG', 'virus'],\n",
       "               [352, 357, 'OG', 'virus']]],\n",
       "             'Cells and culture supernatants were collected at 6, 12, 24, 48, and 72 h after infection, with the exception of culture condition IX, where collection took place at 6, 12, 24, 48, and 72 h after addition of PHA.': [[[79,\n",
       "                88,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'Parameters of hypermutation were evaluated with the Hypermut Program Package (54) (http://www.hiv.lanl.gov/HYPERMUT/hypermut.html) as implemented at the Los Alamos HIV Sequence Database, Los Alamos, N.': [[[164,\n",
       "                167,\n",
       "                'OG',\n",
       "                'HIV']]],\n",
       "             'This program identifies mutations and their dinucleotide context with respect to a reference sequence that is provided with each alignment.': [[[44,\n",
       "                56,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'For virus cultures, the sequence of NL4-3, obtained by RT-PCR and sequencing of the virus stock, was used as the reference.': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [36, 41, 'GP', 'NL4-3'],\n",
       "               [84, 89, 'OG', 'virus']]],\n",
       "             'Nucleotide sequence accession numbers.': [[[0,\n",
       "                10,\n",
       "                'CD',\n",
       "                'Nucleotide']]],\n",
       "             'All the HIV-1 sequences related to this work have been submitted to GenBank and were given accession numbers A Y036228 through A Y036577.': [[[8,\n",
       "                13,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Nested primers which amplify a 297-bp segment encoding HIV-1 protease were examined for GA and GG dinucleotides, which are susceptible to hypermutation.': [[[55,\n",
       "                60,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [61, 69, 'GP', 'protease'],\n",
       "               [95, 97, 'CD', 'GG'],\n",
       "               [98, 111, 'CD', 'dinucleotides']]],\n",
       "             'New primers were designed with either incorporated GA or CT mixtures at some of these sites (hyp primers) or replaced G with A or C with T (hypa primers).': [[[93,\n",
       "                96,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [140, 144, 'GP', 'hypa']]],\n",
       "             'The hyp and hypa primers were also shortened to bring potential mispairs closer to the 3′ end of the primer, where they could exert the maximal effect in destabilizing interaction with normal sequences.': [[[4,\n",
       "                7,\n",
       "                'GP',\n",
       "                'hyp'],\n",
       "               [12, 16, 'GP', 'hypa']]],\n",
       "             'The dye HA yellow, in which consists of the DNA-binding ligand bisbenzamide coupled to PEG, was incorporated into 1% agarose gels.': [[[8,\n",
       "                17,\n",
       "                'CD',\n",
       "                'HA yellow']]],\n",
       "             'A series of cloned protease sequences representing different A+T contents, amplified and sequenced previously in the pilot phase of this project, were used for evaluation of the gel system (Fig. 1).': [[[19,\n",
       "                27,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'HIV-1 proviral DNA sequences from 53 HIV-1-positive patients were investigated (Table 2).': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [37, 42, 'OG', 'HIV-1']]],\n",
       "             'The appearance of protease genes PCR amplified from patient PBMC is illustrated in Fig. 3.': [[[18,\n",
       "                26,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'While all of the protease genes were of uniform length without HA yellow, they often separated into multiple bands of different A+T content on HA yellow gels.': [[[17,\n",
       "                25,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'The sequence data established that, among the 53 patients, 23 (43%) harbored a mixture of normal and hypermutated HIV-1 sequences, while from the remaining 30 patients, only normal sequences were recovered by the techniques used (Table 2).': [[[114,\n",
       "                119,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Hypermutated sequences were found in both asymptomatic and AIDS patients, and in patients with infections with four different HIV-1 subtypes.': [[[59,\n",
       "                63,\n",
       "                'DS',\n",
       "                'AIDS'],\n",
       "               [95, 105, 'DS', 'infections'],\n",
       "               [126, 131, 'OG', 'HIV-1']]],\n",
       "             'This permitted, for the first time, a systematic evaluation of the parameters of HIV-1 hypermutation in vivo.': [[[81,\n",
       "                86,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The dinucleotide context of hypermutation was then considered (Fig. 5, bottom).': [[[4,\n",
       "                16,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'With respect to the saturation of the available GA and GG dinucleotides, at all levels of G-to-A replacement, the percentage of G replaced by A in the dinucleotide context GA was about twofold higher than the percentage of G replaced by A in the GG context.': [[[58,\n",
       "                71,\n",
       "                'CD',\n",
       "                'dinucleotides'],\n",
       "               [151, 163, 'CD', 'dinucleotide']]],\n",
       "             'Although G-to-A substitutions in the GC and GT dinucleotide contexts were previously referred as hypermutation related (60), the observed number of these replacements was limited, ranging from 0 to 3 (mean, 0.6) per hypermutated sequence.': [[[47,\n",
       "                59,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'At the highest levels of G-to-A mutation, virtually all of the GA and GG dinucleotides were mutated, while G-to-A mutations in GC or GT (other mutations; Fig. 5) were not elevated.': [[[73,\n",
       "                86,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'In summary, hypermutated HIV-1 protease sequences recovered from patient PBMC exhibited a broad range of G-to-A mutation that was entirely restricted to the GA or GG dinucleotide context, with a consistent preference for G-to-A mutation in the dinucleotide GA relative to GG at all levels of G-to-A saturation.': [[[25,\n",
       "                30,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [31, 39, 'GP', 'protease'],\n",
       "               [166, 178, 'CD', 'dinucleotide'],\n",
       "               [244, 256, 'CD', 'dinucleotide']]],\n",
       "             'In addition, there was a net accumulation of positively charged amino acids, mostly from decreased Gly, Asp, and Arg and increased Lys and Asn in hypermutants.': [[[64,\n",
       "                75,\n",
       "                'CD',\n",
       "                'amino acids']]],\n",
       "             'The proteases encoded by hypermutated sequences had, on average, a fivefold increase in net charge (Table 3).': [[[4,\n",
       "                13,\n",
       "                'GP',\n",
       "                'proteases']]],\n",
       "             'It is noteworthy that hypermutated sequences with the lowest levels of G-to-A substitutions sustained inactivating in-frame stop codons and many nonsynonymous nucleotide substitutions.': [[[159,\n",
       "                169,\n",
       "                'CD',\n",
       "                'nucleotide']]],\n",
       "             'These combined effects make it highly unlikely that functional proteases, and, consequently, viable viruses can be derived from hypermutated sequences.': [[[63,\n",
       "                72,\n",
       "                'GP',\n",
       "                'proteases'],\n",
       "               [100, 107, 'OG', 'viruses']]],\n",
       "             'The relationships between hypermutation, clinical stage, and HIV-1 subtype were investigated (Table 2).': [[[61,\n",
       "                66,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Hypermutated sequences were recovered at essentially the same rate from AIDS patients (42%) and early asymptomatic individuals (44%).': [[[72,\n",
       "                76,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Hypermutation was found in 57% of subtype A, 67% of subtype D, and 44% of subtype B infections, as well as in A+D dually infected and unclassified samples.': [[[84,\n",
       "                94,\n",
       "                'DS',\n",
       "                'infections']]],\n",
       "             'Although subtype C was the most frequently represented subtype among AIDS patients, hypermutation was detected only in 3 of 14 (21%) patients infected with this subtype.': [[[69,\n",
       "                73,\n",
       "                'DS',\n",
       "                'AIDS']]],\n",
       "             'Experiments with cell-free systems have shown that unbalances in dNTP pools during reverse transcription can lead to hypermutation, but the specificity for the GA and GG dinucleotide context has been difficult to reproduce.': [[[170,\n",
       "                182,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'On the other hand, hypermutants in the correct dinucleotide context have been recovered from HIV-1 cultures in PBMC, but the exact nature of the cell population that produced them and the status of the dNTP pools in this population have not been elucidated.': [[[47,\n",
       "                59,\n",
       "                'CD',\n",
       "                'dinucleotide'],\n",
       "               [93, 98, 'OG', 'HIV-1']]],\n",
       "             'In CD4+ T cells, which are a main target of HIV 1 replication in vivo, dNTP pools are elevated when the cells are activated and in cycle (20) but are unbalanced and fluctuating in resting T cells and during the transition from a resting to an activated state (5, 10, 20, 58).': [[[3,\n",
       "                6,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [44, 49, 'OG', 'HIV 1']]],\n",
       "             'We designed an experiment in which the timing of HIV-1 infection was varied with respect to activation of resting CD4+ T cells by PHA; the outcome measure was recovery of hypermutated sequences whose characteristics corresponded exactly to those found in vivo.': [[[49,\n",
       "                64,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [114, 117, 'GP', 'CD4']]],\n",
       "             'Proviral DNA was established in all culture conditions, as evidenced by its recovery by DNA PCR; a mock-infected culture was HIV-1 DNA PCR negative (data not shown).': [[[125,\n",
       "                130,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'p24 antigen in the culture supernatant, indicative of virus production, was observed only when cultures were stimulated with PHA either before, or simultaneously with, virus infection.': [[[0,\n",
       "                3,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [54, 59, 'OG', 'virus'],\n",
       "               [168, 183, 'DS', 'virus infection']]],\n",
       "             'The HIV-1 protease gene was amplified from all time points and culture conditions (Materials and Methods) and evaluated for the presence of normal and hypermutated sequences using the HA yellow gel system (Fig. 6).': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [10, 18, 'GP', 'protease']]],\n",
       "             'The richest source of hypermutants was culture II, in which infection and activation occurred simultaneously (Fig. 6A, lanes b through e).': [[[60,\n",
       "                69,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'Hypermutated sequences could also be visualized in conditions III and IV (lanes g and h), which represented PHA activation for less than 8 h before infection.': [[[108,\n",
       "                111,\n",
       "                'GP',\n",
       "                'PHA'],\n",
       "               [148, 157, 'DS', 'infection']]],\n",
       "             'However, by pre-enrichment for hypermutants using ScrF1 and AvaII digestion (see Materials and Methods), rare hypermutants could be observed in late time points in most cultures (Fig. 6, lanes f∗, i∗, j∗, k∗, and l∗).': [[[50,\n",
       "                55,\n",
       "                'GP',\n",
       "                'ScrF1'],\n",
       "               [60, 65, 'GP', 'AvaII']]],\n",
       "             'Finally, hypermutated sequences were found in condition IX, where virus infection preceded PHA activation by 24 h.': [[[66,\n",
       "                81,\n",
       "                'DS',\n",
       "                'virus infection'],\n",
       "               [91, 94, 'GP', 'PHA']]],\n",
       "             'Overall, these analyses establish that HIV-1 infection of resting CD4+ T cells either shortly before or shortly after activation generated abundant hypermutated sequences with high levels of G-to-A mutation, while infection after longer periods of activation yielded principally normal sequences.': [[[39,\n",
       "                54,\n",
       "                'DS',\n",
       "                'HIV-1 infection'],\n",
       "               [66, 69, 'GP', 'CD4'],\n",
       "               [214, 223, 'DS', 'infection']]],\n",
       "             'These combined approaches provided 131 different HIV-1 protease sequences from virus cultures for analysis.': [[[49,\n",
       "                54,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [55, 63, 'GP', 'protease'],\n",
       "               [79, 84, 'OG', 'virus']]],\n",
       "             'The normal sequences were narrowly distributed around 62% A+T, that of the NL4-3 virus innoculum.': [[[75,\n",
       "                86,\n",
       "                'OG',\n",
       "                'NL4-3 virus']]],\n",
       "             'However, when sequences from each culture condition were analyzed separately, an interesting pattern of dinucleotide context utilization was noted (Fig. 7B, lower graph).': [[[104,\n",
       "                116,\n",
       "                'CD',\n",
       "                'dinucleotide']]],\n",
       "             'In cultures IX, I, and II, where PHA either was not added or was added simultaneously with or after infection with the virus, the preference for GA over GG was not strong.': [[[100,\n",
       "                109,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [119, 124, 'OG', 'virus']]],\n",
       "             'With prior PHA stimulation for 3 h or more (cultures III to VIII), the saturation of available GA dinucleotides occurred at 4 to 6 times that of available GG dinucleotides.': [[[95,\n",
       "                111,\n",
       "                'CD',\n",
       "                'GA dinucleotides'],\n",
       "               [158, 171, 'CD', 'dinucleotides']]],\n",
       "             'Finally, hypermutated and normal protease gene sequences recovered from cell culture were examined for the loss of informational content (Table 3).': [[[33,\n",
       "                41,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'This study represents the first systematic evaluation of patient PBMC and virus cultures with a consistently applied methodology optimized for detection of hypermutated HIV-1 sequences.': [[[74,\n",
       "                79,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [169, 174, 'OG', 'HIV-1']]],\n",
       "             'Hypermutated sequences could be detected in 43% of patients and in virus cultures when T-cell stimulation was initiated at the time of, or shortly before, virus infection.': [[[67,\n",
       "                72,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [155, 170, 'DS', 'virus infection']]],\n",
       "             'The nearly identical quantitative parameters of the hypermutants recovered from patients and from virus cultures suggest that T cells in the process of becoming activated are also the main source of hypermutation in vivo.': [[[98,\n",
       "                103,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'The excessive accumulation of nonsynonymous substitutions and stop codon formation observed in hypermutants reinforces the perception that they can no longer generate viable viruses.': [[[174,\n",
       "                181,\n",
       "                'OG',\n",
       "                'viruses']]],\n",
       "             'We examined the question of whether hypermutation affects GC and GT dinucleotides, although at lower levels, in addition to the preferred GA and GG contexts.': [[[68,\n",
       "                81,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'However, our observations are limited to the HIV-1 protease gene, and examination of full HIV-1 genomes for the context of hypermutation would be required to fortify this conclusion.': [[[45,\n",
       "                50,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [51, 59, 'GP', 'protease'],\n",
       "               [90, 95, 'OG', 'HIV-1']]],\n",
       "             'A consistent preference for saturation of available GA over GG dinucleotides within hypermutated sequences was also noted.': [[[63,\n",
       "                76,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'Indeed, the DNA-binding preferences of bisbenzamide for runs of AT, particularly AAAA and AATT, would suggest that GA to AA would be more efficiently detected than GG to AG, but the fact that the vast majority of hypermutants we recovered used both contexts suggests that this was not a major obstacle to recovery.': [[[39,\n",
       "                51,\n",
       "                'CD',\n",
       "                'bisbenzamide']]],\n",
       "             'Many hypermutated proviruses may have been missed by our procedure.': [[[18,\n",
       "                28,\n",
       "                'OG',\n",
       "                'proviruses']]],\n",
       "             'Here we show that the process is completely independent of the normal accumulation of mutations in HIV-1 and covers a much broader dynamic range.': [[[99,\n",
       "                104,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The preferential utilization of available GA over GG dinucleotides over the whole dynamic range, and the fact that the vast majority of hypermutants utilized both GA and GG contexts, suggests that both contexts are recognized during a single reverse transcription event.': [[[50,\n",
       "                66,\n",
       "                'CD',\n",
       "                'GG dinucleotides']]],\n",
       "             'The high error rate of HIV-1 RT is apparently necessary to permit both “normal” mutation and hypermutation, but hypermutation should not be thought of as an elevated error rate for RT; it is clearly an independent process with separate biochemical requirements which must vary over a wide range during HIV-1 replication in vivo.': [[[23,\n",
       "                28,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [29, 31, 'GP', 'RT'],\n",
       "               [302, 307, 'OG', 'HIV-1']]],\n",
       "             'Could hypermutation be influenced by viral variation, or is it essentially a passive function of the condition of the cell population in which HIV-1 replicates?': [[[143,\n",
       "                148,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Stronger evidence would be needed, including larger numbers of patients with a broad range of clinical stages, to establish a definitive relationship between hypermutation and HIV-1 subtype.': [[[176,\n",
       "                181,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'The experiments with virus cultures are illuminating with respect to the conditions under which hypermutation occurs.': [[[21,\n",
       "                26,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'First, the evidence that true hypermutants were generated in CD4+ T cells in culture should be reiterated.': [[[61,\n",
       "                64,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'The target cell populations were free of HIV-1 because DNA extracted from the cells of mock-infected cultures carried in parallel were HIV-1 PCR negative.': [[[41,\n",
       "                46,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [135, 140, 'OG', 'HIV-1']]],\n",
       "             'The virus stock was molecularly cloned and verified to be free of hypermutation by sequencing of the viral RNA by RT-PCR and by DNase treatment to remove any contaminating proviral DNA remaining from virus propagation.': [[[4,\n",
       "                9,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [128, 133, 'GP', 'DNase'],\n",
       "               [200, 205, 'OG', 'virus']]],\n",
       "             'DNA PCR of the culture supernatants after infection was negative.': [[[42,\n",
       "                51,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'The identity of the target cell population was carefully established; the culture contained principally, if not exclusively, CD4+ T cells.': [[[125,\n",
       "                128,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'The major contaminant was B cells, which do not support HIV-1 replication.': [[[56,\n",
       "                61,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Most, if not all, CD4+ T cells were quiescent.': [[[18,\n",
       "                21,\n",
       "                'GP',\n",
       "                'CD4']]],\n",
       "             'It has been previously documented that infection of quiescent cells results in the absence of virus progeny (68) and that viral DNA can be detected after infection of unstimulated cells although it is not associated with virus production (69, 70).': [[[39,\n",
       "                48,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [94, 99, 'OG', 'virus'],\n",
       "               [154, 163, 'DS', 'infection'],\n",
       "               [221, 226, 'OG', 'virus']]],\n",
       "             'In our experiments, proviral DNA was established but there was no detectable virus production in the absence of PHA stimulation, which is consistent with the infection of quiescent cells.': [[[77,\n",
       "                82,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [158, 167, 'DS', 'infection']]],\n",
       "             'Most of the cells were HLA-DR negative.': [[[23,\n",
       "                29,\n",
       "                'GP',\n",
       "                'HLA-DR']]],\n",
       "             'It is highly probable that those cells that were HLA-DR+, ranging from 15 to 18% in different experiments, represented recently, but not currently, activated cells.': [[[49,\n",
       "                55,\n",
       "                'GP',\n",
       "                'HLA-DR']]],\n",
       "             'Hypermutants were recovered principally from virus cultures in which HIV-1 infection was simultaneous with, or briefly preceded by, PHA stimulation (Fig. 6).': [[[45,\n",
       "                50,\n",
       "                'OG',\n",
       "                'virus'],\n",
       "               [69, 84, 'DS', 'HIV-1 infection']]],\n",
       "             'Even the dinucleotide context preference is partially parsed out by these experiments; we note that in culture conditions XI, I, and II, where PHA either was not added or was added along with or after the virus, utilization of available GA and GG was almost equivalent.': [[[9,\n",
       "                21,\n",
       "                'CD',\n",
       "                'dinucleotide'],\n",
       "               [205, 210, 'OG', 'virus']]],\n",
       "             'The strong bias for GA appeared once PHA preceded HIV-1 infection by only 3 h (Figure 7B).': [[[50,\n",
       "                65,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'A previous study demonstrated that hypermutants were recovered, although at low frequencies, after infection of unstimulated (2%) and stimulated (0.9%) bulk PBMC (62).': [[[99,\n",
       "                108,\n",
       "                'DS',\n",
       "                'infection']]],\n",
       "             'Such recovery of hypermutants was ascribed to virus infection of a small proportion of PBMC with distorted dNTP pools, reinforcing the concept of an occasional and sporadic event.': [[[46,\n",
       "                61,\n",
       "                'DS',\n",
       "                'virus infection']]],\n",
       "             'Because the composition of dNTP pools fluctuates as cells progress through the cell cycle (5, 20, 45), we investigated if the timing of HIV-1 infection with respect to T-cell stimulation could influence the generation of hypermutation.': [[[136,\n",
       "                151,\n",
       "                'DS',\n",
       "                'HIV-1 infection']]],\n",
       "             'The target cell population for hypermutation must be continually renewed in vivo, commensurate with the proportion of resting T cells that become newly activated during active virus replication.': [[[176,\n",
       "                181,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'Such sequences should be included in conceptualizations of the latent reservoir of HIV-1 in patients.': [[[83,\n",
       "                88,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Is hypermutation related to the observation that the genomes of HIV-1 and other lentiviruses are AT rich?': [[[64,\n",
       "                69,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [80, 92, 'OG', 'lentiviruses']]],\n",
       "             'The total and unrecoverable loss of coding potential of hypermutated HIV sequences (Table 3) would suggest that they are incapable of generating progeny virions and cannot contribute to the HIV-1 gene pool.': [[[69,\n",
       "                72,\n",
       "                'OG',\n",
       "                'HIV'],\n",
       "               [190, 195, 'OG', 'HIV-1']]],\n",
       "             'It will be important to determine what fraction of the genome in hypermutants remains unaffected, as well as the potential for hypermutated proviruses to generate viable viruses through recombination.': [[[140,\n",
       "                150,\n",
       "                'OG',\n",
       "                'proviruses'],\n",
       "               [170, 177, 'OG', 'viruses']]],\n",
       "             'Hypermutated proviruses could represent a difficult-to-eradicate remnant of HIV-1 during highly active antiretroviral therapy.': [[[13,\n",
       "                23,\n",
       "                'OG',\n",
       "                'proviruses'],\n",
       "               [76, 81, 'OG', 'HIV-1']]],\n",
       "             'Another aspect is the fact that the vast majority of hypermutated protease sequences present in this study were associated with a disrupted coding potential.': [[[66,\n",
       "                74,\n",
       "                'GP',\n",
       "                'protease']]],\n",
       "             'Since protease has a crucial role during the HIV-1 replication cycle, the loss of its coding potential integrity should lead to abortive infections.': [[[6,\n",
       "                14,\n",
       "                'GP',\n",
       "                'protease'],\n",
       "               [45, 50, 'OG', 'HIV-1'],\n",
       "               [137, 147, 'DS', 'infections']]],\n",
       "             'Our data demonstrated that hypermutation may reduce the pool of viable replicating genomes in an infecting HIV-1 population.': [[[107,\n",
       "                112,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Detection of HIV-1 hypermutated sequences in at least 43% of patients studied suggests that hypermutation may happen in a systematic way in HIV-1-infected individuals.': [[[13,\n",
       "                18,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [140, 145, 'OG', 'HIV-1']]],\n",
       "             'Based on these two observations one can speculate that hypermutation may be seen as a result of a host-associated mechanism which may decrease virus replication.': [[[143,\n",
       "                148,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'A better understanding of the metabolic profile of early proliferating T cells in relation to their expanding dNTP pools would lead to a better interpretation of the correlation between temporal aspects of cell cycle metabolism and HIV-1 hypermutation.': [[[232,\n",
       "                237,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'This interpretation could develop into new strategies of promoting a hypermutation-inducible state in HIV-1 target cells, leading to irreversible mutagenesis of viral genomes.': [[[102,\n",
       "                107,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'A better comprehension of the cellular mechanisms involved in HIV-1 hypermutation could open new avenues for virus control.': [[[62,\n",
       "                67,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [109, 114, 'OG', 'virus']]],\n",
       "             'They amplify a 297-bp segment of the HIV-1 genome encoding the protease gene, as indicated by their location in the reference sequence HXB2 (GenBank accession number AF033819).': [[[37,\n",
       "                42,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [63, 71, 'GP', 'protease']]],\n",
       "             'HIV-1 protease sequences amplified from patient PBMC and representing a range of G-to-A hypermutations were used to explore the performance of 1% agarose gels containing HA yellow, a dye that preferentially binds to AT-rich regions in DNA.': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [6, 14, 'GP', 'protease'],\n",
       "               [170, 172, 'GP', 'HA']]],\n",
       "             'The GA and GG dinucleotides that are susceptible to hypermutation are shaded.': [[[14,\n",
       "                27,\n",
       "                'CD',\n",
       "                'dinucleotides']]],\n",
       "             'The gel on the left shows the migration of these PCR products as a single band at 297 bp without HA yellow.': [[[97,\n",
       "                99,\n",
       "                'GP',\n",
       "                'HA']]],\n",
       "             'Primers were designed to incorporate G-to-A substitutions in GA or GG dinucleotides (Materials and Methods) and compared to primers matched to normal sequences for their ability to amplify hypermutated sequences from the PBMC of 10 patients.': [[[67,\n",
       "                69,\n",
       "                'CD',\n",
       "                'GG'],\n",
       "               [70, 83, 'CD', 'dinucleotides']]],\n",
       "             'TABLE 2Evaluation of patient PBMC for the presence of hypermutated HIV-1 sequences CountryStudyDateClinical StageSubtypeaNo. of patientsNo.': [[[67,\n",
       "                72,\n",
       "                'OG',\n",
       "                'HIV-1']]],\n",
       "             'Military Seroconverters1997–1998AsymptomaticB147\\u2009(50) TanzaniaSentinel Surveillance1996AIDSA74\\u2009(57) C143\\u2009(21) D32\\u2009(67) Other22\\u2009(100) \\u2003Total5323\\u2009(43)aThe genetic subtype of the normal HIV-1 protease sequence was determined by phylogenetic analysis (data not shown).': [[[183,\n",
       "                188,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [189, 197, 'GP', 'protease']]],\n",
       "             'Other, mixed infection or unclassified.\\xa0bHypermutated HIV-1 sequences could be recovered from primary PBMC as described in Materials and Methods.': [[[7,\n",
       "                22,\n",
       "                'DS',\n",
       "                'mixed infection'],\n",
       "               [54, 59, 'OG', 'HIV-1']]],\n",
       "             'Products were visualized by staining with ethidium bromide.': [[[42,\n",
       "                58,\n",
       "                'CD',\n",
       "                'ethidium bromide']]],\n",
       "             'HIV-1 protease genes were amplified using DNA extracted from the PBMC of 53 patients ranging from the early, asymptomatic stage of HIV-1 infection to late-stage AIDS, and including infections with HIV-1 subtypes A through D (Table 2).': [[[0,\n",
       "                5,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [6, 14, 'GP', 'protease'],\n",
       "               [131, 146, 'DS', 'HIV-1 infection'],\n",
       "               [161, 165, 'DS', 'AIDS'],\n",
       "               [181, 191, 'DS', 'infections'],\n",
       "               [197, 202, 'OG', 'HIV-1']]],\n",
       "             'TABLE 3Genetic damage in hypermutated sequences SourceSequence typenNo. of in-frame stop codonsMean (range) net charge of protein Patient PBMCNormal14932.5\\u2009(0.0–9.0) Hypermutated706912.6\\u2009(3.0–23.0) Virus culturesNormal7331.9\\u2009(0.0–4.0) Hypermutated584810.9\\u2009(2.0–23.0)': [[[198,\n",
       "                203,\n",
       "                'OG',\n",
       "                'Virus']]],\n",
       "             'Relationship of hypermutation to T-cell activation in virus cultures.': [[[54,\n",
       "                59,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'CD4+ T cells isolated from the PBMC of normal donors were infected with an NL4-3 virus stock for 2 h at a multiplicity of infection of 0.02 (stippled gray bar) under varying conditions of PHA activation (gray bars).': [[[0,\n",
       "                3,\n",
       "                'GP',\n",
       "                'CD4'],\n",
       "               [75, 80, 'OG', 'NL4-3'],\n",
       "               [81, 86, 'OG', 'virus'],\n",
       "               [122, 131, 'DS', 'infection']]],\n",
       "             'Condition I, infection without addition of PHA; condition II, simultaneous infection and PHA addition; conditions III to VIII, infection with prior PHA activation for 3 to 72 h respectively; condition IX, infection and addition of PHA 24 h later.': [[[13,\n",
       "                22,\n",
       "                'DS',\n",
       "                'infection'],\n",
       "               [75, 84, 'DS', 'infection'],\n",
       "               [127, 136, 'DS', 'infection'],\n",
       "               [148, 151, 'GP', 'PHA'],\n",
       "               [205, 214, 'DS', 'infection']]],\n",
       "             'Cultures were sampled at intervals and assayed for the presence of p24 antigen in the culture supernatant (solid lines) and for the presence of hypermutated HIV-1 DNA sequences.': [[[67,\n",
       "                70,\n",
       "                'GP',\n",
       "                'p24'],\n",
       "               [157, 162, 'OG', 'HIV-1']]],\n",
       "             'Samples from time points at which hypermutation was rare (f, i, j, k, and l) are shown both as bulk PCR products (B) and after pre-enrichment by ScrF1 and AvaII digestion (A).': [[[145,\n",
       "                150,\n",
       "                'GP',\n",
       "                'ScrF1'],\n",
       "               [155, 160, 'GP', 'AvaII']]],\n",
       "             'Parameters of hypermutation in virus cultures.': [[[31,\n",
       "                36,\n",
       "                'OG',\n",
       "                'virus']]],\n",
       "             'A total of 227 HIV-1 protease sequences, 131 of which were unique, were obtained from virus cultures.': [[[15,\n",
       "                20,\n",
       "                'OG',\n",
       "                'HIV-1'],\n",
       "               [21, 29, 'GP', 'protease'],\n",
       "               [86, 91, 'OG', 'virus']]],\n",
       "             '(C) Base composition of normal and hypermutated sequences from virus cultures.': [[[63,\n",
       "                68,\n",
       "                'OG',\n",
       "                'virus']]]})"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ml_tagged_sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "interested_sentences=[]\n",
    "for each_sentence, ml_tags in final_sentences.items():\n",
    "    minidict = {}\n",
    "    minidict['text'] = each_sentence\n",
    "    if section_tags[each_sentence]:\n",
    "        minidict['section'] = section_tags[each_sentence]\n",
    "    all_matches = mactch_gp_ds_cd[each_sentence]\n",
    "    if all_matches:\n",
    "        minidict['matches'] = all_matches\n",
    "    all_co_occurances = co_occurance_gp_ds[each_sentence] + co_occurance_gp_cd[each_sentence]+co_occurance_ds_cd[each_sentence]\n",
    "    if all_co_occurances:\n",
    "        minidict['co-occurrence'] = all_co_occurances\n",
    "    if all_co_occurances or all_matches:\n",
    "        interested_sentences.append(minidict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "json_generated = {}\n",
    "try:\n",
    "    json_generated['pmid'] = 'PMC'+soup.find(attrs={\"pub-id-type\" : \"pmcid\"}).text #\"article-id\"\n",
    "except:\n",
    "    json_generated['pmid'] = 'PMC'\n",
    "\n",
    "json_generated['OG_set'] = list(og_set)\n",
    "json_generated['sentences'] = interested_sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"pmid\": \"PMC115041\",\n",
      "    \"OG_set\": [\n",
      "        \"HIV-1 virus\",\n",
      "        \"HIV 1\",\n",
      "        \"Taq\",\n",
      "        \"Virus\",\n",
      "        \"caprine arthritis-encephalitis virus\",\n",
      "        \"RNA viruses\",\n",
      "        \"virus\",\n",
      "        \"measles virus\",\n",
      "        \"HIV\",\n",
      "        \"retrovirus\",\n",
      "        \"Human Immunodeficiency Virus Type 1\",\n",
      "        \"EIAV\",\n",
      "        \"CAEV\",\n",
      "        \"simian immunodeficiency virus\",\n",
      "        \"tobacco mosaic virus\",\n",
      "        \"HIV-1 subtype B\",\n",
      "        \"vesicular stomatitis virus\",\n",
      "        \"hepatitis B virus\",\n",
      "        \"SNV\",\n",
      "        \"SIV\",\n",
      "        \"Escherichia coli\",\n",
      "        \"proviruses\",\n",
      "        \"retroviruses\",\n",
      "        \"equine infectious anemia virus\",\n",
      "        \"NL4-3\",\n",
      "        \"human immunodeficiency virus type 1\",\n",
      "        \"avian leukosis virus\",\n",
      "        \"spleen necrosis virus\",\n",
      "        \"lentivirus\",\n",
      "        \"lentiviruses\",\n",
      "        \"HIV-2\",\n",
      "        \"NL4-3 virus\",\n",
      "        \"viruses\",\n",
      "        \"HIV-1\"\n",
      "    ],\n",
      "    \"sentences\": [\n",
      "        {\n",
      "            \"text\": \"Human Immunodeficiency Virus Type 1 DNA Sequences Genetically Damaged by Hypermutation Are Often Abundant in Patient Peripheral Blood Mononuclear Cells and May Be Generated during Near-Simultaneous Infection and Activation of CD4+ T Cells\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"Infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 198,\n",
      "                    \"endInSentence\": 207\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 226,\n",
      "                    \"endInSentence\": 229\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 226,\n",
      "                    \"end1\": 229,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 198,\n",
      "                    \"end2\": 207,\n",
      "                    \"label2\": \"Infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 1.0,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.59,\n",
      "                        \"Negative\": 0.0,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.15,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    },\n",
      "                    \"relation\": [\n",
      "                        {\n",
      "                            \"startr\": 212,\n",
      "                            \"endr\": 222,\n",
      "                            \"labelr\": \"Activation\",\n",
      "                            \"typer\": \"POSITIVE_REGULATION\"\n",
      "                        }\n",
      "                    ]\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"PCR primers matched to normal and hypermutated sequences were used in conjunction with an agarose gel electrophoresis system incorporating an AT-binding dye to visualize, separate, clone, and sequence hypermutated and normal sequences in the 297-bp HIV-1 protease gene amplified from patient PBMC.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 255,\n",
      "                    \"endInSentence\": 263\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Among 53 patients, including individuals infected with subtypes A through D and at different clinical stages, at least 43% of patients harbored abundant hypermutated, along with normal, protease genes.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 186,\n",
      "                    \"endInSentence\": 194\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In 70 hypermutated sequences, saturation of G residues in the GA or GG dinucleotide context ranged from 20 to 94%.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 71,\n",
      "                    \"endInSentence\": 83\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Levels of other mutants were not elevated, and G-to-A replacement was entirely restricted to GA or GG, and not GC or GT, dinucleotides.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 121,\n",
      "                    \"endInSentence\": 134\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"To investigate the conditions under which hypermutation occurs in cell cultures, purified CD4+ T cells from normal donors were infected with cloned NL4-3 virus stocks at various times before and after phytohemagglutinin (PHA) activation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 90,\n",
      "                    \"endInSentence\": 93\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"phytohemagglutinin\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 201,\n",
      "                    \"endInSentence\": 219\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 90,\n",
      "                    \"end1\": 93,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 201,\n",
      "                    \"end2\": 219,\n",
      "                    \"label2\": \"phytohemagglutinin\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutation was pronounced when HIV-1 infection occurred simultaneously with, or a few hours after, PHA activation, but after 12 h or more after PHA activation, most HIV-1 sequences were normal.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 34,\n",
      "                    \"endInSentence\": 49\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Near-simultaneous activation and infection of CD4+ T cells may represent a window of susceptibility where the informational content of HIV-1 sequences is lost due to hypermutation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 33,\n",
      "                    \"endInSentence\": 42\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 46,\n",
      "                    \"endInSentence\": 49\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 46,\n",
      "                    \"end1\": 49,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 33,\n",
      "                    \"end2\": 42,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.98,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.85,\n",
      "                        \"Negative\": 0.02,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.07,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.06,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"There is loss of the information content of nucleotide sequences once mutation rates are increased beyond a tolerable error threshold (15).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 44,\n",
      "                    \"endInSentence\": 54\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"How surprising, then, that for some of the same viruses with the highest mutation rates, another mutation process has been described in which remarkable levels of one specific type of nucleotide substitution are observed.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 184,\n",
      "                    \"endInSentence\": 194\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"G-to-A hypermutation occurs specifically within the GpA or GpG dinucleotide context (18, 60).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"GpA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 52,\n",
      "                    \"endInSentence\": 55\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 63,\n",
      "                    \"endInSentence\": 75\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 52,\n",
      "                    \"end1\": 55,\n",
      "                    \"label1\": \"GpA\",\n",
      "                    \"start2\": 63,\n",
      "                    \"end2\": 75,\n",
      "                    \"label2\": \"dinucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The susceptibility of lentiviruses to hypermutation is thought to be a property of their reverse transcriptase (RT), which, compared to nonlentiviral RTs, has a great capacity to elongate beyond nucleotide mismatches (51) and an increased ability to hypermutate in the presence of unbalanced nucleotide pools in vitro (43).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"reverse transcriptase\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 89,\n",
      "                    \"endInSentence\": 110\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"RT\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 112,\n",
      "                    \"endInSentence\": 114\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"RTs\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 150,\n",
      "                    \"endInSentence\": 153\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 195,\n",
      "                    \"endInSentence\": 205\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 292,\n",
      "                    \"endInSentence\": 302\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 89,\n",
      "                    \"end1\": 110,\n",
      "                    \"label1\": \"reverse transcriptase\",\n",
      "                    \"start2\": 195,\n",
      "                    \"end2\": 205,\n",
      "                    \"label2\": \"nucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 89,\n",
      "                    \"end1\": 110,\n",
      "                    \"label1\": \"reverse transcriptase\",\n",
      "                    \"start2\": 292,\n",
      "                    \"end2\": 302,\n",
      "                    \"label2\": \"nucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 112,\n",
      "                    \"end1\": 114,\n",
      "                    \"label1\": \"RT\",\n",
      "                    \"start2\": 195,\n",
      "                    \"end2\": 205,\n",
      "                    \"label2\": \"nucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 112,\n",
      "                    \"end1\": 114,\n",
      "                    \"label1\": \"RT\",\n",
      "                    \"start2\": 292,\n",
      "                    \"end2\": 302,\n",
      "                    \"label2\": \"nucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 150,\n",
      "                    \"end1\": 153,\n",
      "                    \"label1\": \"RTs\",\n",
      "                    \"start2\": 195,\n",
      "                    \"end2\": 205,\n",
      "                    \"label2\": \"nucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 150,\n",
      "                    \"end1\": 153,\n",
      "                    \"label1\": \"RTs\",\n",
      "                    \"start2\": 292,\n",
      "                    \"end2\": 302,\n",
      "                    \"label2\": \"nucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"With respect to genome region, hypermutation was initially found in small subregions of gag, env, nef, and U3/R elements of the long terminal repeat (LTR) (11, 21, 60), but others have found hypermutation throughout the genome (4).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"gag\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 88,\n",
      "                    \"endInSentence\": 91\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"env\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 93,\n",
      "                    \"endInSentence\": 96\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"nef\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 98,\n",
      "                    \"endInSentence\": 101\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"First thought to be the work of a mutant RT (18, 50), hypermutation has now been shown to be a property of wild-type RT (42) operating under suboptimal conditions.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"RT\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 41,\n",
      "                    \"endInSentence\": 43\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"RT\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 117,\n",
      "                    \"endInSentence\": 119\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Attempts have been made to generate hypermutants in a cell-free system in vitro with RNA, purified RT, and strongly biased deoxynucleoside triphosphate (dNTP) pools (43, 44, 59, 61).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"RT\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 99,\n",
      "                    \"endInSentence\": 101\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"While G-to-A hypermutants were produced, they lacked the marked preference for the GpA and GpG dinucleotide contexts that typify hypermutants produced in vivo or during virus cultivation in vitro.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"GpA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 83,\n",
      "                    \"endInSentence\": 86\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 95,\n",
      "                    \"endInSentence\": 107\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 83,\n",
      "                    \"end1\": 86,\n",
      "                    \"label1\": \"GpA\",\n",
      "                    \"start2\": 95,\n",
      "                    \"end2\": 107,\n",
      "                    \"label2\": \"dinucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"An intermediate approach where RT in virus particles was allowed to complete first-strand DNA synthesis in the presence of exogenously provided, highly biased dNTP pools also resulted in loss of the GpA and GpG dinucleotide preference (62).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"RT\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 31,\n",
      "                    \"endInSentence\": 33\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"GpA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 199,\n",
      "                    \"endInSentence\": 202\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 211,\n",
      "                    \"endInSentence\": 223\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 31,\n",
      "                    \"end1\": 33,\n",
      "                    \"label1\": \"RT\",\n",
      "                    \"start2\": 211,\n",
      "                    \"end2\": 223,\n",
      "                    \"label2\": \"dinucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 199,\n",
      "                    \"end1\": 202,\n",
      "                    \"label1\": \"GpA\",\n",
      "                    \"start2\": 211,\n",
      "                    \"end2\": 223,\n",
      "                    \"label2\": \"dinucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutants were also recovered at low frequency (0.5%) from cell cultures after the addition of deoxythymidine to supernatants, which resulted in an elevated intracellular [dTTP]/[dCTP] ratio (62), but because all possible dinucleotide contexts were not represented in the sequence that was studied, a complete analysis of context preference was not possible.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"deoxythymidine\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 98,\n",
      "                    \"endInSentence\": 112\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 225,\n",
      "                    \"endInSentence\": 237\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"More subtle perturbations in the intracellular environment may trigger dNTP pool biases sufficient to generate hypermutation while maintaining the proper dinucleotide context (67).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 154,\n",
      "                    \"endInSentence\": 166\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Indeed, a 40- fold [dTTP]/[dCTP] ratio produced hypermutants in which the GpA context preference was maintained (62), and hypermutants were recovered from 1 to 2% of unstimulated and phytohemagglutinin (PHA)-stimulated peripheral blood mononuclear cells (PBMC) without manipulation of dNTP pools (62).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"GpA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 74,\n",
      "                    \"endInSentence\": 77\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"phytohemagglutinin\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 183,\n",
      "                    \"endInSentence\": 201\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 74,\n",
      "                    \"end1\": 77,\n",
      "                    \"label1\": \"GpA\",\n",
      "                    \"start2\": 183,\n",
      "                    \"end2\": 201,\n",
      "                    \"label2\": \"phytohemagglutinin\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Imbalanced and fluctuating nucleotide pools are a key element of many types of mutation, including hypermutation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 27,\n",
      "                    \"endInSentence\": 37\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Since the composition of dNTP pools changes as cells progress through the cell cycle (5, 20, 45), the timing of HIV-1 infection with respect to T-cell activation and entry into the cell cycle could be an important factor determining the generation of hypermutants.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 112,\n",
      "                    \"endInSentence\": 127\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"For example, a method based on the blue/white \\u03b2-galactosidase complementation assay yielded a low recovery of hypermutants (0.5 to 2%) (62), not higher than that obtained when clones were randomly picked and sequenced.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"blue/white \\u03b2-galactosidase\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 35,\n",
      "                    \"endInSentence\": 61\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Here we describe the design and application of powerful new methods for systematic detection and recovery of hypermutants and their application to clinical samples and to the products of HIV-1 infection in cell culture.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 187,\n",
      "                    \"endInSentence\": 202\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Twenty-six samples were from patients hospitalized in Tanzania in 1996 with symptoms compatible with AIDS (25).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"AIDS\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 101,\n",
      "                    \"endInSentence\": 105\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The latter two groups were in the early, asymptomatic stage of HIV-1 infection, and the sample was typically drawn within 6 months of HIV-1 seroconversion.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 63,\n",
      "                    \"endInSentence\": 78\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"DNA extracted from patient PBMC or virus cultures was the template for PCR amplification of HIV-1 protease sequences.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 98,\n",
      "                    \"endInSentence\": 106\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The primers for amplification of hypermutants contained either mixed bases (hyp primers) or G-to-A replacements (hypa primers) at some of the sites (GpA or GpG) susceptible to hypermutation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"hyp\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 76,\n",
      "                    \"endInSentence\": 79\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"hypa\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 113,\n",
      "                    \"endInSentence\": 117\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"GpA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 149,\n",
      "                    \"endInSentence\": 152\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"GpG\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 156,\n",
      "                    \"endInSentence\": 159\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Each PCR amplification was run in duplicate, once with normal primers and a second time with an equal mixture of hyp and hypa primers.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"hyp\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 113,\n",
      "                    \"endInSentence\": 116\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"hypa\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 121,\n",
      "                    \"endInSentence\": 125\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"PCR products were analyzed in parallel 1% agarose gels, with or without HA yellow, a compound which consists of the DNA ligand bisbenzamide covalently linked to polyethylene glycol (PEG) (Hanse Analytik, Bremen, Germany).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"bisbenzamide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 127,\n",
      "                    \"endInSentence\": 139\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Bisbenzamide binds preferentially to AT-rich regions in DNA (in order from highest to lowest preference, AATT and AAAA, TAAT, ATAT, TATA, and TTAA) (1) and, when coupled to PEG, retards DNA mobility during gel electrophoresis according to AT content (41, 48, 65).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"Bisbenzamide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 0,\n",
      "                    \"endInSentence\": 12\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Products were visualized by poststaining with ethidium bromide.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"ethidium bromide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 46,\n",
      "                    \"endInSentence\": 62\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In experiments with virus cultivation in cell cultures, enrichment for hypermutated sequences by predigestion of the cellular DNA with restriction endonucleases ScrF1(CCAG\\u2217G\\u2217A) and AvaII (G\\u2217G\\u2217ACC), which preferentially cleave normal, as opposed to hypermutated, HIV-1 protease sequences, was used to recover rare hypermutated sequences from some samples.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"restriction endonucleases\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 135,\n",
      "                    \"endInSentence\": 160\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"ScrF1\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 161,\n",
      "                    \"endInSentence\": 166\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"AvaII\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 181,\n",
      "                    \"endInSentence\": 186\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 268,\n",
      "                    \"endInSentence\": 276\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"DNA was codigested with ScrF1 and AvaII at 37\\u00b0C overnight prior to PCR amplification of HIV-1 protease.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"ScrF1\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 24,\n",
      "                    \"endInSentence\": 29\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"AvaII\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 34,\n",
      "                    \"endInSentence\": 39\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 94,\n",
      "                    \"endInSentence\": 102\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Eight hundred nanograms of purified plasmids was digested using 20 U of EcoR1 (New England Biolabs, Beverly, Mass.) for 1 h at 37\\u00b0C to release the cloned protease gene, which was then rescreened on agarose gels with HA yellow.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"EcoR1\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 72,\n",
      "                    \"endInSentence\": 77\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 154,\n",
      "                    \"endInSentence\": 162\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The genetic subtypes of normal protease sequences from the 53 patients were established by phylogenetic analysis (see Table 2).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 31,\n",
      "                    \"endInSentence\": 39\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Among the AIDS patients more than half harbored subtype C; the remainder harbored mostly subtype A or D.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"AIDS\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 10,\n",
      "                    \"endInSentence\": 14\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Virus stocks generated from an infectious molecular clone of HIV-1 subtype B isolate NL4-3 (56, 57) were used to infect PBMC from a seronegative donor.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"NL4-3\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 85,\n",
      "                    \"endInSentence\": 90\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Cells were activated with 1 \\u03bcg of PHA/ml for 72 h, infected for 24 h, washed twice, and maintained in RPMI 1640 with 15% fetal calf serum (FCS), 1% PenStrep (Quality Biological, Gaithersburg, Md.), 1% l-glutamine, and 20 U of interleukin-2 (IL-2)/ml.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"l-glutamine\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 201,\n",
      "                    \"endInSentence\": 212\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"interleukin-2\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 226,\n",
      "                    \"endInSentence\": 239\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"IL-2\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 241,\n",
      "                    \"endInSentence\": 245\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 226,\n",
      "                    \"end1\": 239,\n",
      "                    \"label1\": \"interleukin-2\",\n",
      "                    \"start2\": 201,\n",
      "                    \"end2\": 212,\n",
      "                    \"label2\": \"l-glutamine\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Culture supernatants were assayed for p24 antigen (HIV-1 p24 Antigen Assay Kit; Coulter Corporation, Miami, Fla.) at 3-day intervals.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 38,\n",
      "                    \"endInSentence\": 41\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 57,\n",
      "                    \"endInSentence\": 60\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Culture supernatants collected near the peak of p24 antigen production were treated with 50 U of DNAse I/ml (Boehringer Mannheim) for 30 min at room temperature and filtered sterilized (pore size, 0.22 \\u03bcm).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 48,\n",
      "                    \"endInSentence\": 51\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"DNAse I\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 97,\n",
      "                    \"endInSentence\": 104\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"CD4+ T cells were purified by negative selection on magnetic beads with the MACS CD4+ T Cell Isolation Kit (Miltenyi Biotec, Auburn, Calif.).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 0,\n",
      "                    \"endInSentence\": 3\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 81,\n",
      "                    \"endInSentence\": 84\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Prior to infection, an aliquot of the purified CD4+ T cells was fixed with 2% formaldehyde and stained with fluorescein isothiocyanate (FITC)- or phycoerythrin (PE)-conjugated monoclonal antibodies specific for cell surface molecules CD3, CD4, CD8, CD14, CD20, CD56, and HLA-DR (Becton Dickinson, San Jose, Calif.).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 9,\n",
      "                    \"endInSentence\": 18\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 47,\n",
      "                    \"endInSentence\": 50\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"phycoerythrin\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 146,\n",
      "                    \"endInSentence\": 159\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD3\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 234,\n",
      "                    \"endInSentence\": 237\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 239,\n",
      "                    \"endInSentence\": 242\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD8\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 244,\n",
      "                    \"endInSentence\": 247\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD14\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 249,\n",
      "                    \"endInSentence\": 253\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD20\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 255,\n",
      "                    \"endInSentence\": 259\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD56\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 261,\n",
      "                    \"endInSentence\": 265\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"HLA-DR\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 271,\n",
      "                    \"endInSentence\": 277\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 47,\n",
      "                    \"end1\": 50,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 9,\n",
      "                    \"end2\": 18,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 1.0,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.86,\n",
      "                        \"Negative\": 0.01,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.13,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 47,\n",
      "                    \"end1\": 50,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 146,\n",
      "                    \"end2\": 159,\n",
      "                    \"label2\": \"phycoerythrin\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 9,\n",
      "                    \"end1\": 18,\n",
      "                    \"label1\": \"infection\",\n",
      "                    \"start2\": 146,\n",
      "                    \"end2\": 159,\n",
      "                    \"label2\": \"phycoerythrin\",\n",
      "                    \"type\": \"DS-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The majority of the cells in each experiment were CD4+ T cells (mean, 81%; range, 69 to 99%) representing most of the CD3+ T-cell subset (mean, 83%; range, 76 to 100%).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 50,\n",
      "                    \"endInSentence\": 53\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD3\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 118,\n",
      "                    \"endInSentence\": 121\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The other major cell type was B cells (CD20+; mean, 10.8%; range, 4.5 to 14.9%).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD20\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 39,\n",
      "                    \"endInSentence\": 43\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Monocytes (CD14+), CD8+ T cells (CD8+), and NK cells (CD56+) were, on average, 2 to 4% of the population.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD14\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 11,\n",
      "                    \"endInSentence\": 15\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD8\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 19,\n",
      "                    \"endInSentence\": 22\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD8\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 33,\n",
      "                    \"endInSentence\": 36\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD56\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 54,\n",
      "                    \"endInSentence\": 58\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Between 14.6 and 17.7% of the cells were HLA-DR positive.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HLA-DR\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 41,\n",
      "                    \"endInSentence\": 47\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Cells were either left unstimulated or stimulated with 1 \\u03bcg of PHA (Pharmacia)/ml for various intervals before, during, or after virus infection at a multiplicity of infection of 0.02 for 2 h at 37\\u00b0C.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 129,\n",
      "                    \"endInSentence\": 144\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 166,\n",
      "                    \"endInSentence\": 175\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Cells were washed twice and maintained in RPMI 1640 containing 10% autologous plasma, and IL-2 at 20 U/ml (Amersham).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"IL-2\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 90,\n",
      "                    \"endInSentence\": 94\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Autologous plasma, necessary to prevent activation of the unstimulated CD4+ T cells by foreign antigens, was used throughout.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 71,\n",
      "                    \"endInSentence\": 74\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Supernatants from cell cultures were evaluated for p24 antigen production using the Coulter HIV-1 p24 Antigen Assay Kit.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 51,\n",
      "                    \"endInSentence\": 54\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 98,\n",
      "                    \"endInSentence\": 101\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Nine different culture conditions were established (see Fig. 6) as follows: Culture I, virus infection without PHA stimulation; culture II, PHA and virus added simultaneously for 2 h; cultures III through VIII, PHA stimulation preceding virus infection by 3, 6, 12, 24, 48, or 72 h, respectively; culture IX, addition of virus for 2 h, washing away of virus, and PHA addition 24 h later.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 87,\n",
      "                    \"endInSentence\": 102\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"PHA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 111,\n",
      "                    \"endInSentence\": 114\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"PHA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 211,\n",
      "                    \"endInSentence\": 214\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 237,\n",
      "                    \"endInSentence\": 252\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 111,\n",
      "                    \"end1\": 114,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 87,\n",
      "                    \"end2\": 102,\n",
      "                    \"label2\": \"virus infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.01,\n",
      "                        \"Positive\": 0.21,\n",
      "                        \"Negative\": 0.02,\n",
      "                        \"Neutral\": 0.02,\n",
      "                        \"Altered Expression\": 0.1,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 111,\n",
      "                    \"end1\": 114,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 237,\n",
      "                    \"end2\": 252,\n",
      "                    \"label2\": \"virus infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.01,\n",
      "                        \"Positive\": 0.21,\n",
      "                        \"Negative\": 0.02,\n",
      "                        \"Neutral\": 0.02,\n",
      "                        \"Altered Expression\": 0.1,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 211,\n",
      "                    \"end1\": 214,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 237,\n",
      "                    \"end2\": 252,\n",
      "                    \"label2\": \"virus infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.01,\n",
      "                        \"Positive\": 0.21,\n",
      "                        \"Negative\": 0.02,\n",
      "                        \"Neutral\": 0.02,\n",
      "                        \"Altered Expression\": 0.1,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Cells and culture supernatants were collected at 6, 12, 24, 48, and 72 h after infection, with the exception of culture condition IX, where collection took place at 6, 12, 24, 48, and 72 h after addition of PHA.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 79,\n",
      "                    \"endInSentence\": 88\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"This program identifies mutations and their dinucleotide context with respect to a reference sequence that is provided with each alignment.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 44,\n",
      "                    \"endInSentence\": 56\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"For virus cultures, the sequence of NL4-3, obtained by RT-PCR and sequencing of the virus stock, was used as the reference.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"NL4-3\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 36,\n",
      "                    \"endInSentence\": 41\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Nucleotide sequence accession numbers.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"Nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 0,\n",
      "                    \"endInSentence\": 10\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Nested primers which amplify a 297-bp segment encoding HIV-1 protease were examined for GA and GG dinucleotides, which are susceptible to hypermutation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 61,\n",
      "                    \"endInSentence\": 69\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"GG\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 95,\n",
      "                    \"endInSentence\": 97\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 98,\n",
      "                    \"endInSentence\": 111\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 61,\n",
      "                    \"end1\": 69,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 95,\n",
      "                    \"end2\": 97,\n",
      "                    \"label2\": \"GG\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 61,\n",
      "                    \"end1\": 69,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 98,\n",
      "                    \"end2\": 111,\n",
      "                    \"label2\": \"dinucleotides\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"New primers were designed with either incorporated GA or CT mixtures at some of these sites (hyp primers) or replaced G with A or C with T (hypa primers).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"hyp\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 93,\n",
      "                    \"endInSentence\": 96\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"hypa\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 140,\n",
      "                    \"endInSentence\": 144\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The hyp and hypa primers were also shortened to bring potential mispairs closer to the 3\\u2032 end of the primer, where they could exert the maximal effect in destabilizing interaction with normal sequences.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"hyp\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 4,\n",
      "                    \"endInSentence\": 7\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"hypa\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 12,\n",
      "                    \"endInSentence\": 16\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The dye HA yellow, in which consists of the DNA-binding ligand bisbenzamide coupled to PEG, was incorporated into 1% agarose gels.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HA yellow\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 8,\n",
      "                    \"endInSentence\": 17\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"A series of cloned protease sequences representing different A+T contents, amplified and sequenced previously in the pilot phase of this project, were used for evaluation of the gel system (Fig. 1).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 19,\n",
      "                    \"endInSentence\": 27\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The appearance of protease genes PCR amplified from patient PBMC is illustrated in Fig. 3.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 18,\n",
      "                    \"endInSentence\": 26\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"While all of the protease genes were of uniform length without HA yellow, they often separated into multiple bands of different A+T content on HA yellow gels.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 17,\n",
      "                    \"endInSentence\": 25\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutated sequences were found in both asymptomatic and AIDS patients, and in patients with infections with four different HIV-1 subtypes.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"AIDS\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 59,\n",
      "                    \"endInSentence\": 63\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infections\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 95,\n",
      "                    \"endInSentence\": 105\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The dinucleotide context of hypermutation was then considered (Fig. 5, bottom).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 4,\n",
      "                    \"endInSentence\": 16\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"With respect to the saturation of the available GA and GG dinucleotides, at all levels of G-to-A replacement, the percentage of G replaced by A in the dinucleotide context GA was about twofold higher than the percentage of G replaced by A in the GG context.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 58,\n",
      "                    \"endInSentence\": 71\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 151,\n",
      "                    \"endInSentence\": 163\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Although G-to-A substitutions in the GC and GT dinucleotide contexts were previously referred as hypermutation related (60), the observed number of these replacements was limited, ranging from 0 to 3 (mean, 0.6) per hypermutated sequence.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 47,\n",
      "                    \"endInSentence\": 59\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"At the highest levels of G-to-A mutation, virtually all of the GA and GG dinucleotides were mutated, while G-to-A mutations in GC or GT (other mutations; Fig. 5) were not elevated.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 73,\n",
      "                    \"endInSentence\": 86\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In summary, hypermutated HIV-1 protease sequences recovered from patient PBMC exhibited a broad range of G-to-A mutation that was entirely restricted to the GA or GG dinucleotide context, with a consistent preference for G-to-A mutation in the dinucleotide GA relative to GG at all levels of G-to-A saturation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 31,\n",
      "                    \"endInSentence\": 39\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 166,\n",
      "                    \"endInSentence\": 178\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 244,\n",
      "                    \"endInSentence\": 256\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 31,\n",
      "                    \"end1\": 39,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 166,\n",
      "                    \"end2\": 178,\n",
      "                    \"label2\": \"dinucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 31,\n",
      "                    \"end1\": 39,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 244,\n",
      "                    \"end2\": 256,\n",
      "                    \"label2\": \"dinucleotide\",\n",
      "                    \"type\": \"GP-CD\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In addition, there was a net accumulation of positively charged amino acids, mostly from decreased Gly, Asp, and Arg and increased Lys and Asn in hypermutants.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"amino acids\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 64,\n",
      "                    \"endInSentence\": 75\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The proteases encoded by hypermutated sequences had, on average, a fivefold increase in net charge (Table 3).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"proteases\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 4,\n",
      "                    \"endInSentence\": 13\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"It is noteworthy that hypermutated sequences with the lowest levels of G-to-A substitutions sustained inactivating in-frame stop codons and many nonsynonymous nucleotide substitutions.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"nucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 159,\n",
      "                    \"endInSentence\": 169\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"These combined effects make it highly unlikely that functional proteases, and, consequently, viable viruses can be derived from hypermutated sequences.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"proteases\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 63,\n",
      "                    \"endInSentence\": 72\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutated sequences were recovered at essentially the same rate from AIDS patients (42%) and early asymptomatic individuals (44%).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"AIDS\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 72,\n",
      "                    \"endInSentence\": 76\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutation was found in 57% of subtype A, 67% of subtype D, and 44% of subtype B infections, as well as in A+D dually infected and unclassified samples.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infections\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 84,\n",
      "                    \"endInSentence\": 94\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Although subtype C was the most frequently represented subtype among AIDS patients, hypermutation was detected only in 3 of 14 (21%) patients infected with this subtype.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"AIDS\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 69,\n",
      "                    \"endInSentence\": 73\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Experiments with cell-free systems have shown that unbalances in dNTP pools during reverse transcription can lead to hypermutation, but the specificity for the GA and GG dinucleotide context has been difficult to reproduce.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 170,\n",
      "                    \"endInSentence\": 182\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"On the other hand, hypermutants in the correct dinucleotide context have been recovered from HIV-1 cultures in PBMC, but the exact nature of the cell population that produced them and the status of the dNTP pools in this population have not been elucidated.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 47,\n",
      "                    \"endInSentence\": 59\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In CD4+ T cells, which are a main target of HIV 1 replication in vivo, dNTP pools are elevated when the cells are activated and in cycle (20) but are unbalanced and fluctuating in resting T cells and during the transition from a resting to an activated state (5, 10, 20, 58).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 3,\n",
      "                    \"endInSentence\": 6\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"We designed an experiment in which the timing of HIV-1 infection was varied with respect to activation of resting CD4+ T cells by PHA; the outcome measure was recovery of hypermutated sequences whose characteristics corresponded exactly to those found in vivo.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 49,\n",
      "                    \"endInSentence\": 64\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 114,\n",
      "                    \"endInSentence\": 117\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 114,\n",
      "                    \"end1\": 117,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 49,\n",
      "                    \"end2\": 64,\n",
      "                    \"label2\": \"HIV-1 infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.21,\n",
      "                        \"Negative\": 0.01,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.09,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.02,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    },\n",
      "                    \"relation\": [\n",
      "                        {\n",
      "                            \"startr\": 92,\n",
      "                            \"endr\": 102,\n",
      "                            \"labelr\": \"activation\",\n",
      "                            \"typer\": \"POSITIVE_REGULATION\"\n",
      "                        }\n",
      "                    ]\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"p24 antigen in the culture supernatant, indicative of virus production, was observed only when cultures were stimulated with PHA either before, or simultaneously with, virus infection.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 0,\n",
      "                    \"endInSentence\": 3\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 168,\n",
      "                    \"endInSentence\": 183\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 0,\n",
      "                    \"end1\": 3,\n",
      "                    \"label1\": \"p24\",\n",
      "                    \"start2\": 168,\n",
      "                    \"end2\": 183,\n",
      "                    \"label2\": \"virus infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.98,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.03,\n",
      "                        \"Negative\": 0.02,\n",
      "                        \"Neutral\": 0.04,\n",
      "                        \"Altered Expression\": 0.01,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.05,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The HIV-1 protease gene was amplified from all time points and culture conditions (Materials and Methods) and evaluated for the presence of normal and hypermutated sequences using the HA yellow gel system (Fig. 6).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 10,\n",
      "                    \"endInSentence\": 18\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The richest source of hypermutants was culture II, in which infection and activation occurred simultaneously (Fig. 6A, lanes b through e).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 60,\n",
      "                    \"endInSentence\": 69\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutated sequences could also be visualized in conditions III and IV (lanes g and h), which represented PHA activation for less than 8 h before infection.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"PHA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 108,\n",
      "                    \"endInSentence\": 111\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 148,\n",
      "                    \"endInSentence\": 157\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 108,\n",
      "                    \"end1\": 111,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 148,\n",
      "                    \"end2\": 157,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.98,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.01,\n",
      "                        \"Negative\": 0.01,\n",
      "                        \"Neutral\": 0.56,\n",
      "                        \"Altered Expression\": 0.06,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.01,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"However, by pre-enrichment for hypermutants using ScrF1 and AvaII digestion (see Materials and Methods), rare hypermutants could be observed in late time points in most cultures (Fig. 6, lanes f\\u2217, i\\u2217, j\\u2217, k\\u2217, and l\\u2217).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"ScrF1\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 50,\n",
      "                    \"endInSentence\": 55\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"AvaII\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 60,\n",
      "                    \"endInSentence\": 65\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Finally, hypermutated sequences were found in condition IX, where virus infection preceded PHA activation by 24 h.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 66,\n",
      "                    \"endInSentence\": 81\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"PHA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 91,\n",
      "                    \"endInSentence\": 94\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 91,\n",
      "                    \"end1\": 94,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 66,\n",
      "                    \"end2\": 81,\n",
      "                    \"label2\": \"virus infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 1.0,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.0,\n",
      "                        \"Negative\": 0.0,\n",
      "                        \"Neutral\": 0.87,\n",
      "                        \"Altered Expression\": 0.01,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Overall, these analyses establish that HIV-1 infection of resting CD4+ T cells either shortly before or shortly after activation generated abundant hypermutated sequences with high levels of G-to-A mutation, while infection after longer periods of activation yielded principally normal sequences.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 39,\n",
      "                    \"endInSentence\": 54\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 66,\n",
      "                    \"endInSentence\": 69\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 214,\n",
      "                    \"endInSentence\": 223\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 66,\n",
      "                    \"end1\": 69,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 39,\n",
      "                    \"end2\": 54,\n",
      "                    \"label2\": \"HIV-1 infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.55,\n",
      "                        \"Negative\": 0.03,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.02,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.13,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 66,\n",
      "                    \"end1\": 69,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 214,\n",
      "                    \"end2\": 223,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.55,\n",
      "                        \"Negative\": 0.03,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.02,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.13,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"These combined approaches provided 131 different HIV-1 protease sequences from virus cultures for analysis.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 55,\n",
      "                    \"endInSentence\": 63\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"However, when sequences from each culture condition were analyzed separately, an interesting pattern of dinucleotide context utilization was noted (Fig. 7B, lower graph).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 104,\n",
      "                    \"endInSentence\": 116\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In cultures IX, I, and II, where PHA either was not added or was added simultaneously with or after infection with the virus, the preference for GA over GG was not strong.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 100,\n",
      "                    \"endInSentence\": 109\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"With prior PHA stimulation for 3 h or more (cultures III to VIII), the saturation of available GA dinucleotides occurred at 4 to 6 times that of available GG dinucleotides.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"GA dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 95,\n",
      "                    \"endInSentence\": 111\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 158,\n",
      "                    \"endInSentence\": 171\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Finally, hypermutated and normal protease gene sequences recovered from cell culture were examined for the loss of informational content (Table 3).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 33,\n",
      "                    \"endInSentence\": 41\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutated sequences could be detected in 43% of patients and in virus cultures when T-cell stimulation was initiated at the time of, or shortly before, virus infection.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 155,\n",
      "                    \"endInSentence\": 170\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"We examined the question of whether hypermutation affects GC and GT dinucleotides, although at lower levels, in addition to the preferred GA and GG contexts.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 68,\n",
      "                    \"endInSentence\": 81\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"However, our observations are limited to the HIV-1 protease gene, and examination of full HIV-1 genomes for the context of hypermutation would be required to fortify this conclusion.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 51,\n",
      "                    \"endInSentence\": 59\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"A consistent preference for saturation of available GA over GG dinucleotides within hypermutated sequences was also noted.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 63,\n",
      "                    \"endInSentence\": 76\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Indeed, the DNA-binding preferences of bisbenzamide for runs of AT, particularly AAAA and AATT, would suggest that GA to AA would be more efficiently detected than GG to AG, but the fact that the vast majority of hypermutants we recovered used both contexts suggests that this was not a major obstacle to recovery.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"bisbenzamide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 39,\n",
      "                    \"endInSentence\": 51\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The preferential utilization of available GA over GG dinucleotides over the whole dynamic range, and the fact that the vast majority of hypermutants utilized both GA and GG contexts, suggests that both contexts are recognized during a single reverse transcription event.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"GG dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 50,\n",
      "                    \"endInSentence\": 66\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The high error rate of HIV-1 RT is apparently necessary to permit both \\u201cnormal\\u201d mutation and hypermutation, but hypermutation should not be thought of as an elevated error rate for RT; it is clearly an independent process with separate biochemical requirements which must vary over a wide range during HIV-1 replication in vivo.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"RT\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 29,\n",
      "                    \"endInSentence\": 31\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"First, the evidence that true hypermutants were generated in CD4+ T cells in culture should be reiterated.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 61,\n",
      "                    \"endInSentence\": 64\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The virus stock was molecularly cloned and verified to be free of hypermutation by sequencing of the viral RNA by RT-PCR and by DNase treatment to remove any contaminating proviral DNA remaining from virus propagation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"DNase\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 128,\n",
      "                    \"endInSentence\": 133\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"DNA PCR of the culture supernatants after infection was negative.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 42,\n",
      "                    \"endInSentence\": 51\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The identity of the target cell population was carefully established; the culture contained principally, if not exclusively, CD4+ T cells.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 125,\n",
      "                    \"endInSentence\": 128\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Most, if not all, CD4+ T cells were quiescent.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 18,\n",
      "                    \"endInSentence\": 21\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"It has been previously documented that infection of quiescent cells results in the absence of virus progeny (68) and that viral DNA can be detected after infection of unstimulated cells although it is not associated with virus production (69, 70).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 39,\n",
      "                    \"endInSentence\": 48\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 154,\n",
      "                    \"endInSentence\": 163\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"In our experiments, proviral DNA was established but there was no detectable virus production in the absence of PHA stimulation, which is consistent with the infection of quiescent cells.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 158,\n",
      "                    \"endInSentence\": 167\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Most of the cells were HLA-DR negative.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HLA-DR\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 23,\n",
      "                    \"endInSentence\": 29\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"It is highly probable that those cells that were HLA-DR+, ranging from 15 to 18% in different experiments, represented recently, but not currently, activated cells.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HLA-DR\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 49,\n",
      "                    \"endInSentence\": 55\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Hypermutants were recovered principally from virus cultures in which HIV-1 infection was simultaneous with, or briefly preceded by, PHA stimulation (Fig. 6).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 69,\n",
      "                    \"endInSentence\": 84\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Even the dinucleotide context preference is partially parsed out by these experiments; we note that in culture conditions XI, I, and II, where PHA either was not added or was added along with or after the virus, utilization of available GA and GG was almost equivalent.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 9,\n",
      "                    \"endInSentence\": 21\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The strong bias for GA appeared once PHA preceded HIV-1 infection by only 3 h (Figure 7B).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 50,\n",
      "                    \"endInSentence\": 65\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"A previous study demonstrated that hypermutants were recovered, although at low frequencies, after infection of unstimulated (2%) and stimulated (0.9%) bulk PBMC (62).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 99,\n",
      "                    \"endInSentence\": 108\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Such recovery of hypermutants was ascribed to virus infection of a small proportion of PBMC with distorted dNTP pools, reinforcing the concept of an occasional and sporadic event.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"virus infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 46,\n",
      "                    \"endInSentence\": 61\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Because the composition of dNTP pools fluctuates as cells progress through the cell cycle (5, 20, 45), we investigated if the timing of HIV-1 infection with respect to T-cell stimulation could influence the generation of hypermutation.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 136,\n",
      "                    \"endInSentence\": 151\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Another aspect is the fact that the vast majority of hypermutated protease sequences present in this study were associated with a disrupted coding potential.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 66,\n",
      "                    \"endInSentence\": 74\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Since protease has a crucial role during the HIV-1 replication cycle, the loss of its coding potential integrity should lead to abortive infections.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 6,\n",
      "                    \"endInSentence\": 14\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infections\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 137,\n",
      "                    \"endInSentence\": 147\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 6,\n",
      "                    \"end1\": 14,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 137,\n",
      "                    \"end2\": 147,\n",
      "                    \"label2\": \"infections\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.02,\n",
      "                        \"Negative\": 0.01,\n",
      "                        \"Neutral\": 0.01,\n",
      "                        \"Altered Expression\": 0.0,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.02,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"They amplify a 297-bp segment of the HIV-1 genome encoding the protease gene, as indicated by their location in the reference sequence HXB2 (GenBank accession number AF033819).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 63,\n",
      "                    \"endInSentence\": 71\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"HIV-1 protease sequences amplified from patient PBMC and representing a range of G-to-A hypermutations were used to explore the performance of 1% agarose gels containing HA yellow, a dye that preferentially binds to AT-rich regions in DNA.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 6,\n",
      "                    \"endInSentence\": 14\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"HA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 170,\n",
      "                    \"endInSentence\": 172\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The GA and GG dinucleotides that are susceptible to hypermutation are shaded.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 14,\n",
      "                    \"endInSentence\": 27\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"The gel on the left shows the migration of these PCR products as a single band at 297 bp without HA yellow.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"HA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 97,\n",
      "                    \"endInSentence\": 99\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Primers were designed to incorporate G-to-A substitutions in GA or GG dinucleotides (Materials and Methods) and compared to primers matched to normal sequences for their ability to amplify hypermutated sequences from the PBMC of 10 patients.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"GG\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 67,\n",
      "                    \"endInSentence\": 69\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"dinucleotides\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 70,\n",
      "                    \"endInSentence\": 83\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Military Seroconverters1997\\u20131998AsymptomaticB147\\u2009(50) TanzaniaSentinel Surveillance1996AIDSA74\\u2009(57) C143\\u2009(21) D32\\u2009(67) Other22\\u2009(100) \\u2003Total5323\\u2009(43)aThe genetic subtype of the normal HIV-1 protease sequence was determined by phylogenetic analysis (data not shown).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 189,\n",
      "                    \"endInSentence\": 197\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Other, mixed infection or unclassified.\\u00a0bHypermutated HIV-1 sequences could be recovered from primary PBMC as described in Materials and Methods.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"mixed infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 7,\n",
      "                    \"endInSentence\": 22\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Products were visualized by staining with ethidium bromide.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"ethidium bromide\",\n",
      "                    \"type\": \"CD\",\n",
      "                    \"startInSentence\": 42,\n",
      "                    \"endInSentence\": 58\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"HIV-1 protease genes were amplified using DNA extracted from the PBMC of 53 patients ranging from the early, asymptomatic stage of HIV-1 infection to late-stage AIDS, and including infections with HIV-1 subtypes A through D (Table 2).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 6,\n",
      "                    \"endInSentence\": 14\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"HIV-1 infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 131,\n",
      "                    \"endInSentence\": 146\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"AIDS\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 161,\n",
      "                    \"endInSentence\": 165\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infections\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 181,\n",
      "                    \"endInSentence\": 191\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 6,\n",
      "                    \"end1\": 14,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 131,\n",
      "                    \"end2\": 146,\n",
      "                    \"label2\": \"HIV-1 infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.96,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.0,\n",
      "                        \"Negative\": 0.04,\n",
      "                        \"Neutral\": 0.93,\n",
      "                        \"Altered Expression\": 0.0,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.04,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 6,\n",
      "                    \"end1\": 14,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 161,\n",
      "                    \"end2\": 165,\n",
      "                    \"label2\": \"AIDS\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.96,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.0,\n",
      "                        \"Negative\": 0.04,\n",
      "                        \"Neutral\": 0.93,\n",
      "                        \"Altered Expression\": 0.0,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.04,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 6,\n",
      "                    \"end1\": 14,\n",
      "                    \"label1\": \"protease\",\n",
      "                    \"start2\": 181,\n",
      "                    \"end2\": 191,\n",
      "                    \"label2\": \"infections\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.96,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.0,\n",
      "                        \"Negative\": 0.04,\n",
      "                        \"Neutral\": 0.93,\n",
      "                        \"Altered Expression\": 0.0,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.04,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"CD4+ T cells isolated from the PBMC of normal donors were infected with an NL4-3 virus stock for 2 h at a multiplicity of infection of 0.02 (stippled gray bar) under varying conditions of PHA activation (gray bars).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"CD4\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 0,\n",
      "                    \"endInSentence\": 3\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 122,\n",
      "                    \"endInSentence\": 131\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 0,\n",
      "                    \"end1\": 3,\n",
      "                    \"label1\": \"CD4\",\n",
      "                    \"start2\": 122,\n",
      "                    \"end2\": 131,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.84,\n",
      "                        \"Negative\": 0.0,\n",
      "                        \"Neutral\": 0.0,\n",
      "                        \"Altered Expression\": 0.47,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.01,\n",
      "                        \"Regulatory modification\": 0.0\n",
      "                    }\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Condition I, infection without addition of PHA; condition II, simultaneous infection and PHA addition; conditions III to VIII, infection with prior PHA activation for 3 to 72 h respectively; condition IX, infection and addition of PHA 24 h later.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 13,\n",
      "                    \"endInSentence\": 22\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 75,\n",
      "                    \"endInSentence\": 84\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 127,\n",
      "                    \"endInSentence\": 136\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"PHA\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 148,\n",
      "                    \"endInSentence\": 151\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"infection\",\n",
      "                    \"type\": \"DS\",\n",
      "                    \"startInSentence\": 205,\n",
      "                    \"endInSentence\": 214\n",
      "                }\n",
      "            ],\n",
      "            \"co-occurrence\": [\n",
      "                {\n",
      "                    \"start1\": 148,\n",
      "                    \"end1\": 151,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 13,\n",
      "                    \"end2\": 22,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.09,\n",
      "                        \"Negative\": 0.01,\n",
      "                        \"Neutral\": 0.64,\n",
      "                        \"Altered Expression\": 0.08,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.01\n",
      "                    },\n",
      "                    \"relation\": [\n",
      "                        {\n",
      "                            \"startr\": 152,\n",
      "                            \"endr\": 162,\n",
      "                            \"labelr\": \"activation\",\n",
      "                            \"typer\": \"POSITIVE_REGULATION\"\n",
      "                        }\n",
      "                    ]\n",
      "                },\n",
      "                {\n",
      "                    \"start1\": 148,\n",
      "                    \"end1\": 151,\n",
      "                    \"label1\": \"PHA\",\n",
      "                    \"start2\": 205,\n",
      "                    \"end2\": 214,\n",
      "                    \"label2\": \"infection\",\n",
      "                    \"type\": \"GP-DS\",\n",
      "                    \"association\": {\n",
      "                        \"Yes\": 0.99,\n",
      "                        \"No\": 0.0,\n",
      "                        \"Positive\": 0.09,\n",
      "                        \"Negative\": 0.01,\n",
      "                        \"Neutral\": 0.64,\n",
      "                        \"Altered Expression\": 0.08,\n",
      "                        \"Genetic Variation\": 0.0,\n",
      "                        \"Any\": 0.0,\n",
      "                        \"Regulatory modification\": 0.01\n",
      "                    },\n",
      "                    \"relation\": [\n",
      "                        {\n",
      "                            \"startr\": 152,\n",
      "                            \"endr\": 162,\n",
      "                            \"labelr\": \"activation\",\n",
      "                            \"typer\": \"POSITIVE_REGULATION\"\n",
      "                        }\n",
      "                    ]\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Cultures were sampled at intervals and assayed for the presence of p24 antigen in the culture supernatant (solid lines) and for the presence of hypermutated HIV-1 DNA sequences.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"p24\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 67,\n",
      "                    \"endInSentence\": 70\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"Samples from time points at which hypermutation was rare (f, i, j, k, and l) are shown both as bulk PCR products (B) and after pre-enrichment by ScrF1 and AvaII digestion (A).\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"ScrF1\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 145,\n",
      "                    \"endInSentence\": 150\n",
      "                },\n",
      "                {\n",
      "                    \"label\": \"AvaII\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 155,\n",
      "                    \"endInSentence\": 160\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"text\": \"A total of 227 HIV-1 protease sequences, 131 of which were unique, were obtained from virus cultures.\",\n",
      "            \"matches\": [\n",
      "                {\n",
      "                    \"label\": \"protease\",\n",
      "                    \"type\": \"GP\",\n",
      "                    \"startInSentence\": 21,\n",
      "                    \"endInSentence\": 29\n",
      "                }\n",
      "            ]\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "json_object = json.dumps(json_generated, indent = 4)   \n",
    "print(json_object)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=EXT_ID%3A2109978%20AND%20SRC%3AMED&resultType=core&cursorMark=*&pageSize=25&format=json\n",
    "\n",
    "import requests\n",
    "query = '2109978'\n",
    "url = 'https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=EXT_ID:'+query+' AND SRC:MED&resultType=core&cursorMark=*&pageSize=25&format=json'\n",
    "response = requests.get(url)\n",
    "rjson = response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rjson['resultList']['result'][0]['abstractText']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xx_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xx_==yy_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FP Analysis on test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pmc_meta(query):\n",
    "        \n",
    "        url = 'https://www.ebi.ac.uk/europepmc/webservices/rest/search?'\n",
    "        params = {'query':query, 'resultType':'core','cursorMark':'*','pageSize':'25','format':'json'}\n",
    "        response = requests.get(url,params)\n",
    "        rjson = response.json()\n",
    "\n",
    "        for rslt in rjson['resultList']['result']:\n",
    "            pmeta = {}\n",
    "            pmeta['pmid'] = ''\n",
    "            pmeta['pmcid'] = ''\n",
    "            pmeta['mesh'] = []\n",
    "            pmeta['pmc_title'] = ''\n",
    "            pmeta['pmc_abstract']= ''\n",
    "            pmeta['text_urls'] = []\n",
    "            pmeta['oa'] =  ''\n",
    "            pmeta['authMan'] = ''\n",
    "            if 'pmid' in rslt.keys():\n",
    "                pmeta['pmid'] = rslt['pmid']\n",
    "                pmids.append(rslt['pmid'])                \n",
    "            if 'meshHeadingList' in rslt.keys():\n",
    "                for m in rslt['meshHeadingList']['meshHeading']:\n",
    "                    if 'meshQualifierList' in m.keys():\n",
    "                        for q in m['meshQualifierList']['meshQualifier']:\n",
    "                            pmeta['mesh'].append(m['descriptorName'])\n",
    "                            pmeta['mesh'].append(q['qualifierName'])\n",
    "                    else:\n",
    "                        pmeta['mesh'].append(m['descriptorName'])\n",
    "            if 'title' in rslt.keys(): pmeta['pmc_title'] = rslt['title']\n",
    "            if 'abstractText' in rslt.keys(): pmeta['pmc_abstract'] = rslt['abstractText']\n",
    "            if 'fullTextUrlList' in rslt.keys():            \n",
    "                for u in rslt['fullTextUrlList']['fullTextUrl']:\n",
    "                    pmeta['text_urls'].append(u['url'])\n",
    "            if 'isOpenAccess' in rslt.keys():pmeta['oa'] =  rslt['isOpenAccess']\n",
    "            if 'authMan' in rslt.keys():pmeta['authMan'] = rslt['authMan']\n",
    "            if 'pmcid' in rslt.keys() and rslt['pmcid']!='':\n",
    "                pmeta['pmcid'] = rslt['pmcid']\n",
    "                pmcids.append(rslt['pmcid'])\n",
    "                if rslt['isOpenAccess'] == 'Y' or rslt['authMan'] == 'Y':\n",
    "                #if rslt['isOpenAccess'] == 'Y':    \n",
    "                    oapids.append(rslt['pmcid'])                        \n",
    "            pmetas.append(pmeta)\n",
    "    except:\n",
    "        print('ERROR IN PMC ID:'+query)\n",
    "        pass\n",
    "    return [pmids,pmcids,oapids,pmetas]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "path_EPMC_text = '/mnt/droplet/nfs/gns/literature/machine-learning/evaluation/300articles/europePMC-NER/annotations_API/full_sentences/test_annotations/Europe_PMC_annotation.csv'\n",
    "colNames = ['pmc_id', 'section', 'sentence','ner'] \n",
    "    \n",
    "test_df = pd.read_csv(path_EPMC_text,sep ='\\t', names=colNames) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ast import literal_eval\n",
    "\n",
    "def deleting_epmc_GPS(list_1,del_name):\n",
    "   \n",
    "    for sub_list in list_1:\n",
    "        if del_name in sub_list:\n",
    "            list_1.remove(sub_list)\n",
    "    return list_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fuzzywuzzy import fuzz\n",
    "\n",
    "def remove_FP(epmc_list, ml_json):\n",
    "    all_ml_gps = []\n",
    "    if ml_json['annotations']:\n",
    "        for each_ml_annotation in ml_json['annotations']:\n",
    "            if each_ml_annotation[2] == 'GP':\n",
    "                all_ml_gps.append(each_ml_annotation[3])\n",
    "    \n",
    "    non_FP_removed =[]\n",
    "              \n",
    "\n",
    "    for each_ner in epmc_list:\n",
    "        if each_ner[2] == 'Gene_Proteins':\n",
    "            for each_ml_gp in all_ml_gps:         \n",
    "                score = fuzz.token_set_ratio(each_ml_gp, each_ner[1])\n",
    "                if score == 100:\n",
    "                    non_FP_removed.append(each_ner)\n",
    "\n",
    "    non_gp_tags =  deleting_epmc_GPS(epmc_list,'Gene_Proteins')  \n",
    "\n",
    "    fp_removed_tags = non_gp_tags+non_FP_removed\n",
    "\n",
    "    return fp_removed_tags\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spacy_annotations(text_sentence):\n",
    "    data_dict ={}\n",
    "    doc = nlp2(text_sentence)\n",
    "    terms_entities = []\n",
    "    \n",
    "    for ent in doc.ents:\n",
    "        terms_entities.append(\n",
    "            [ent.start_char, ent.end_char, ent.label_, ent.text])\n",
    "    \n",
    "    data_dict['annotations'] = terms_entities\n",
    "    \n",
    "    return data_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from tqdm import tqdm\n",
    "with open(result_path + 'spacy_fp_removal_80.tsv', 'w', newline='\\n') as f1:\n",
    "    public_writer = csv.writer(f1, delimiter='\\t', lineterminator='\\n')\n",
    "    \n",
    "    for index,row in tqdm(test_df.iterrows(),total = len(test_df)):\n",
    "        try:\n",
    "            ml_annotations = get_spacy_annotations(row['sentence'])\n",
    "            fp_removed = remove_FP(literal_eval(row['ner']), ml_annotations)\n",
    "        except ValueError:\n",
    "            fp_removed =''\n",
    "            \n",
    "        public_writer.writerow([row['pmc_id'], row['section'],row['sentence'], fp_removed])   \n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to IOB format\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "\n",
    "sys.path.append('/mnt/droplet/nfs/gns/literature/Santosh_Tirunagari/test Gitlab/epmc-ml-misc-library/')\n",
    "\n",
    "import capo_tools_lib\n",
    "import evaluation_epmc_lib\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iob_result_path = result_path+'iob/'\n",
    "pathlib.Path(iob_result_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "file_path = result_path + 'spacy_fp_removal_80.tsv'\n",
    "capo_tools_lib.annotations_api_tagged_sentences_to_IOB(file_path,\n",
    "                                                       iob_result_path,'spacy_fp_removal_iob.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import metrics.ner as ner_metrics\n",
    "\n",
    "\n",
    "# precision\t0.7\t0.7\t0.72\t0.73\n",
    "# recall\t0.53\t0.53\t0.54\t0.55\n",
    "# f1 score\t0.6\t0.6\t0.62\t0.6\n",
    "\n",
    "#\n",
    "# print(ner_metrics.semeval_scores_report(gold=epmc_labels, response=ml_labels, digits=2))\n",
    "\n",
    "root_path = '/mnt/droplet/nfs/gns/literature/machine-learning/'\n",
    "epmc_path = root_path+'Datasets/NER_Datasets/EBI_standard-IOB/test.csv'\n",
    "all_tags = ['GP', 'DS', 'OG']\n",
    "\n",
    "print('################ Annotation Pipeline Results ########################')\n",
    "CAPO_path = iob_result_path+'spacy_fp_removal_iob.tsv'\n",
    "for each_tag in all_tags:\n",
    "    print('############ '+each_tag+' ####################')\n",
    "    print('\\n')\n",
    "    print(ner_metrics.semeval_report(gold_path=epmc_path, response_path=CAPO_path, targets=[each_tag]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch",
   "language": "python",
   "name": "pytorch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
